// twenty-sdk/dist/front-component/index.mjs
import { isNull as e, isUndefined as t } from "@sniptt/guards";
import { v4 as n } from "uuid";
import { useEffect as r, useRef as i, useState as a } from "react";
//#region ../twenty-shared/dist/isDefined-Dtu5EYqP.mjs
var o = (n) => !t(n) && !e(n);
Object.freeze({ status: "aborted" });
function s(e, t, n) {
	function r(n, r) {
		var i;
		Object.defineProperty(n, "_zod", {
			value: n._zod ?? {},
			enumerable: !1
		}), (i = n._zod).traits ?? (i.traits = /* @__PURE__ */ new Set()), n._zod.traits.add(e), t(n, r);
		for (let e in o.prototype) e in n || Object.defineProperty(n, e, { value: o.prototype[e].bind(n) });
		n._zod.constr = o, n._zod.def = r;
	}
	let i = n?.Parent ?? Object;
	class a extends i {}
	Object.defineProperty(a, "name", { value: e });
	function o(e) {
		var t;
		let i = n?.Parent ? new a() : this;
		r(i, e), (t = i._zod).deferred ?? (t.deferred = []);
		for (let e of i._zod.deferred) e();
		return i;
	}
	return Object.defineProperty(o, "init", { value: r }), Object.defineProperty(o, Symbol.hasInstance, { value: (t) => n?.Parent && t instanceof n.Parent ? !0 : t?._zod?.traits?.has(e) }), Object.defineProperty(o, "name", { value: e }), o;
}
var c = class extends Error {
	constructor() {
		super("Encountered Promise during synchronous parse. Use .parseAsync() instead.");
	}
}, l = class extends Error {
	constructor(e) {
		super(`Encountered unidirectional transform during encode: ${e}`), this.name = "ZodEncodeError";
	}
}, u = {};
function d(e) {
	return e && Object.assign(u, e), u;
}
//#endregion
//#region ../../node_modules/zod/v4/core/util.js
function f(e) {
	let t = Object.values(e).filter((e) => typeof e == "number");
	return Object.entries(e).filter(([e, n]) => t.indexOf(+e) === -1).map(([e, t]) => t);
}
function p(e, t) {
	return typeof t == "bigint" ? t.toString() : t;
}
function m(e) {
	return { get value() {
		{
			let t = e();
			return Object.defineProperty(this, "value", { value: t }), t;
		}
		throw Error("cached value already set");
	} };
}
function h(e) {
	return e == null;
}
function g(e) {
	let t = +!!e.startsWith("^"), n = e.endsWith("$") ? e.length - 1 : e.length;
	return e.slice(t, n);
}
var ee = Symbol("evaluating");
function _(e, t, n) {
	let r;
	Object.defineProperty(e, t, {
		get() {
			if (r !== ee) return r === void 0 && (r = ee, r = n()), r;
		},
		set(n) {
			Object.defineProperty(e, t, { value: n });
		},
		configurable: !0
	});
}
function v(e, t, n) {
	Object.defineProperty(e, t, {
		value: n,
		writable: !0,
		enumerable: !0,
		configurable: !0
	});
}
function y(...e) {
	let t = {};
	for (let n of e) Object.assign(t, Object.getOwnPropertyDescriptors(n));
	return Object.defineProperties({}, t);
}
function te(e) {
	return JSON.stringify(e);
}
var ne = "captureStackTrace" in Error ? Error.captureStackTrace : (...e) => {};
function b(e) {
	return typeof e == "object" && !!e && !Array.isArray(e);
}
var re = m(() => {
	if (typeof navigator < "u" && navigator?.userAgent?.includes("Cloudflare")) return !1;
	try {
		return Function(""), !0;
	} catch {
		return !1;
	}
});
function x(e) {
	if (b(e) === !1) return !1;
	let t = e.constructor;
	if (t === void 0) return !0;
	let n = t.prototype;
	return !(b(n) === !1 || Object.prototype.hasOwnProperty.call(n, "isPrototypeOf") === !1);
}
function ie(e) {
	return x(e) ? { ...e } : Array.isArray(e) ? [...e] : e;
}
var ae = new Set([
	"string",
	"number",
	"symbol"
]);
function S(e) {
	return e.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}
function C(e, t, n) {
	let r = new e._zod.constr(t ?? e._zod.def);
	return (!t || n?.parent) && (r._zod.parent = e), r;
}
function w(e) {
	let t = e;
	if (!t) return {};
	if (typeof t == "string") return { error: () => t };
	if (t?.message !== void 0) {
		if (t?.error !== void 0) throw Error("Cannot specify both `message` and `error` params");
		t.error = t.message;
	}
	return delete t.message, typeof t.error == "string" ? {
		...t,
		error: () => t.error
	} : t;
}
function oe(e) {
	return Object.keys(e).filter((t) => e[t]._zod.optin === "optional" && e[t]._zod.optout === "optional");
}
-Number.MAX_VALUE, Number.MAX_VALUE;
function se(e, t) {
	let n = e._zod.def;
	return C(e, y(e._zod.def, {
		get shape() {
			let e = {};
			for (let r in t) {
				if (!(r in n.shape)) throw Error(`Unrecognized key: "${r}"`);
				t[r] && (e[r] = n.shape[r]);
			}
			return v(this, "shape", e), e;
		},
		checks: []
	}));
}
function ce(e, t) {
	let n = e._zod.def;
	return C(e, y(e._zod.def, {
		get shape() {
			let r = { ...e._zod.def.shape };
			for (let e in t) {
				if (!(e in n.shape)) throw Error(`Unrecognized key: "${e}"`);
				t[e] && delete r[e];
			}
			return v(this, "shape", r), r;
		},
		checks: []
	}));
}
function le(e, t) {
	if (!x(t)) throw Error("Invalid input to extend: expected a plain object");
	let n = e._zod.def.checks;
	if (n && n.length > 0) throw Error("Object schemas containing refinements cannot be extended. Use `.safeExtend()` instead.");
	return C(e, y(e._zod.def, {
		get shape() {
			let n = {
				...e._zod.def.shape,
				...t
			};
			return v(this, "shape", n), n;
		},
		checks: []
	}));
}
function ue(e, t) {
	if (!x(t)) throw Error("Invalid input to safeExtend: expected a plain object");
	return C(e, {
		...e._zod.def,
		get shape() {
			let n = {
				...e._zod.def.shape,
				...t
			};
			return v(this, "shape", n), n;
		},
		checks: e._zod.def.checks
	});
}
function de(e, t) {
	return C(e, y(e._zod.def, {
		get shape() {
			let n = {
				...e._zod.def.shape,
				...t._zod.def.shape
			};
			return v(this, "shape", n), n;
		},
		get catchall() {
			return t._zod.def.catchall;
		},
		checks: []
	}));
}
function fe(e, t, n) {
	return C(t, y(t._zod.def, {
		get shape() {
			let r = t._zod.def.shape, i = { ...r };
			if (n) for (let t in n) {
				if (!(t in r)) throw Error(`Unrecognized key: "${t}"`);
				n[t] && (i[t] = e ? new e({
					type: "optional",
					innerType: r[t]
				}) : r[t]);
			}
			else for (let t in r) i[t] = e ? new e({
				type: "optional",
				innerType: r[t]
			}) : r[t];
			return v(this, "shape", i), i;
		},
		checks: []
	}));
}
function pe(e, t, n) {
	return C(t, y(t._zod.def, {
		get shape() {
			let r = t._zod.def.shape, i = { ...r };
			if (n) for (let t in n) {
				if (!(t in i)) throw Error(`Unrecognized key: "${t}"`);
				n[t] && (i[t] = new e({
					type: "nonoptional",
					innerType: r[t]
				}));
			}
			else for (let t in r) i[t] = new e({
				type: "nonoptional",
				innerType: r[t]
			});
			return v(this, "shape", i), i;
		},
		checks: []
	}));
}
function T(e, t = 0) {
	if (e.aborted === !0) return !0;
	for (let n = t; n < e.issues.length; n++) if (e.issues[n]?.continue !== !0) return !0;
	return !1;
}
function me(e, t) {
	return t.map((t) => {
		var n;
		return (n = t).path ?? (n.path = []), t.path.unshift(e), t;
	});
}
function E(e) {
	return typeof e == "string" ? e : e?.message;
}
function D(e, t, n) {
	let r = {
		...e,
		path: e.path ?? []
	};
	return e.message || (r.message = E(e.inst?._zod.def?.error?.(e)) ?? E(t?.error?.(e)) ?? E(n.customError?.(e)) ?? E(n.localeError?.(e)) ?? "Invalid input"), delete r.inst, delete r.continue, t?.reportInput || delete r.input, r;
}
function he(e) {
	return Array.isArray(e) ? "array" : typeof e == "string" ? "string" : "unknown";
}
function O(...e) {
	let [t, n, r] = e;
	return typeof t == "string" ? {
		message: t,
		code: "custom",
		input: n,
		inst: r
	} : { ...t };
}
//#endregion
//#region ../../node_modules/zod/v4/core/errors.js
var ge = (e, t) => {
	e.name = "$ZodError", Object.defineProperty(e, "_zod", {
		value: e._zod,
		enumerable: !1
	}), Object.defineProperty(e, "issues", {
		value: t,
		enumerable: !1
	}), e.message = JSON.stringify(t, p, 2), Object.defineProperty(e, "toString", {
		value: () => e.message,
		enumerable: !1
	});
}, _e = s("$ZodError", ge), ve = s("$ZodError", ge, { Parent: Error });
function ye(e, t = (e) => e.message) {
	let n = {}, r = [];
	for (let i of e.issues) i.path.length > 0 ? (n[i.path[0]] = n[i.path[0]] || [], n[i.path[0]].push(t(i))) : r.push(t(i));
	return {
		formErrors: r,
		fieldErrors: n
	};
}
function be(e, t) {
	let n = t || function(e) {
		return e.message;
	}, r = { _errors: [] }, i = (e) => {
		for (let t of e.issues) if (t.code === "invalid_union" && t.errors.length) t.errors.map((e) => i({ issues: e }));
		else if (t.code === "invalid_key") i({ issues: t.issues });
		else if (t.code === "invalid_element") i({ issues: t.issues });
		else if (t.path.length === 0) r._errors.push(n(t));
		else {
			let e = r, i = 0;
			for (; i < t.path.length;) {
				let r = t.path[i];
				i === t.path.length - 1 ? (e[r] = e[r] || { _errors: [] }, e[r]._errors.push(n(t))) : e[r] = e[r] || { _errors: [] }, e = e[r], i++;
			}
		}
	};
	return i(e), r;
}
//#endregion
//#region ../../node_modules/zod/v4/core/parse.js
var k = (e) => (t, n, r, i) => {
	let a = r ? Object.assign(r, { async: !1 }) : { async: !1 }, o = t._zod.run({
		value: n,
		issues: []
	}, a);
	if (o instanceof Promise) throw new c();
	if (o.issues.length) {
		let t = new (i?.Err ?? e)(o.issues.map((e) => D(e, a, d())));
		throw ne(t, i?.callee), t;
	}
	return o.value;
}, A = (e) => async (t, n, r, i) => {
	let a = r ? Object.assign(r, { async: !0 }) : { async: !0 }, o = t._zod.run({
		value: n,
		issues: []
	}, a);
	if (o instanceof Promise && (o = await o), o.issues.length) {
		let t = new (i?.Err ?? e)(o.issues.map((e) => D(e, a, d())));
		throw ne(t, i?.callee), t;
	}
	return o.value;
}, j = (e) => (t, n, r) => {
	let i = r ? {
		...r,
		async: !1
	} : { async: !1 }, a = t._zod.run({
		value: n,
		issues: []
	}, i);
	if (a instanceof Promise) throw new c();
	return a.issues.length ? {
		success: !1,
		error: new (e ?? _e)(a.issues.map((e) => D(e, i, d())))
	} : {
		success: !0,
		data: a.value
	};
}, xe = /* @__PURE__*/ j(ve), M = (e) => async (t, n, r) => {
	let i = r ? Object.assign(r, { async: !0 }) : { async: !0 }, a = t._zod.run({
		value: n,
		issues: []
	}, i);
	return a instanceof Promise && (a = await a), a.issues.length ? {
		success: !1,
		error: new e(a.issues.map((e) => D(e, i, d())))
	} : {
		success: !0,
		data: a.value
	};
}, Se = /* @__PURE__*/ M(ve), Ce = (e) => (t, n, r) => {
	let i = r ? Object.assign(r, { direction: "backward" }) : { direction: "backward" };
	return k(e)(t, n, i);
}, we = (e) => (t, n, r) => k(e)(t, n, r), Te = (e) => async (t, n, r) => {
	let i = r ? Object.assign(r, { direction: "backward" }) : { direction: "backward" };
	return A(e)(t, n, i);
}, Ee = (e) => async (t, n, r) => A(e)(t, n, r), De = (e) => (t, n, r) => {
	let i = r ? Object.assign(r, { direction: "backward" }) : { direction: "backward" };
	return j(e)(t, n, i);
}, Oe = (e) => (t, n, r) => j(e)(t, n, r), ke = (e) => async (t, n, r) => {
	let i = r ? Object.assign(r, { direction: "backward" }) : { direction: "backward" };
	return M(e)(t, n, i);
}, Ae = (e) => async (t, n, r) => M(e)(t, n, r), je = /^[cC][^\s-]{8,}$/, Me = /^[0-9a-z]+$/, Ne = /^[0-9A-HJKMNP-TV-Za-hjkmnp-tv-z]{26}$/, Pe = /^[0-9a-vA-V]{20}$/, Fe = /^[A-Za-z0-9]{27}$/, Ie = /^[a-zA-Z0-9_-]{21}$/, Le = /^P(?:(\d+W)|(?!.*W)(?=\d|T\d)(\d+Y)?(\d+M)?(\d+D)?(T(?=\d)(\d+H)?(\d+M)?(\d+([.,]\d+)?S)?)?)$/, Re = /^([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12})$/, ze = (e) => e ? RegExp(`^([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-${e}[0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12})$`) : /^([0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[1-8][0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}|00000000-0000-0000-0000-000000000000|ffffffff-ffff-ffff-ffff-ffffffffffff)$/, Be = /^(?!\.)(?!.*\.\.)([A-Za-z0-9_'+\-\.]*)[A-Za-z0-9_+-]@([A-Za-z0-9][A-Za-z0-9\-]*\.)+[A-Za-z]{2,}$/, Ve = "^(\\p{Extended_Pictographic}|\\p{Emoji_Component})+$";
function He() {
	return new RegExp(Ve, "u");
}
var Ue = /^(?:(?:25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9][0-9]|[0-9])\.){3}(?:25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9][0-9]|[0-9])$/, We = /^(([0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:))$/, Ge = /^((25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9][0-9]|[0-9])\.){3}(25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9][0-9]|[0-9])\/([0-9]|[1-2][0-9]|3[0-2])$/, Ke = /^(([0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}|::|([0-9a-fA-F]{1,4})?::([0-9a-fA-F]{1,4}:?){0,6})\/(12[0-8]|1[01][0-9]|[1-9]?[0-9])$/, qe = /^$|^(?:[0-9a-zA-Z+/]{4})*(?:(?:[0-9a-zA-Z+/]{2}==)|(?:[0-9a-zA-Z+/]{3}=))?$/, Je = /^[A-Za-z0-9_-]*$/, Ye = /^(?=.{1,253}\.?$)[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[-0-9a-zA-Z]{0,61}[0-9a-zA-Z])?)*\.?$/, Xe = /^\+(?:[0-9]){6,14}[0-9]$/, Ze = "(?:(?:\\d\\d[2468][048]|\\d\\d[13579][26]|\\d\\d0[48]|[02468][048]00|[13579][26]00)-02-29|\\d{4}-(?:(?:0[13578]|1[02])-(?:0[1-9]|[12]\\d|3[01])|(?:0[469]|11)-(?:0[1-9]|[12]\\d|30)|(?:02)-(?:0[1-9]|1\\d|2[0-8])))", Qe = /*@__PURE__*/ RegExp(`^${Ze}$`);
function $e(e) {
	let t = "(?:[01]\\d|2[0-3]):[0-5]\\d";
	return typeof e.precision == "number" ? e.precision === -1 ? `${t}` : e.precision === 0 ? `${t}:[0-5]\\d` : `${t}:[0-5]\\d\\.\\d{${e.precision}}` : `${t}(?::[0-5]\\d(?:\\.\\d+)?)?`;
}
function et(e) {
	return RegExp(`^${$e(e)}$`);
}
function tt(e) {
	let t = $e({ precision: e.precision }), n = ["Z"];
	e.local && n.push(""), e.offset && n.push("([+-](?:[01]\\d|2[0-3]):[0-5]\\d)");
	let r = `${t}(?:${n.join("|")})`;
	return RegExp(`^${Ze}T(?:${r})$`);
}
var nt = (e) => {
	let t = e ? `[\\s\\S]{${e?.minimum ?? 0},${e?.maximum ?? ""}}` : "[\\s\\S]*";
	return RegExp(`^${t}$`);
}, rt = /^[^A-Z]*$/, it = /^[^a-z]*$/, N = /*@__PURE__*/ s("$ZodCheck", (e, t) => {
	var n;
	e._zod ??= {}, e._zod.def = t, (n = e._zod).onattach ?? (n.onattach = []);
}), at = /*@__PURE__*/ s("$ZodCheckMaxLength", (e, t) => {
	var n;
	N.init(e, t), (n = e._zod.def).when ?? (n.when = (e) => {
		let t = e.value;
		return !h(t) && t.length !== void 0;
	}), e._zod.onattach.push((e) => {
		let n = e._zod.bag.maximum ?? Infinity;
		t.maximum < n && (e._zod.bag.maximum = t.maximum);
	}), e._zod.check = (n) => {
		let r = n.value;
		if (r.length <= t.maximum) return;
		let i = he(r);
		n.issues.push({
			origin: i,
			code: "too_big",
			maximum: t.maximum,
			inclusive: !0,
			input: r,
			inst: e,
			continue: !t.abort
		});
	};
}), ot = /*@__PURE__*/ s("$ZodCheckMinLength", (e, t) => {
	var n;
	N.init(e, t), (n = e._zod.def).when ?? (n.when = (e) => {
		let t = e.value;
		return !h(t) && t.length !== void 0;
	}), e._zod.onattach.push((e) => {
		let n = e._zod.bag.minimum ?? -Infinity;
		t.minimum > n && (e._zod.bag.minimum = t.minimum);
	}), e._zod.check = (n) => {
		let r = n.value;
		if (r.length >= t.minimum) return;
		let i = he(r);
		n.issues.push({
			origin: i,
			code: "too_small",
			minimum: t.minimum,
			inclusive: !0,
			input: r,
			inst: e,
			continue: !t.abort
		});
	};
}), st = /*@__PURE__*/ s("$ZodCheckLengthEquals", (e, t) => {
	var n;
	N.init(e, t), (n = e._zod.def).when ?? (n.when = (e) => {
		let t = e.value;
		return !h(t) && t.length !== void 0;
	}), e._zod.onattach.push((e) => {
		let n = e._zod.bag;
		n.minimum = t.length, n.maximum = t.length, n.length = t.length;
	}), e._zod.check = (n) => {
		let r = n.value, i = r.length;
		if (i === t.length) return;
		let a = he(r), o = i > t.length;
		n.issues.push({
			origin: a,
			...o ? {
				code: "too_big",
				maximum: t.length
			} : {
				code: "too_small",
				minimum: t.length
			},
			inclusive: !0,
			exact: !0,
			input: n.value,
			inst: e,
			continue: !t.abort
		});
	};
}), P = /*@__PURE__*/ s("$ZodCheckStringFormat", (e, t) => {
	var n, r;
	N.init(e, t), e._zod.onattach.push((e) => {
		let n = e._zod.bag;
		n.format = t.format, t.pattern && (n.patterns ??= /* @__PURE__ */ new Set(), n.patterns.add(t.pattern));
	}), t.pattern ? (n = e._zod).check ?? (n.check = (n) => {
		t.pattern.lastIndex = 0, !t.pattern.test(n.value) && n.issues.push({
			origin: "string",
			code: "invalid_format",
			format: t.format,
			input: n.value,
			...t.pattern ? { pattern: t.pattern.toString() } : {},
			inst: e,
			continue: !t.abort
		});
	}) : (r = e._zod).check ?? (r.check = () => {});
}), ct = /*@__PURE__*/ s("$ZodCheckRegex", (e, t) => {
	P.init(e, t), e._zod.check = (n) => {
		t.pattern.lastIndex = 0, !t.pattern.test(n.value) && n.issues.push({
			origin: "string",
			code: "invalid_format",
			format: "regex",
			input: n.value,
			pattern: t.pattern.toString(),
			inst: e,
			continue: !t.abort
		});
	};
}), lt = /*@__PURE__*/ s("$ZodCheckLowerCase", (e, t) => {
	t.pattern ??= rt, P.init(e, t);
}), ut = /*@__PURE__*/ s("$ZodCheckUpperCase", (e, t) => {
	t.pattern ??= it, P.init(e, t);
}), dt = /*@__PURE__*/ s("$ZodCheckIncludes", (e, t) => {
	N.init(e, t);
	let n = S(t.includes), r = new RegExp(typeof t.position == "number" ? `^.{${t.position}}${n}` : n);
	t.pattern = r, e._zod.onattach.push((e) => {
		let t = e._zod.bag;
		t.patterns ??= /* @__PURE__ */ new Set(), t.patterns.add(r);
	}), e._zod.check = (n) => {
		n.value.includes(t.includes, t.position) || n.issues.push({
			origin: "string",
			code: "invalid_format",
			format: "includes",
			includes: t.includes,
			input: n.value,
			inst: e,
			continue: !t.abort
		});
	};
}), ft = /*@__PURE__*/ s("$ZodCheckStartsWith", (e, t) => {
	N.init(e, t);
	let n = RegExp(`^${S(t.prefix)}.*`);
	t.pattern ??= n, e._zod.onattach.push((e) => {
		let t = e._zod.bag;
		t.patterns ??= /* @__PURE__ */ new Set(), t.patterns.add(n);
	}), e._zod.check = (n) => {
		n.value.startsWith(t.prefix) || n.issues.push({
			origin: "string",
			code: "invalid_format",
			format: "starts_with",
			prefix: t.prefix,
			input: n.value,
			inst: e,
			continue: !t.abort
		});
	};
}), pt = /*@__PURE__*/ s("$ZodCheckEndsWith", (e, t) => {
	N.init(e, t);
	let n = RegExp(`.*${S(t.suffix)}$`);
	t.pattern ??= n, e._zod.onattach.push((e) => {
		let t = e._zod.bag;
		t.patterns ??= /* @__PURE__ */ new Set(), t.patterns.add(n);
	}), e._zod.check = (n) => {
		n.value.endsWith(t.suffix) || n.issues.push({
			origin: "string",
			code: "invalid_format",
			format: "ends_with",
			suffix: t.suffix,
			input: n.value,
			inst: e,
			continue: !t.abort
		});
	};
}), mt = /*@__PURE__*/ s("$ZodCheckOverwrite", (e, t) => {
	N.init(e, t), e._zod.check = (e) => {
		e.value = t.tx(e.value);
	};
}), ht = class {
	constructor(e = []) {
		this.content = [], this.indent = 0, this && (this.args = e);
	}
	indented(e) {
		this.indent += 1, e(this), --this.indent;
	}
	write(e) {
		if (typeof e == "function") {
			e(this, { execution: "sync" }), e(this, { execution: "async" });
			return;
		}
		let t = e.split("\n").filter((e) => e), n = Math.min(...t.map((e) => e.length - e.trimStart().length)), r = t.map((e) => e.slice(n)).map((e) => " ".repeat(this.indent * 2) + e);
		for (let e of r) this.content.push(e);
	}
	compile() {
		let e = Function, t = this?.args, n = [...(this?.content ?? [""]).map((e) => `  ${e}`)];
		return new e(...t, n.join("\n"));
	}
}, gt = {
	major: 4,
	minor: 1,
	patch: 11
}, F = /*@__PURE__*/ s("$ZodType", (e, t) => {
	var n;
	e ??= {}, e._zod.def = t, e._zod.bag = e._zod.bag || {}, e._zod.version = gt;
	let r = [...e._zod.def.checks ?? []];
	e._zod.traits.has("$ZodCheck") && r.unshift(e);
	for (let t of r) for (let n of t._zod.onattach) n(e);
	if (r.length === 0) (n = e._zod).deferred ?? (n.deferred = []), e._zod.deferred?.push(() => {
		e._zod.run = e._zod.parse;
	});
	else {
		let t = (e, t, n) => {
			let r = T(e), i;
			for (let a of t) {
				if (a._zod.def.when) {
					if (!a._zod.def.when(e)) continue;
				} else if (r) continue;
				let t = e.issues.length, o = a._zod.check(e);
				if (o instanceof Promise && n?.async === !1) throw new c();
				if (i || o instanceof Promise) i = (i ?? Promise.resolve()).then(async () => {
					await o, e.issues.length !== t && (r ||= T(e, t));
				});
				else {
					if (e.issues.length === t) continue;
					r ||= T(e, t);
				}
			}
			return i ? i.then(() => e) : e;
		}, n = (n, i, a) => {
			if (T(n)) return n.aborted = !0, n;
			let o = t(i, r, a);
			if (o instanceof Promise) {
				if (a.async === !1) throw new c();
				return o.then((t) => e._zod.parse(t, a));
			}
			return e._zod.parse(o, a);
		};
		e._zod.run = (i, a) => {
			if (a.skipChecks) return e._zod.parse(i, a);
			if (a.direction === "backward") {
				let t = e._zod.parse({
					value: i.value,
					issues: []
				}, {
					...a,
					skipChecks: !0
				});
				return t instanceof Promise ? t.then((e) => n(e, i, a)) : n(t, i, a);
			}
			let o = e._zod.parse(i, a);
			if (o instanceof Promise) {
				if (a.async === !1) throw new c();
				return o.then((e) => t(e, r, a));
			}
			return t(o, r, a);
		};
	}
	e["~standard"] = {
		validate: (t) => {
			try {
				let n = xe(e, t);
				return n.success ? { value: n.data } : { issues: n.error?.issues };
			} catch {
				return Se(e, t).then((e) => e.success ? { value: e.data } : { issues: e.error?.issues });
			}
		},
		vendor: "zod",
		version: 1
	};
}), I = /*@__PURE__*/ s("$ZodString", (e, t) => {
	F.init(e, t), e._zod.pattern = [...e?._zod.bag?.patterns ?? []].pop() ?? nt(e._zod.bag), e._zod.parse = (n, r) => {
		if (t.coerce) try {
			n.value = String(n.value);
		} catch {}
		return typeof n.value == "string" || n.issues.push({
			expected: "string",
			code: "invalid_type",
			input: n.value,
			inst: e
		}), n;
	};
}), L = /*@__PURE__*/ s("$ZodStringFormat", (e, t) => {
	P.init(e, t), I.init(e, t);
}), _t = /*@__PURE__*/ s("$ZodGUID", (e, t) => {
	t.pattern ??= Re, L.init(e, t);
}), vt = /*@__PURE__*/ s("$ZodUUID", (e, t) => {
	if (t.version) {
		let e = {
			v1: 1,
			v2: 2,
			v3: 3,
			v4: 4,
			v5: 5,
			v6: 6,
			v7: 7,
			v8: 8
		}[t.version];
		if (e === void 0) throw Error(`Invalid UUID version: "${t.version}"`);
		t.pattern ??= ze(e);
	} else t.pattern ??= ze();
	L.init(e, t);
}), yt = /*@__PURE__*/ s("$ZodEmail", (e, t) => {
	t.pattern ??= Be, L.init(e, t);
}), bt = /*@__PURE__*/ s("$ZodURL", (e, t) => {
	L.init(e, t), e._zod.check = (n) => {
		try {
			let r = n.value.trim(), i = new URL(r);
			t.hostname && (t.hostname.lastIndex = 0, t.hostname.test(i.hostname) || n.issues.push({
				code: "invalid_format",
				format: "url",
				note: "Invalid hostname",
				pattern: Ye.source,
				input: n.value,
				inst: e,
				continue: !t.abort
			})), t.protocol && (t.protocol.lastIndex = 0, t.protocol.test(i.protocol.endsWith(":") ? i.protocol.slice(0, -1) : i.protocol) || n.issues.push({
				code: "invalid_format",
				format: "url",
				note: "Invalid protocol",
				pattern: t.protocol.source,
				input: n.value,
				inst: e,
				continue: !t.abort
			})), t.normalize ? n.value = i.href : n.value = r;
			return;
		} catch {
			n.issues.push({
				code: "invalid_format",
				format: "url",
				input: n.value,
				inst: e,
				continue: !t.abort
			});
		}
	};
}), xt = /*@__PURE__*/ s("$ZodEmoji", (e, t) => {
	t.pattern ??= He(), L.init(e, t);
}), St = /*@__PURE__*/ s("$ZodNanoID", (e, t) => {
	t.pattern ??= Ie, L.init(e, t);
}), Ct = /*@__PURE__*/ s("$ZodCUID", (e, t) => {
	t.pattern ??= je, L.init(e, t);
}), wt = /*@__PURE__*/ s("$ZodCUID2", (e, t) => {
	t.pattern ??= Me, L.init(e, t);
}), Tt = /*@__PURE__*/ s("$ZodULID", (e, t) => {
	t.pattern ??= Ne, L.init(e, t);
}), Et = /*@__PURE__*/ s("$ZodXID", (e, t) => {
	t.pattern ??= Pe, L.init(e, t);
}), Dt = /*@__PURE__*/ s("$ZodKSUID", (e, t) => {
	t.pattern ??= Fe, L.init(e, t);
}), Ot = /*@__PURE__*/ s("$ZodISODateTime", (e, t) => {
	t.pattern ??= tt(t), L.init(e, t);
}), kt = /*@__PURE__*/ s("$ZodISODate", (e, t) => {
	t.pattern ??= Qe, L.init(e, t);
}), At = /*@__PURE__*/ s("$ZodISOTime", (e, t) => {
	t.pattern ??= et(t), L.init(e, t);
}), jt = /*@__PURE__*/ s("$ZodISODuration", (e, t) => {
	t.pattern ??= Le, L.init(e, t);
}), Mt = /*@__PURE__*/ s("$ZodIPv4", (e, t) => {
	t.pattern ??= Ue, L.init(e, t), e._zod.onattach.push((e) => {
		let t = e._zod.bag;
		t.format = "ipv4";
	});
}), Nt = /*@__PURE__*/ s("$ZodIPv6", (e, t) => {
	t.pattern ??= We, L.init(e, t), e._zod.onattach.push((e) => {
		let t = e._zod.bag;
		t.format = "ipv6";
	}), e._zod.check = (n) => {
		try {
			new URL(`http://[${n.value}]`);
		} catch {
			n.issues.push({
				code: "invalid_format",
				format: "ipv6",
				input: n.value,
				inst: e,
				continue: !t.abort
			});
		}
	};
}), Pt = /*@__PURE__*/ s("$ZodCIDRv4", (e, t) => {
	t.pattern ??= Ge, L.init(e, t);
}), Ft = /*@__PURE__*/ s("$ZodCIDRv6", (e, t) => {
	t.pattern ??= Ke, L.init(e, t), e._zod.check = (n) => {
		let r = n.value.split("/");
		try {
			if (r.length !== 2) throw Error();
			let [e, t] = r;
			if (!t) throw Error();
			let n = Number(t);
			if (`${n}` !== t || n < 0 || n > 128) throw Error();
			new URL(`http://[${e}]`);
		} catch {
			n.issues.push({
				code: "invalid_format",
				format: "cidrv6",
				input: n.value,
				inst: e,
				continue: !t.abort
			});
		}
	};
});
function It(e) {
	if (e === "") return !0;
	if (e.length % 4 != 0) return !1;
	try {
		return atob(e), !0;
	} catch {
		return !1;
	}
}
var Lt = /*@__PURE__*/ s("$ZodBase64", (e, t) => {
	t.pattern ??= qe, L.init(e, t), e._zod.onattach.push((e) => {
		e._zod.bag.contentEncoding = "base64";
	}), e._zod.check = (n) => {
		It(n.value) || n.issues.push({
			code: "invalid_format",
			format: "base64",
			input: n.value,
			inst: e,
			continue: !t.abort
		});
	};
});
function Rt(e) {
	if (!Je.test(e)) return !1;
	let t = e.replace(/[-_]/g, (e) => e === "-" ? "+" : "/");
	return It(t.padEnd(Math.ceil(t.length / 4) * 4, "="));
}
var zt = /*@__PURE__*/ s("$ZodBase64URL", (e, t) => {
	t.pattern ??= Je, L.init(e, t), e._zod.onattach.push((e) => {
		e._zod.bag.contentEncoding = "base64url";
	}), e._zod.check = (n) => {
		Rt(n.value) || n.issues.push({
			code: "invalid_format",
			format: "base64url",
			input: n.value,
			inst: e,
			continue: !t.abort
		});
	};
}), Bt = /*@__PURE__*/ s("$ZodE164", (e, t) => {
	t.pattern ??= Xe, L.init(e, t);
});
function Vt(e, t = null) {
	try {
		let n = e.split(".");
		if (n.length !== 3) return !1;
		let [r] = n;
		if (!r) return !1;
		let i = JSON.parse(atob(r));
		return !("typ" in i && i?.typ !== "JWT" || !i.alg || t && (!("alg" in i) || i.alg !== t));
	} catch {
		return !1;
	}
}
var Ht = /*@__PURE__*/ s("$ZodJWT", (e, t) => {
	L.init(e, t), e._zod.check = (n) => {
		Vt(n.value, t.alg) || n.issues.push({
			code: "invalid_format",
			format: "jwt",
			input: n.value,
			inst: e,
			continue: !t.abort
		});
	};
}), Ut = /*@__PURE__*/ s("$ZodUnknown", (e, t) => {
	F.init(e, t), e._zod.parse = (e) => e;
}), Wt = /*@__PURE__*/ s("$ZodNever", (e, t) => {
	F.init(e, t), e._zod.parse = (t, n) => (t.issues.push({
		expected: "never",
		code: "invalid_type",
		input: t.value,
		inst: e
	}), t);
});
function Gt(e, t, n) {
	e.issues.length && t.issues.push(...me(n, e.issues)), t.value[n] = e.value;
}
var Kt = /*@__PURE__*/ s("$ZodArray", (e, t) => {
	F.init(e, t), e._zod.parse = (n, r) => {
		let i = n.value;
		if (!Array.isArray(i)) return n.issues.push({
			expected: "array",
			code: "invalid_type",
			input: i,
			inst: e
		}), n;
		n.value = Array(i.length);
		let a = [];
		for (let e = 0; e < i.length; e++) {
			let o = i[e], s = t.element._zod.run({
				value: o,
				issues: []
			}, r);
			s instanceof Promise ? a.push(s.then((t) => Gt(t, n, e))) : Gt(s, n, e);
		}
		return a.length ? Promise.all(a).then(() => n) : n;
	};
});
function R(e, t, n, r) {
	e.issues.length && t.issues.push(...me(n, e.issues)), e.value === void 0 ? n in r && (t.value[n] = void 0) : t.value[n] = e.value;
}
function qt(e) {
	let t = Object.keys(e.shape);
	for (let n of t) if (!e.shape?.[n]?._zod?.traits?.has("$ZodType")) throw Error(`Invalid element at key "${n}": expected a Zod schema`);
	let n = oe(e.shape);
	return {
		...e,
		keys: t,
		keySet: new Set(t),
		numKeys: t.length,
		optionalKeys: new Set(n)
	};
}
function Jt(e, t, n, r, i, a) {
	let o = [], s = i.keySet, c = i.catchall._zod, l = c.def.type;
	for (let i of Object.keys(t)) {
		if (s.has(i)) continue;
		if (l === "never") {
			o.push(i);
			continue;
		}
		let a = c.run({
			value: t[i],
			issues: []
		}, r);
		a instanceof Promise ? e.push(a.then((e) => R(e, n, i, t))) : R(a, n, i, t);
	}
	return o.length && n.issues.push({
		code: "unrecognized_keys",
		keys: o,
		input: t,
		inst: a
	}), e.length ? Promise.all(e).then(() => n) : n;
}
var Yt = /*@__PURE__*/ s("$ZodObject", (e, t) => {
	if (F.init(e, t), !Object.getOwnPropertyDescriptor(t, "shape")?.get) {
		let e = t.shape;
		Object.defineProperty(t, "shape", { get: () => {
			let n = { ...e };
			return Object.defineProperty(t, "shape", { value: n }), n;
		} });
	}
	let n = m(() => qt(t));
	_(e._zod, "propValues", () => {
		let e = t.shape, n = {};
		for (let t in e) {
			let r = e[t]._zod;
			if (r.values) {
				n[t] ?? (n[t] = /* @__PURE__ */ new Set());
				for (let e of r.values) n[t].add(e);
			}
		}
		return n;
	});
	let r = b, i = t.catchall, a;
	e._zod.parse = (t, o) => {
		a ??= n.value;
		let s = t.value;
		if (!r(s)) return t.issues.push({
			expected: "object",
			code: "invalid_type",
			input: s,
			inst: e
		}), t;
		t.value = {};
		let c = [], l = a.shape;
		for (let e of a.keys) {
			let n = l[e]._zod.run({
				value: s[e],
				issues: []
			}, o);
			n instanceof Promise ? c.push(n.then((n) => R(n, t, e, s))) : R(n, t, e, s);
		}
		return i ? Jt(c, s, t, o, n.value, e) : c.length ? Promise.all(c).then(() => t) : t;
	};
}), Xt = /*@__PURE__*/ s("$ZodObjectJIT", (e, t) => {
	Yt.init(e, t);
	let n = e._zod.parse, r = m(() => qt(t)), i = (e) => {
		let t = new ht([
			"shape",
			"payload",
			"ctx"
		]), n = r.value, i = (e) => {
			let t = te(e);
			return `shape[${t}]._zod.run({ value: input[${t}], issues: [] }, ctx)`;
		};
		t.write("const input = payload.value;");
		let a = Object.create(null), o = 0;
		for (let e of n.keys) a[e] = `key_${o++}`;
		t.write("const newResult = {};");
		for (let e of n.keys) {
			let n = a[e], r = te(e);
			t.write(`const ${n} = ${i(e)};`), t.write(`
        if (${n}.issues.length) {
          payload.issues = payload.issues.concat(${n}.issues.map(iss => ({
            ...iss,
            path: iss.path ? [${r}, ...iss.path] : [${r}]
          })));
        }
        
        
        if (${n}.value === undefined) {
          if (${r} in input) {
            newResult[${r}] = undefined;
          }
        } else {
          newResult[${r}] = ${n}.value;
        }
        
      `);
		}
		t.write("payload.value = newResult;"), t.write("return payload;");
		let s = t.compile();
		return (t, n) => s(e, t, n);
	}, a, o = b, s = !u.jitless, c = s && re.value, l = t.catchall, d;
	e._zod.parse = (u, f) => {
		d ??= r.value;
		let p = u.value;
		return o(p) ? s && c && f?.async === !1 && f.jitless !== !0 ? (a ||= i(t.shape), u = a(u, f), l ? Jt([], p, u, f, d, e) : u) : n(u, f) : (u.issues.push({
			expected: "object",
			code: "invalid_type",
			input: p,
			inst: e
		}), u);
	};
});
function Zt(e, t, n, r) {
	for (let n of e) if (n.issues.length === 0) return t.value = n.value, t;
	let i = e.filter((e) => !T(e));
	return i.length === 1 ? (t.value = i[0].value, i[0]) : (t.issues.push({
		code: "invalid_union",
		input: t.value,
		inst: n,
		errors: e.map((e) => e.issues.map((e) => D(e, r, d())))
	}), t);
}
var Qt = /*@__PURE__*/ s("$ZodUnion", (e, t) => {
	F.init(e, t), _(e._zod, "optin", () => t.options.some((e) => e._zod.optin === "optional") ? "optional" : void 0), _(e._zod, "optout", () => t.options.some((e) => e._zod.optout === "optional") ? "optional" : void 0), _(e._zod, "values", () => {
		if (t.options.every((e) => e._zod.values)) return new Set(t.options.flatMap((e) => Array.from(e._zod.values)));
	}), _(e._zod, "pattern", () => {
		if (t.options.every((e) => e._zod.pattern)) {
			let e = t.options.map((e) => e._zod.pattern);
			return RegExp(`^(${e.map((e) => g(e.source)).join("|")})$`);
		}
	});
	let n = t.options.length === 1, r = t.options[0]._zod.run;
	e._zod.parse = (i, a) => {
		if (n) return r(i, a);
		let o = !1, s = [];
		for (let e of t.options) {
			let t = e._zod.run({
				value: i.value,
				issues: []
			}, a);
			if (t instanceof Promise) s.push(t), o = !0;
			else {
				if (t.issues.length === 0) return t;
				s.push(t);
			}
		}
		return o ? Promise.all(s).then((t) => Zt(t, i, e, a)) : Zt(s, i, e, a);
	};
}), $t = /*@__PURE__*/ s("$ZodIntersection", (e, t) => {
	F.init(e, t), e._zod.parse = (e, n) => {
		let r = e.value, i = t.left._zod.run({
			value: r,
			issues: []
		}, n), a = t.right._zod.run({
			value: r,
			issues: []
		}, n);
		return i instanceof Promise || a instanceof Promise ? Promise.all([i, a]).then(([t, n]) => en(e, t, n)) : en(e, i, a);
	};
});
function z(e, t) {
	if (e === t || e instanceof Date && t instanceof Date && +e == +t) return {
		valid: !0,
		data: e
	};
	if (x(e) && x(t)) {
		let n = Object.keys(t), r = Object.keys(e).filter((e) => n.indexOf(e) !== -1), i = {
			...e,
			...t
		};
		for (let n of r) {
			let r = z(e[n], t[n]);
			if (!r.valid) return {
				valid: !1,
				mergeErrorPath: [n, ...r.mergeErrorPath]
			};
			i[n] = r.data;
		}
		return {
			valid: !0,
			data: i
		};
	}
	if (Array.isArray(e) && Array.isArray(t)) {
		if (e.length !== t.length) return {
			valid: !1,
			mergeErrorPath: []
		};
		let n = [];
		for (let r = 0; r < e.length; r++) {
			let i = e[r], a = t[r], o = z(i, a);
			if (!o.valid) return {
				valid: !1,
				mergeErrorPath: [r, ...o.mergeErrorPath]
			};
			n.push(o.data);
		}
		return {
			valid: !0,
			data: n
		};
	}
	return {
		valid: !1,
		mergeErrorPath: []
	};
}
function en(e, t, n) {
	if (t.issues.length && e.issues.push(...t.issues), n.issues.length && e.issues.push(...n.issues), T(e)) return e;
	let r = z(t.value, n.value);
	if (!r.valid) throw Error(`Unmergable intersection. Error path: ${JSON.stringify(r.mergeErrorPath)}`);
	return e.value = r.data, e;
}
var tn = /*@__PURE__*/ s("$ZodEnum", (e, t) => {
	F.init(e, t);
	let n = f(t.entries), r = new Set(n);
	e._zod.values = r, e._zod.pattern = RegExp(`^(${n.filter((e) => ae.has(typeof e)).map((e) => typeof e == "string" ? S(e) : e.toString()).join("|")})$`), e._zod.parse = (t, i) => {
		let a = t.value;
		return r.has(a) || t.issues.push({
			code: "invalid_value",
			values: n,
			input: a,
			inst: e
		}), t;
	};
}), nn = /*@__PURE__*/ s("$ZodTransform", (e, t) => {
	F.init(e, t), e._zod.parse = (n, r) => {
		if (r.direction === "backward") throw new l(e.constructor.name);
		let i = t.transform(n.value, n);
		if (r.async) return (i instanceof Promise ? i : Promise.resolve(i)).then((e) => (n.value = e, n));
		if (i instanceof Promise) throw new c();
		return n.value = i, n;
	};
});
function rn(e, t) {
	return e.issues.length && t === void 0 ? {
		issues: [],
		value: void 0
	} : e;
}
var an = /*@__PURE__*/ s("$ZodOptional", (e, t) => {
	F.init(e, t), e._zod.optin = "optional", e._zod.optout = "optional", _(e._zod, "values", () => t.innerType._zod.values ? new Set([...t.innerType._zod.values, void 0]) : void 0), _(e._zod, "pattern", () => {
		let e = t.innerType._zod.pattern;
		return e ? RegExp(`^(${g(e.source)})?$`) : void 0;
	}), e._zod.parse = (e, n) => {
		if (t.innerType._zod.optin === "optional") {
			let r = t.innerType._zod.run(e, n);
			return r instanceof Promise ? r.then((t) => rn(t, e.value)) : rn(r, e.value);
		}
		return e.value === void 0 ? e : t.innerType._zod.run(e, n);
	};
}), on = /*@__PURE__*/ s("$ZodNullable", (e, t) => {
	F.init(e, t), _(e._zod, "optin", () => t.innerType._zod.optin), _(e._zod, "optout", () => t.innerType._zod.optout), _(e._zod, "pattern", () => {
		let e = t.innerType._zod.pattern;
		return e ? RegExp(`^(${g(e.source)}|null)$`) : void 0;
	}), _(e._zod, "values", () => t.innerType._zod.values ? new Set([...t.innerType._zod.values, null]) : void 0), e._zod.parse = (e, n) => e.value === null ? e : t.innerType._zod.run(e, n);
}), sn = /*@__PURE__*/ s("$ZodDefault", (e, t) => {
	F.init(e, t), e._zod.optin = "optional", _(e._zod, "values", () => t.innerType._zod.values), e._zod.parse = (e, n) => {
		if (n.direction === "backward") return t.innerType._zod.run(e, n);
		if (e.value === void 0) return e.value = t.defaultValue, e;
		let r = t.innerType._zod.run(e, n);
		return r instanceof Promise ? r.then((e) => cn(e, t)) : cn(r, t);
	};
});
function cn(e, t) {
	return e.value === void 0 && (e.value = t.defaultValue), e;
}
var ln = /*@__PURE__*/ s("$ZodPrefault", (e, t) => {
	F.init(e, t), e._zod.optin = "optional", _(e._zod, "values", () => t.innerType._zod.values), e._zod.parse = (e, n) => (n.direction === "backward" || e.value === void 0 && (e.value = t.defaultValue), t.innerType._zod.run(e, n));
}), un = /*@__PURE__*/ s("$ZodNonOptional", (e, t) => {
	F.init(e, t), _(e._zod, "values", () => {
		let e = t.innerType._zod.values;
		return e ? new Set([...e].filter((e) => e !== void 0)) : void 0;
	}), e._zod.parse = (n, r) => {
		let i = t.innerType._zod.run(n, r);
		return i instanceof Promise ? i.then((t) => dn(t, e)) : dn(i, e);
	};
});
function dn(e, t) {
	return !e.issues.length && e.value === void 0 && e.issues.push({
		code: "invalid_type",
		expected: "nonoptional",
		input: e.value,
		inst: t
	}), e;
}
var fn = /*@__PURE__*/ s("$ZodCatch", (e, t) => {
	F.init(e, t), _(e._zod, "optin", () => t.innerType._zod.optin), _(e._zod, "optout", () => t.innerType._zod.optout), _(e._zod, "values", () => t.innerType._zod.values), e._zod.parse = (e, n) => {
		if (n.direction === "backward") return t.innerType._zod.run(e, n);
		let r = t.innerType._zod.run(e, n);
		return r instanceof Promise ? r.then((r) => (e.value = r.value, r.issues.length && (e.value = t.catchValue({
			...e,
			error: { issues: r.issues.map((e) => D(e, n, d())) },
			input: e.value
		}), e.issues = []), e)) : (e.value = r.value, r.issues.length && (e.value = t.catchValue({
			...e,
			error: { issues: r.issues.map((e) => D(e, n, d())) },
			input: e.value
		}), e.issues = []), e);
	};
}), pn = /*@__PURE__*/ s("$ZodPipe", (e, t) => {
	F.init(e, t), _(e._zod, "values", () => t.in._zod.values), _(e._zod, "optin", () => t.in._zod.optin), _(e._zod, "optout", () => t.out._zod.optout), _(e._zod, "propValues", () => t.in._zod.propValues), e._zod.parse = (e, n) => {
		if (n.direction === "backward") {
			let r = t.out._zod.run(e, n);
			return r instanceof Promise ? r.then((e) => B(e, t.in, n)) : B(r, t.in, n);
		}
		let r = t.in._zod.run(e, n);
		return r instanceof Promise ? r.then((e) => B(e, t.out, n)) : B(r, t.out, n);
	};
});
function B(e, t, n) {
	return e.issues.length ? (e.aborted = !0, e) : t._zod.run({
		value: e.value,
		issues: e.issues
	}, n);
}
var mn = /*@__PURE__*/ s("$ZodReadonly", (e, t) => {
	F.init(e, t), _(e._zod, "propValues", () => t.innerType._zod.propValues), _(e._zod, "values", () => t.innerType._zod.values), _(e._zod, "optin", () => t.innerType._zod.optin), _(e._zod, "optout", () => t.innerType._zod.optout), e._zod.parse = (e, n) => {
		if (n.direction === "backward") return t.innerType._zod.run(e, n);
		let r = t.innerType._zod.run(e, n);
		return r instanceof Promise ? r.then(hn) : hn(r);
	};
});
function hn(e) {
	return e.value = Object.freeze(e.value), e;
}
var gn = /*@__PURE__*/ s("$ZodCustom", (e, t) => {
	N.init(e, t), F.init(e, t), e._zod.parse = (e, t) => e, e._zod.check = (n) => {
		let r = n.value, i = t.fn(r);
		if (i instanceof Promise) return i.then((t) => _n(t, n, r, e));
		_n(i, n, r, e);
	};
});
function _n(e, t, n, r) {
	if (!e) {
		let e = {
			code: "custom",
			input: n,
			inst: r,
			path: [...r._zod.def.path ?? []],
			continue: !r._zod.def.abort
		};
		r._zod.def.params && (e.params = r._zod.def.params), t.issues.push(O(e));
	}
}
//#endregion
//#region ../../node_modules/zod/v4/core/registries.js
var vn = class {
	constructor() {
		this._map = /* @__PURE__ */ new WeakMap(), this._idmap = /* @__PURE__ */ new Map();
	}
	add(e, ...t) {
		let n = t[0];
		if (this._map.set(e, n), n && typeof n == "object" && "id" in n) {
			if (this._idmap.has(n.id)) throw Error(`ID ${n.id} already exists in the registry`);
			this._idmap.set(n.id, e);
		}
		return this;
	}
	clear() {
		return this._map = /* @__PURE__ */ new WeakMap(), this._idmap = /* @__PURE__ */ new Map(), this;
	}
	remove(e) {
		let t = this._map.get(e);
		return t && typeof t == "object" && "id" in t && this._idmap.delete(t.id), this._map.delete(e), this;
	}
	get(e) {
		let t = e._zod.parent;
		if (t) {
			let n = { ...this.get(t) ?? {} };
			delete n.id;
			let r = {
				...n,
				...this._map.get(e)
			};
			return Object.keys(r).length ? r : void 0;
		}
		return this._map.get(e);
	}
	has(e) {
		return this._map.has(e);
	}
};
function yn() {
	return new vn();
}
var V = /*@__PURE__*/ yn();
//#endregion
//#region ../../node_modules/zod/v4/core/api.js
function bn(e, t) {
	return new e({
		type: "string",
		...w(t)
	});
}
function xn(e, t) {
	return new e({
		type: "string",
		format: "email",
		check: "string_format",
		abort: !1,
		...w(t)
	});
}
function Sn(e, t) {
	return new e({
		type: "string",
		format: "guid",
		check: "string_format",
		abort: !1,
		...w(t)
	});
}
function Cn(e, t) {
	return new e({
		type: "string",
		format: "uuid",
		check: "string_format",
		abort: !1,
		...w(t)
	});
}
function wn(e, t) {
	return new e({
		type: "string",
		format: "uuid",
		check: "string_format",
		abort: !1,
		version: "v4",
		...w(t)
	});
}
function Tn(e, t) {
	return new e({
		type: "string",
		format: "uuid",
		check: "string_format",
		abort: !1,
		version: "v6",
		...w(t)
	});
}
function En(e, t) {
	return new e({
		type: "string",
		format: "uuid",
		check: "string_format",
		abort: !1,
		version: "v7",
		...w(t)
	});
}
function Dn(e, t) {
	return new e({
		type: "string",
		format: "url",
		check: "string_format",
		abort: !1,
		...w(t)
	});
}
function On(e, t) {
	return new e({
		type: "string",
		format: "emoji",
		check: "string_format",
		abort: !1,
		...w(t)
	});
}
function kn(e, t) {
	return new e({
		type: "string",
		format: "nanoid",
		check: "string_format",
		abort: !1,
		...w(t)
	});
}
function An(e, t) {
	return new e({
		type: "string",
		format: "cuid",
		check: "string_format",
		abort: !1,
		...w(t)
	});
}
function jn(e, t) {
	return new e({
		type: "string",
		format: "cuid2",
		check: "string_format",
		abort: !1,
		...w(t)
	});
}
function Mn(e, t) {
	return new e({
		type: "string",
		format: "ulid",
		check: "string_format",
		abort: !1,
		...w(t)
	});
}
function Nn(e, t) {
	return new e({
		type: "string",
		format: "xid",
		check: "string_format",
		abort: !1,
		...w(t)
	});
}
function Pn(e, t) {
	return new e({
		type: "string",
		format: "ksuid",
		check: "string_format",
		abort: !1,
		...w(t)
	});
}
function Fn(e, t) {
	return new e({
		type: "string",
		format: "ipv4",
		check: "string_format",
		abort: !1,
		...w(t)
	});
}
function In(e, t) {
	return new e({
		type: "string",
		format: "ipv6",
		check: "string_format",
		abort: !1,
		...w(t)
	});
}
function Ln(e, t) {
	return new e({
		type: "string",
		format: "cidrv4",
		check: "string_format",
		abort: !1,
		...w(t)
	});
}
function Rn(e, t) {
	return new e({
		type: "string",
		format: "cidrv6",
		check: "string_format",
		abort: !1,
		...w(t)
	});
}
function zn(e, t) {
	return new e({
		type: "string",
		format: "base64",
		check: "string_format",
		abort: !1,
		...w(t)
	});
}
function Bn(e, t) {
	return new e({
		type: "string",
		format: "base64url",
		check: "string_format",
		abort: !1,
		...w(t)
	});
}
function Vn(e, t) {
	return new e({
		type: "string",
		format: "e164",
		check: "string_format",
		abort: !1,
		...w(t)
	});
}
function Hn(e, t) {
	return new e({
		type: "string",
		format: "jwt",
		check: "string_format",
		abort: !1,
		...w(t)
	});
}
function Un(e, t) {
	return new e({
		type: "string",
		format: "datetime",
		check: "string_format",
		offset: !1,
		local: !1,
		precision: null,
		...w(t)
	});
}
function Wn(e, t) {
	return new e({
		type: "string",
		format: "date",
		check: "string_format",
		...w(t)
	});
}
function Gn(e, t) {
	return new e({
		type: "string",
		format: "time",
		check: "string_format",
		precision: null,
		...w(t)
	});
}
function Kn(e, t) {
	return new e({
		type: "string",
		format: "duration",
		check: "string_format",
		...w(t)
	});
}
function qn(e) {
	return new e({ type: "unknown" });
}
function Jn(e, t) {
	return new e({
		type: "never",
		...w(t)
	});
}
function Yn(e, t) {
	return new at({
		check: "max_length",
		...w(t),
		maximum: e
	});
}
function H(e, t) {
	return new ot({
		check: "min_length",
		...w(t),
		minimum: e
	});
}
function Xn(e, t) {
	return new st({
		check: "length_equals",
		...w(t),
		length: e
	});
}
function Zn(e, t) {
	return new ct({
		check: "string_format",
		format: "regex",
		...w(t),
		pattern: e
	});
}
function Qn(e) {
	return new lt({
		check: "string_format",
		format: "lowercase",
		...w(e)
	});
}
function $n(e) {
	return new ut({
		check: "string_format",
		format: "uppercase",
		...w(e)
	});
}
function er(e, t) {
	return new dt({
		check: "string_format",
		format: "includes",
		...w(t),
		includes: e
	});
}
function tr(e, t) {
	return new ft({
		check: "string_format",
		format: "starts_with",
		...w(t),
		prefix: e
	});
}
function nr(e, t) {
	return new pt({
		check: "string_format",
		format: "ends_with",
		...w(t),
		suffix: e
	});
}
function U(e) {
	return new mt({
		check: "overwrite",
		tx: e
	});
}
function rr(e) {
	return U((t) => t.normalize(e));
}
function ir() {
	return U((e) => e.trim());
}
function ar() {
	return U((e) => e.toLowerCase());
}
function or() {
	return U((e) => e.toUpperCase());
}
function sr(e, t, n) {
	return new e({
		type: "array",
		element: t,
		...w(n)
	});
}
function cr(e, t, n) {
	return new e({
		type: "custom",
		check: "custom",
		fn: t,
		...w(n)
	});
}
function lr(e) {
	let t = ur((n) => (n.addIssue = (e) => {
		if (typeof e == "string") n.issues.push(O(e, n.value, t._zod.def));
		else {
			let r = e;
			r.fatal && (r.continue = !1), r.code ??= "custom", r.input ??= n.value, r.inst ??= t, r.continue ??= !t._zod.def.abort, n.issues.push(O(r));
		}
	}, e(n.value, n)));
	return t;
}
function ur(e, t) {
	let n = new N({
		check: "custom",
		...w(t)
	});
	return n._zod.check = e, n;
}
//#endregion
//#region ../../node_modules/zod/v4/classic/iso.js
var dr = /*@__PURE__*/ s("ZodISODateTime", (e, t) => {
	Ot.init(e, t), K.init(e, t);
});
function fr(e) {
	return Un(dr, e);
}
var pr = /*@__PURE__*/ s("ZodISODate", (e, t) => {
	kt.init(e, t), K.init(e, t);
});
function mr(e) {
	return Wn(pr, e);
}
var hr = /*@__PURE__*/ s("ZodISOTime", (e, t) => {
	At.init(e, t), K.init(e, t);
});
function gr(e) {
	return Gn(hr, e);
}
var _r = /*@__PURE__*/ s("ZodISODuration", (e, t) => {
	jt.init(e, t), K.init(e, t);
});
function vr(e) {
	return Kn(_r, e);
}
//#endregion
//#region ../../node_modules/zod/v4/classic/errors.js
var yr = (e, t) => {
	_e.init(e, t), e.name = "ZodError", Object.defineProperties(e, {
		format: { value: (t) => be(e, t) },
		flatten: { value: (t) => ye(e, t) },
		addIssue: { value: (t) => {
			e.issues.push(t), e.message = JSON.stringify(e.issues, p, 2);
		} },
		addIssues: { value: (t) => {
			e.issues.push(...t), e.message = JSON.stringify(e.issues, p, 2);
		} },
		isEmpty: { get() {
			return e.issues.length === 0;
		} }
	});
};
s("ZodError", yr);
var W = s("ZodError", yr, { Parent: Error }), br = /* @__PURE__ */ k(W), xr = /* @__PURE__ */ A(W), Sr = /* @__PURE__ */ j(W), Cr = /* @__PURE__ */ M(W), wr = /* @__PURE__ */ Ce(W), Tr = /* @__PURE__ */ we(W), Er = /* @__PURE__ */ Te(W), Dr = /* @__PURE__ */ Ee(W), Or = /* @__PURE__ */ De(W), kr = /* @__PURE__ */ Oe(W), Ar = /* @__PURE__ */ ke(W), jr = /* @__PURE__ */ Ae(W), G = /*@__PURE__*/ s("ZodType", (e, t) => (F.init(e, t), e.def = t, e.type = t.type, Object.defineProperty(e, "_def", { value: t }), e.check = (...n) => e.clone(y(t, { checks: [...t.checks ?? [], ...n.map((e) => typeof e == "function" ? { _zod: {
	check: e,
	def: { check: "custom" },
	onattach: []
} } : e)] })), e.clone = (t, n) => C(e, t, n), e.brand = () => e, e.register = ((t, n) => (t.add(e, n), e)), e.parse = (t, n) => br(e, t, n, { callee: e.parse }), e.safeParse = (t, n) => Sr(e, t, n), e.parseAsync = async (t, n) => xr(e, t, n, { callee: e.parseAsync }), e.safeParseAsync = async (t, n) => Cr(e, t, n), e.spa = e.safeParseAsync, e.encode = (t, n) => wr(e, t, n), e.decode = (t, n) => Tr(e, t, n), e.encodeAsync = async (t, n) => Er(e, t, n), e.decodeAsync = async (t, n) => Dr(e, t, n), e.safeEncode = (t, n) => Or(e, t, n), e.safeDecode = (t, n) => kr(e, t, n), e.safeEncodeAsync = async (t, n) => Ar(e, t, n), e.safeDecodeAsync = async (t, n) => jr(e, t, n), e.refine = (t, n) => e.check(Mi(t, n)), e.superRefine = (t) => e.check(Ni(t)), e.overwrite = (t) => e.check(U(t)), e.optional = () => gi(e), e.nullable = () => vi(e), e.nullish = () => gi(vi(e)), e.nonoptional = (t) => wi(e, t), e.array = () => ii(e), e.or = (t) => ci([e, t]), e.and = (t) => ui(e, t), e.transform = (t) => Oi(e, mi(t)), e.default = (t) => bi(e, t), e.prefault = (t) => Si(e, t), e.catch = (t) => Ei(e, t), e.pipe = (t) => Oi(e, t), e.readonly = () => Ai(e), e.describe = (t) => {
	let n = e.clone();
	return V.add(n, { description: t }), n;
}, Object.defineProperty(e, "description", {
	get() {
		return V.get(e)?.description;
	},
	configurable: !0
}), e.meta = (...t) => {
	if (t.length === 0) return V.get(e);
	let n = e.clone();
	return V.add(n, t[0]), n;
}, e.isOptional = () => e.safeParse(void 0).success, e.isNullable = () => e.safeParse(null).success, e)), Mr = /*@__PURE__*/ s("_ZodString", (e, t) => {
	I.init(e, t), G.init(e, t);
	let n = e._zod.bag;
	e.format = n.format ?? null, e.minLength = n.minimum ?? null, e.maxLength = n.maximum ?? null, e.regex = (...t) => e.check(Zn(...t)), e.includes = (...t) => e.check(er(...t)), e.startsWith = (...t) => e.check(tr(...t)), e.endsWith = (...t) => e.check(nr(...t)), e.min = (...t) => e.check(H(...t)), e.max = (...t) => e.check(Yn(...t)), e.length = (...t) => e.check(Xn(...t)), e.nonempty = (...t) => e.check(H(1, ...t)), e.lowercase = (t) => e.check(Qn(t)), e.uppercase = (t) => e.check($n(t)), e.trim = () => e.check(ir()), e.normalize = (...t) => e.check(rr(...t)), e.toLowerCase = () => e.check(ar()), e.toUpperCase = () => e.check(or());
}), Nr = /*@__PURE__*/ s("ZodString", (e, t) => {
	I.init(e, t), Mr.init(e, t), e.email = (t) => e.check(xn(Fr, t)), e.url = (t) => e.check(Dn(Lr, t)), e.jwt = (t) => e.check(Hn(Qr, t)), e.emoji = (t) => e.check(On(Rr, t)), e.guid = (t) => e.check(Sn(Ir, t)), e.uuid = (t) => e.check(Cn(q, t)), e.uuidv4 = (t) => e.check(wn(q, t)), e.uuidv6 = (t) => e.check(Tn(q, t)), e.uuidv7 = (t) => e.check(En(q, t)), e.nanoid = (t) => e.check(kn(zr, t)), e.guid = (t) => e.check(Sn(Ir, t)), e.cuid = (t) => e.check(An(Br, t)), e.cuid2 = (t) => e.check(jn(Vr, t)), e.ulid = (t) => e.check(Mn(Hr, t)), e.base64 = (t) => e.check(zn(Yr, t)), e.base64url = (t) => e.check(Bn(Xr, t)), e.xid = (t) => e.check(Nn(Ur, t)), e.ksuid = (t) => e.check(Pn(Wr, t)), e.ipv4 = (t) => e.check(Fn(Gr, t)), e.ipv6 = (t) => e.check(In(Kr, t)), e.cidrv4 = (t) => e.check(Ln(qr, t)), e.cidrv6 = (t) => e.check(Rn(Jr, t)), e.e164 = (t) => e.check(Vn(Zr, t)), e.datetime = (t) => e.check(fr(t)), e.date = (t) => e.check(mr(t)), e.time = (t) => e.check(gr(t)), e.duration = (t) => e.check(vr(t));
});
function Pr(e) {
	return bn(Nr, e);
}
var K = /*@__PURE__*/ s("ZodStringFormat", (e, t) => {
	L.init(e, t), Mr.init(e, t);
}), Fr = /*@__PURE__*/ s("ZodEmail", (e, t) => {
	yt.init(e, t), K.init(e, t);
}), Ir = /*@__PURE__*/ s("ZodGUID", (e, t) => {
	_t.init(e, t), K.init(e, t);
}), q = /*@__PURE__*/ s("ZodUUID", (e, t) => {
	vt.init(e, t), K.init(e, t);
}), Lr = /*@__PURE__*/ s("ZodURL", (e, t) => {
	bt.init(e, t), K.init(e, t);
}), Rr = /*@__PURE__*/ s("ZodEmoji", (e, t) => {
	xt.init(e, t), K.init(e, t);
}), zr = /*@__PURE__*/ s("ZodNanoID", (e, t) => {
	St.init(e, t), K.init(e, t);
}), Br = /*@__PURE__*/ s("ZodCUID", (e, t) => {
	Ct.init(e, t), K.init(e, t);
}), Vr = /*@__PURE__*/ s("ZodCUID2", (e, t) => {
	wt.init(e, t), K.init(e, t);
}), Hr = /*@__PURE__*/ s("ZodULID", (e, t) => {
	Tt.init(e, t), K.init(e, t);
}), Ur = /*@__PURE__*/ s("ZodXID", (e, t) => {
	Et.init(e, t), K.init(e, t);
}), Wr = /*@__PURE__*/ s("ZodKSUID", (e, t) => {
	Dt.init(e, t), K.init(e, t);
}), Gr = /*@__PURE__*/ s("ZodIPv4", (e, t) => {
	Mt.init(e, t), K.init(e, t);
}), Kr = /*@__PURE__*/ s("ZodIPv6", (e, t) => {
	Nt.init(e, t), K.init(e, t);
}), qr = /*@__PURE__*/ s("ZodCIDRv4", (e, t) => {
	Pt.init(e, t), K.init(e, t);
}), Jr = /*@__PURE__*/ s("ZodCIDRv6", (e, t) => {
	Ft.init(e, t), K.init(e, t);
}), Yr = /*@__PURE__*/ s("ZodBase64", (e, t) => {
	Lt.init(e, t), K.init(e, t);
}), Xr = /*@__PURE__*/ s("ZodBase64URL", (e, t) => {
	zt.init(e, t), K.init(e, t);
}), Zr = /*@__PURE__*/ s("ZodE164", (e, t) => {
	Bt.init(e, t), K.init(e, t);
}), Qr = /*@__PURE__*/ s("ZodJWT", (e, t) => {
	Ht.init(e, t), K.init(e, t);
}), $r = /*@__PURE__*/ s("ZodUnknown", (e, t) => {
	Ut.init(e, t), G.init(e, t);
});
function ei() {
	return qn($r);
}
var ti = /*@__PURE__*/ s("ZodNever", (e, t) => {
	Wt.init(e, t), G.init(e, t);
});
function ni(e) {
	return Jn(ti, e);
}
var ri = /*@__PURE__*/ s("ZodArray", (e, t) => {
	Kt.init(e, t), G.init(e, t), e.element = t.element, e.min = (t, n) => e.check(H(t, n)), e.nonempty = (t) => e.check(H(1, t)), e.max = (t, n) => e.check(Yn(t, n)), e.length = (t, n) => e.check(Xn(t, n)), e.unwrap = () => e.element;
});
function ii(e, t) {
	return sr(ri, e, t);
}
var ai = /*@__PURE__*/ s("ZodObject", (e, t) => {
	Xt.init(e, t), G.init(e, t), _(e, "shape", () => t.shape), e.keyof = () => fi(Object.keys(e._zod.def.shape)), e.catchall = (t) => e.clone({
		...e._zod.def,
		catchall: t
	}), e.passthrough = () => e.clone({
		...e._zod.def,
		catchall: ei()
	}), e.loose = () => e.clone({
		...e._zod.def,
		catchall: ei()
	}), e.strict = () => e.clone({
		...e._zod.def,
		catchall: ni()
	}), e.strip = () => e.clone({
		...e._zod.def,
		catchall: void 0
	}), e.extend = (t) => le(e, t), e.safeExtend = (t) => ue(e, t), e.merge = (t) => de(e, t), e.pick = (t) => se(e, t), e.omit = (t) => ce(e, t), e.partial = (...t) => fe(hi, e, t[0]), e.required = (...t) => pe(Ci, e, t[0]);
});
function oi(e, t) {
	return new ai({
		type: "object",
		shape: e ?? {},
		...w(t)
	});
}
var si = /*@__PURE__*/ s("ZodUnion", (e, t) => {
	Qt.init(e, t), G.init(e, t), e.options = t.options;
});
function ci(e, t) {
	return new si({
		type: "union",
		options: e,
		...w(t)
	});
}
var li = /*@__PURE__*/ s("ZodIntersection", (e, t) => {
	$t.init(e, t), G.init(e, t);
});
function ui(e, t) {
	return new li({
		type: "intersection",
		left: e,
		right: t
	});
}
var di = /*@__PURE__*/ s("ZodEnum", (e, t) => {
	tn.init(e, t), G.init(e, t), e.enum = t.entries, e.options = Object.values(t.entries);
	let n = new Set(Object.keys(t.entries));
	e.extract = (e, r) => {
		let i = {};
		for (let r of e) if (n.has(r)) i[r] = t.entries[r];
		else throw Error(`Key ${r} not found in enum`);
		return new di({
			...t,
			checks: [],
			...w(r),
			entries: i
		});
	}, e.exclude = (e, r) => {
		let i = { ...t.entries };
		for (let t of e) if (n.has(t)) delete i[t];
		else throw Error(`Key ${t} not found in enum`);
		return new di({
			...t,
			checks: [],
			...w(r),
			entries: i
		});
	};
});
function fi(e, t) {
	return new di({
		type: "enum",
		entries: Array.isArray(e) ? Object.fromEntries(e.map((e) => [e, e])) : e,
		...w(t)
	});
}
var pi = /*@__PURE__*/ s("ZodTransform", (e, t) => {
	nn.init(e, t), G.init(e, t), e._zod.parse = (n, r) => {
		if (r.direction === "backward") throw new l(e.constructor.name);
		n.addIssue = (r) => {
			if (typeof r == "string") n.issues.push(O(r, n.value, t));
			else {
				let t = r;
				t.fatal && (t.continue = !1), t.code ??= "custom", t.input ??= n.value, t.inst ??= e, n.issues.push(O(t));
			}
		};
		let i = t.transform(n.value, n);
		return i instanceof Promise ? i.then((e) => (n.value = e, n)) : (n.value = i, n);
	};
});
function mi(e) {
	return new pi({
		type: "transform",
		transform: e
	});
}
var hi = /*@__PURE__*/ s("ZodOptional", (e, t) => {
	an.init(e, t), G.init(e, t), e.unwrap = () => e._zod.def.innerType;
});
function gi(e) {
	return new hi({
		type: "optional",
		innerType: e
	});
}
var _i = /*@__PURE__*/ s("ZodNullable", (e, t) => {
	on.init(e, t), G.init(e, t), e.unwrap = () => e._zod.def.innerType;
});
function vi(e) {
	return new _i({
		type: "nullable",
		innerType: e
	});
}
var yi = /*@__PURE__*/ s("ZodDefault", (e, t) => {
	sn.init(e, t), G.init(e, t), e.unwrap = () => e._zod.def.innerType, e.removeDefault = e.unwrap;
});
function bi(e, t) {
	return new yi({
		type: "default",
		innerType: e,
		get defaultValue() {
			return typeof t == "function" ? t() : ie(t);
		}
	});
}
var xi = /*@__PURE__*/ s("ZodPrefault", (e, t) => {
	ln.init(e, t), G.init(e, t), e.unwrap = () => e._zod.def.innerType;
});
function Si(e, t) {
	return new xi({
		type: "prefault",
		innerType: e,
		get defaultValue() {
			return typeof t == "function" ? t() : ie(t);
		}
	});
}
var Ci = /*@__PURE__*/ s("ZodNonOptional", (e, t) => {
	un.init(e, t), G.init(e, t), e.unwrap = () => e._zod.def.innerType;
});
function wi(e, t) {
	return new Ci({
		type: "nonoptional",
		innerType: e,
		...w(t)
	});
}
var Ti = /*@__PURE__*/ s("ZodCatch", (e, t) => {
	fn.init(e, t), G.init(e, t), e.unwrap = () => e._zod.def.innerType, e.removeCatch = e.unwrap;
});
function Ei(e, t) {
	return new Ti({
		type: "catch",
		innerType: e,
		catchValue: typeof t == "function" ? t : () => t
	});
}
var Di = /*@__PURE__*/ s("ZodPipe", (e, t) => {
	pn.init(e, t), G.init(e, t), e.in = t.in, e.out = t.out;
});
function Oi(e, t) {
	return new Di({
		type: "pipe",
		in: e,
		out: t
	});
}
var ki = /*@__PURE__*/ s("ZodReadonly", (e, t) => {
	mn.init(e, t), G.init(e, t), e.unwrap = () => e._zod.def.innerType;
});
function Ai(e) {
	return new ki({
		type: "readonly",
		innerType: e
	});
}
var ji = /*@__PURE__*/ s("ZodCustom", (e, t) => {
	gn.init(e, t), G.init(e, t);
});
function Mi(e, t = {}) {
	return cr(ji, e, t);
}
function Ni(e) {
	return lr(e);
}
//#endregion
//#region ../twenty-shared/node_modules/class-validator/esm5/metadata/ValidationMetadata.js
var Pi = function() {
	function e(e) {
		this.groups = [], this.each = !1, this.context = void 0, this.type = e.type, this.name = e.name, this.target = e.target, this.propertyName = e.propertyName, this.constraints = e?.constraints, this.constraintCls = e.constraintCls, this.validationTypeOptions = e.validationTypeOptions, e.validationOptions && (this.message = e.validationOptions.message, this.groups = e.validationOptions.groups, this.always = e.validationOptions.always, this.each = e.validationOptions.each, this.context = e.validationOptions.context);
	}
	return e;
}(), Fi = function() {
	function e() {}
	return e.prototype.transform = function(e) {
		var t = [];
		return Object.keys(e.properties).forEach(function(n) {
			e.properties[n].forEach(function(r) {
				var i = {
					message: r.message,
					groups: r.groups,
					always: r.always,
					each: r.each
				}, a = {
					type: r.type,
					name: r.name,
					target: e.name,
					propertyName: n,
					constraints: r.constraints,
					validationTypeOptions: r.options,
					validationOptions: i
				};
				t.push(new Pi(a));
			});
		}), t;
	}, e;
}(), Ii = function(e) {
	var t = typeof Symbol == "function" && Symbol.iterator, n = t && e[t], r = 0;
	if (n) return n.call(e);
	if (e && typeof e.length == "number") return { next: function() {
		return e && r >= e.length && (e = void 0), {
			value: e && e[r++],
			done: !e
		};
	} };
	throw TypeError(t ? "Object is not iterable." : "Symbol.iterator is not defined.");
}, Li = function(e, t) {
	var n = typeof Symbol == "function" && e[Symbol.iterator];
	if (!n) return e;
	var r = n.call(e), i, a = [], o;
	try {
		for (; (t === void 0 || t-- > 0) && !(i = r.next()).done;) a.push(i.value);
	} catch (e) {
		o = { error: e };
	} finally {
		try {
			i && !i.done && (n = r.return) && n.call(r);
		} finally {
			if (o) throw o.error;
		}
	}
	return a;
}, Ri = function(e, t, n) {
	if (n || arguments.length === 2) for (var r = 0, i = t.length, a; r < i; r++) (a || !(r in t)) && (a ||= Array.prototype.slice.call(t, 0, r), a[r] = t[r]);
	return e.concat(a || Array.prototype.slice.call(t));
};
(function() {
	function e() {
		this.validationMetadatas = /* @__PURE__ */ new Map(), this.constraintMetadatas = /* @__PURE__ */ new Map();
	}
	return Object.defineProperty(e.prototype, "hasValidationMetaData", {
		get: function() {
			return !!this.validationMetadatas.size;
		},
		enumerable: !1,
		configurable: !0
	}), e.prototype.addValidationSchema = function(e) {
		var t = this;
		new Fi().transform(e).forEach(function(e) {
			return t.addValidationMetadata(e);
		});
	}, e.prototype.addValidationMetadata = function(e) {
		var t = this.validationMetadatas.get(e.target);
		t ? t.push(e) : this.validationMetadatas.set(e.target, [e]);
	}, e.prototype.addConstraintMetadata = function(e) {
		var t = this.constraintMetadatas.get(e.target);
		t ? t.push(e) : this.constraintMetadatas.set(e.target, [e]);
	}, e.prototype.groupByPropertyName = function(e) {
		var t = {};
		return e.forEach(function(e) {
			t[e.propertyName] || (t[e.propertyName] = []), t[e.propertyName].push(e);
		}), t;
	}, e.prototype.getTargetValidationMetadatas = function(e, t, n, r, i) {
		var a, o, s = function(e) {
			return e.always === void 0 ? e.groups && e.groups.length ? !1 : n : e.always;
		}, c = function(e) {
			return !!(r && (!i || !i.length) && e.groups && e.groups.length);
		}, l = (this.validationMetadatas.get(e) || []).filter(function(n) {
			return n.target !== e && n.target !== t ? !1 : s(n) ? !0 : c(n) ? !1 : i && i.length > 0 ? n.groups && !!n.groups.find(function(e) {
				return i.indexOf(e) !== -1;
			}) : !0;
		}), u = [];
		try {
			for (var d = Ii(this.validationMetadatas.entries()), f = d.next(); !f.done; f = d.next()) {
				var p = Li(f.value, 2), m = p[0], h = p[1];
				e.prototype instanceof m && u.push.apply(u, Ri([], Li(h), !1));
			}
		} catch (e) {
			a = { error: e };
		} finally {
			try {
				f && !f.done && (o = d.return) && o.call(d);
			} finally {
				if (a) throw a.error;
			}
		}
		var g = u.filter(function(t) {
			return typeof t.target == "string" || t.target === e || t.target instanceof Function && !(e.prototype instanceof t.target) ? !1 : s(t) ? !0 : c(t) ? !1 : i && i.length > 0 ? t.groups && !!t.groups.find(function(e) {
				return i.indexOf(e) !== -1;
			}) : !0;
		}).filter(function(e) {
			return !l.find(function(t) {
				return t.propertyName === e.propertyName && t.type === e.type;
			});
		});
		return l.concat(g);
	}, e.prototype.getTargetValidatorConstraints = function(e) {
		return this.constraintMetadatas.get(e) || [];
	}, e;
})(), function() {
	function e() {}
	return e.isValid = function(e) {
		var t = this;
		return e !== "isValid" && e !== "getMessage" && Object.keys(this).map(function(e) {
			return t[e];
		}).indexOf(e) !== -1;
	}, e.CUSTOM_VALIDATION = "customValidation", e.NESTED_VALIDATION = "nestedValidation", e.PROMISE_VALIDATION = "promiseValidation", e.CONDITIONAL_VALIDATION = "conditionalValidation", e.WHITELIST = "whitelistValidation", e.IS_DEFINED = "isDefined", e;
}();
//#endregion
//#region ../twenty-shared/node_modules/class-validator/esm5/validation/ValidationTypes.js
var zi = new (function() {
	function e() {
		this.instances = [];
	}
	return e.prototype.get = function(e) {
		var t = this.instances.find(function(t) {
			return t.type === e;
		});
		return t || (t = {
			type: e,
			object: new e()
		}, this.instances.push(t)), t.object;
	}, e;
}())(), Bi, J;
function Vi(e) {
	if (Bi) try {
		var t = Bi.get(e);
		if (t || !J || !J.fallback) return t;
	} catch (e) {
		if (!J || !J.fallbackOnErrors) throw e;
	}
	return zi.get(e);
}
(function() {
	function e(e, t, n) {
		n === void 0 && (n = !1), this.target = e, this.name = t, this.async = n;
	}
	return Object.defineProperty(e.prototype, "instance", {
		get: function() {
			return Vi(this.target);
		},
		enumerable: !1,
		configurable: !0
	}), e;
})();
//#endregion
//#region ../twenty-shared/node_modules/class-validator/esm5/metadata/ConstraintMetadata.js
var Hi = /* @__PURE__ */ function(e) {
	return e.Verify = "/verify", e.VerifyEmail = "/verify-email", e.SignInUp = "/welcome", e.SignInUpV2 = "/welcome-v2", e.Invite = "/invite/:workspaceInviteHash", e.ResetPassword = "/reset-password/:passwordResetToken", e.WorkspaceActivation = "/workspace-activation", e.CreateProfile = "/create/profile", e.SyncEmails = "/sync/emails", e.InviteTeam = "/invite-team", e.PlanRequired = "/plan-required", e.PlanRequiredSuccess = "/plan-required/payment-success", e.BookCallDecision = "/book-call-decision", e.BookCall = "/book-call", e.Index = "/", e.TasksPage = "/objects/tasks", e.OpportunitiesPage = "/objects/opportunities", e.RecordIndexPage = "/objects/:objectNamePlural", e.RecordShowPage = "/object/:objectNameSingular/:objectRecordId", e.PageLayoutPage = "/page/:pageLayoutId", e.Settings = "settings", e.SettingsCatchAll = "/settings/*", e.Developers = "developers", e.DevelopersCatchAll = "/developers/*", e.Authorize = "/authorize", e.NotFoundWildcard = "*", e.NotFound = "/not-found", e;
}({}), Y = /* @__PURE__ */ function(e) {
	return e.ACTOR = "ACTOR", e.ADDRESS = "ADDRESS", e.ARRAY = "ARRAY", e.BOOLEAN = "BOOLEAN", e.CURRENCY = "CURRENCY", e.DATE = "DATE", e.DATE_TIME = "DATE_TIME", e.EMAILS = "EMAILS", e.FILES = "FILES", e.FULL_NAME = "FULL_NAME", e.LINKS = "LINKS", e.MORPH_RELATION = "MORPH_RELATION", e.MULTI_SELECT = "MULTI_SELECT", e.NUMBER = "NUMBER", e.NUMERIC = "NUMERIC", e.PHONES = "PHONES", e.POSITION = "POSITION", e.RATING = "RATING", e.RAW_JSON = "RAW_JSON", e.RELATION = "RELATION", e.RICH_TEXT = "RICH_TEXT", e.SELECT = "SELECT", e.TEXT = "TEXT", e.TS_VECTOR = "TS_VECTOR", e.UUID = "UUID", e;
}({}), Ui = /* @__PURE__ */ function(e) {
	return e.EMAIL = "EMAIL", e.CALENDAR = "CALENDAR", e.WORKFLOW = "WORKFLOW", e.AGENT = "AGENT", e.API = "API", e.IMPORT = "IMPORT", e.MANUAL = "MANUAL", e.SYSTEM = "SYSTEM", e.WEBHOOK = "WEBHOOK", e.APPLICATION = "APPLICATION", e;
}({}), Wi = {
	type: Y.ACTOR,
	properties: [
		{
			name: "source",
			type: Y.SELECT,
			hidden: !1,
			isRequired: !0,
			options: Object.keys(Ui).map((e, t) => ({
				id: n(),
				label: `${Ui[e].toLowerCase()}`,
				value: e,
				position: t
			}))
		},
		{
			name: "workspaceMemberId",
			type: Y.UUID,
			hidden: "input",
			isRequired: !1
		},
		{
			name: "name",
			type: Y.TEXT,
			hidden: "input",
			isRequired: !0
		},
		{
			name: "context",
			type: Y.RAW_JSON,
			hidden: !1,
			isRequired: !1
		}
	]
}, Gi = {
	type: Y.ADDRESS,
	properties: [
		{
			name: "addressStreet1",
			type: Y.TEXT,
			hidden: !1,
			isRequired: !1
		},
		{
			name: "addressStreet2",
			type: Y.TEXT,
			hidden: !1,
			isRequired: !1
		},
		{
			name: "addressCity",
			type: Y.TEXT,
			hidden: !1,
			isRequired: !1
		},
		{
			name: "addressPostcode",
			type: Y.TEXT,
			hidden: !1,
			isRequired: !1
		},
		{
			name: "addressState",
			type: Y.TEXT,
			hidden: !1,
			isRequired: !1
		},
		{
			name: "addressCountry",
			type: Y.TEXT,
			hidden: !1,
			isRequired: !1
		},
		{
			name: "addressLat",
			type: Y.NUMERIC,
			hidden: !1,
			isRequired: !1
		},
		{
			name: "addressLng",
			type: Y.NUMERIC,
			hidden: !1,
			isRequired: !1
		}
	]
}, Ki = {
	type: Y.CURRENCY,
	properties: [{
		name: "amountMicros",
		type: Y.NUMERIC,
		hidden: !1,
		isRequired: !1
	}, {
		name: "currencyCode",
		type: Y.TEXT,
		hidden: !1,
		isRequired: !1
	}]
}, qi = {
	type: Y.EMAILS,
	properties: [{
		name: "primaryEmail",
		type: Y.TEXT,
		hidden: !1,
		isRequired: !1,
		isIncludedInUniqueConstraint: !0
	}, {
		name: "additionalEmails",
		type: Y.RAW_JSON,
		hidden: !1,
		isRequired: !1
	}]
}, Ji = {
	type: Y.FULL_NAME,
	properties: [{
		name: "firstName",
		type: Y.TEXT,
		hidden: !1,
		isRequired: !1,
		isIncludedInUniqueConstraint: !1
	}, {
		name: "lastName",
		type: Y.TEXT,
		hidden: !1,
		isRequired: !1,
		isIncludedInUniqueConstraint: !1
	}]
}, Yi = {
	type: Y.LINKS,
	properties: [
		{
			name: "primaryLinkLabel",
			type: Y.TEXT,
			hidden: !1,
			isRequired: !1
		},
		{
			name: "primaryLinkUrl",
			type: Y.TEXT,
			hidden: !1,
			isRequired: !1,
			isIncludedInUniqueConstraint: !0
		},
		{
			name: "secondaryLinks",
			type: Y.RAW_JSON,
			hidden: !1,
			isRequired: !1
		}
	]
}, Xi = {
	type: Y.PHONES,
	properties: [
		{
			name: "primaryPhoneNumber",
			type: Y.TEXT,
			hidden: !1,
			isRequired: !1,
			isIncludedInUniqueConstraint: !0
		},
		{
			name: "primaryPhoneCountryCode",
			type: Y.TEXT,
			hidden: !1,
			isRequired: !1,
			isIncludedInUniqueConstraint: !0
		},
		{
			name: "primaryPhoneCallingCode",
			type: Y.TEXT,
			hidden: !1,
			isRequired: !1,
			isIncludedInUniqueConstraint: !0
		},
		{
			name: "additionalPhones",
			type: Y.RAW_JSON,
			hidden: !1,
			isRequired: !1
		}
	]
}, Zi = {
	type: Y.RICH_TEXT,
	properties: [{
		name: "blocknote",
		type: Y.TEXT,
		hidden: !1,
		isRequired: !1
	}, {
		name: "markdown",
		type: Y.TEXT,
		hidden: !1,
		isRequired: !1
	}]
};
oi({
	blocknote: Pr().nullable().optional(),
	markdown: Pr().nullable()
}), new Map([
	[Y.LINKS, Yi],
	[Y.CURRENCY, Ki],
	[Y.FULL_NAME, Ji],
	[Y.ADDRESS, Gi],
	[Y.ACTOR, Wi],
	[Y.EMAILS, qi],
	[Y.PHONES, Xi],
	[Y.RICH_TEXT, Zi]
]);
//#endregion
//#region ../twenty-shared/dist/types-CbtF-GT7.mjs
var Qi = /* @__PURE__ */ function(e) {
	return e.CommandMenuDisplay = "command-menu-display", e.ViewRecord = "view-record", e.MergeRecords = "merge-records", e.UpdateRecords = "update-records", e.EditRichText = "edit-rich-text", e.Copilot = "copilot", e.WorkflowTriggerSelectType = "workflow-trigger-select-type", e.WorkflowStepCreate = "workflow-step-create", e.WorkflowStepEditType = "workflow-step-edit-type", e.WorkflowStepView = "workflow-step-view", e.WorkflowStepEdit = "workflow-step-edit", e.WorkflowRunStepView = "workflow-run-step-view", e.SearchRecords = "search-records", e.AskAI = "ask-ai", e.ViewPreviousAiChats = "view-previous-ai-chats", e.PageLayoutDashboardWidgetTypeSelect = "page-layout-dashboard-widget-type-select", e.PageLayoutTabSettings = "page-layout-tab-settings", e.DashboardChartSettings = "dashboard-chart-settings", e.DashboardIframeSettings = "dashboard-iframe-settings", e.DashboardRecordTableSettings = "dashboard-record-table-settings", e.RecordPageFieldsSettings = "record-page-fields-settings", e.RecordPageFieldSettings = "record-page-field-settings", e.ViewFrontComponent = "view-front-component", e.NavigationMenuItemEdit = "navigation-menu-item-edit", e.NavigationMenuAddItem = "navigation-menu-add-item", e.CommandMenuEdit = "command-menu-edit", e.PageLayoutRecordPageWidgetTypeSelect = "page-layout-record-page-widget-type-select", e.ComposeEmail = "compose-email", e.ComposeCampaign = "compose-campaign", e;
}({}), $i = "frontComponentHostCommunicationApi";
//#endregion
//#region src/sdk/front-component/globals/frontComponentHostCommunicationApi.ts
globalThis[$i] ??= {};
var X = globalThis.frontComponentHostCommunicationApi, ea = () => {
	let e = X.closeSidePanel;
	if (!o(e)) throw Error("closeSidePanelFunction is not set");
	return e();
}, ta = (e) => {
	let t = X.copyToClipboard;
	if (!o(t)) throw Error("copyToClipboardFunction is not set");
	return t(e);
}, na = (e) => {
	let t = process.env.applicationVariables;
	if (t) return JSON.parse(t)[e];
}, ra = (e) => {
	let t = X.enqueueSnackbar;
	if (!o(t)) throw Error("enqueueSnackbarFunction is not set");
	return t(e);
}, ia = (e, t, n, r) => {
	let i = X.navigate;
	if (!o(i)) throw Error("navigateFunction is not set");
	return i(e, t, n, r);
}, aa = (e) => {
	let t = X.openCommandConfirmationModal;
	if (!o(t)) throw Error("openCommandConfirmationModalFunction is not set");
	return t(e);
}, oa = (e) => {
	let t = X.openSidePanelPage;
	if (!o(t)) throw Error("openSidePanelPageFunction is not set");
	return t(e);
}, Z = () => {
	let e = X.unmountFrontComponent;
	if (!o(e)) throw Error("unmountFrontComponentFunction is not set");
	return e();
}, sa = (e) => {
	let t = X.updateProgress;
	if (!o(t)) throw Error("updateProgressFunction is not set");
	return t(e);
}, ca = "__twentySdkExecutionContext__", la = "__twentySdkContextListeners__", ua = () => (globalThis.__twentySdkContextListeners__ || (globalThis[la] = /* @__PURE__ */ new Set()), globalThis[la]), da = () => globalThis[ca], fa = (e) => {
	ua().add(e);
}, pa = (e) => {
	ua().delete(e);
}, Q = (e) => {
	let [t, n] = a(() => e(da())), o = i(t);
	return r(() => {
		let t = () => {
			let t = e(da());
			t !== o.current && (o.current = t, n(t));
		};
		return fa(t), t(), () => pa(t);
	}, [e]), t;
}, ma = (e) => e.colorScheme, ha = () => Q(ma), ga = (e) => e.frontComponentId, $ = () => Q(ga), _a = (e) => e.selectedRecordIds.length === 1 ? e.selectedRecordIds[0] : null, va = () => Q(_a), ya = (e) => e.selectedRecordIds, ba = () => Q(ya), xa = (e) => e.userId, Sa = () => Q(xa), Ca = (e) => `${e}-command-error`, wa = ({ execute: e }) => {
	let [t, n] = a(!1);
	return r(() => {
		t || (n(!0), (async () => {
			await e(), await Z();
		})());
	}, [
		e,
		t,
		$()
	]), null;
}, Ta = ({ to: e, params: t, queryParams: n, options: i }) => {
	let [o, s] = a(!1);
	return r(() => {
		o || (s(!0), (async () => {
			await ia(e, t, n, i), await Z();
		})());
	}, [
		e,
		t,
		n,
		i,
		o,
		$()
	]), null;
}, Ea = ({ page: e, pageTitle: t, pageIcon: n, onClick: i, shouldResetSearchState: o = !1 }) => {
	let [s, c] = a(!1);
	return r(() => {
		s || (c(!0), (async () => {
			i?.(), await oa({
				page: e,
				pageTitle: t,
				pageIcon: n,
				shouldResetSearchState: o
			}), await Z();
		})());
	}, [
		e,
		t,
		n,
		o,
		i,
		s,
		$()
	]), null;
}, Da = ({ title: e, subtitle: t, execute: n, confirmButtonText: i, confirmButtonAccent: o }) => {
	let [s, c] = a(!1);
	return r(() => {
		s || (c(!0), (async () => {
			await aa({
				title: e,
				subtitle: t,
				confirmButtonText: i,
				confirmButtonAccent: o
			}) === "confirm" && await n(), await Z();
		})());
	}, [
		e,
		t,
		n,
		i,
		o,
		s,
		$()
	]), null;
};
//#endregion
export { Hi as AppPath, wa as Command, Ta as CommandLink, Da as CommandModal, Ea as CommandOpenSidePanelPage, Qi as SidePanelPages, ea as closeSidePanel, ta as copyToClipboard, ra as enqueueSnackbar, na as getApplicationVariable, Ca as getFrontComponentCommandErrorDedupeKey, ia as navigate, aa as openCommandConfirmationModal, oa as openSidePanelPage, Z as unmountFrontComponent, sa as updateProgress, ha as useColorScheme, Q as useFrontComponentExecutionContext, $ as useFrontComponentId, va as useRecordId, ba as useSelectedRecordIds, Sa as useUserId };
