// twenty-client-sdk/dist/rest.mjs
import { n as e, t } from "./application-CbP6KYwU.mjs";
//#region src/rest/index.ts
var n = (e) => e != null, r = class extends Error {
	constructor(e, t) {
		super(e), this.name = "RestApiClientError", this.status = t?.status, this.statusText = t?.statusText, this.url = t?.url, this.body = t?.body;
	}
}, i = () => globalThis.process?.env ?? {}, a = (e, t, r) => {
	let i = /^https?:\/\//i.test(t) ? t : `${e}${t.startsWith("/") ? "" : "/"}${t}`;
	if (!n(r)) return i;
	let a = new URLSearchParams();
	for (let [e, t] of Object.entries(r)) n(t) && a.append(e, String(t));
	let o = a.toString();
	return o.length === 0 ? i : i.includes("?") ? `${i}&${o}` : `${i}?${o}`;
}, o = class {
	constructor(e) {
		this.refreshAccessTokenPromise = null, this.baseUrl = e?.baseUrl, this.token = e?.token, this.defaultHeaders = e?.defaultHeaders, this.fetchImplementation = e?.fetch ?? globalThis.fetch ?? null, this.authorizationToken = e?.token ?? null;
	}
	request(e, t, n) {
		return this.execute(e, t, n?.body, n);
	}
	get(e, t) {
		return this.execute("GET", e, void 0, t);
	}
	post(e, t, n) {
		return this.execute("POST", e, t, n);
	}
	put(e, t, n) {
		return this.execute("PUT", e, t, n);
	}
	patch(e, t, n) {
		return this.execute("PATCH", e, t, n);
	}
	delete(e, t) {
		return this.execute("DELETE", e, void 0, t);
	}
	resolveBaseUrl() {
		let t = this.baseUrl ?? i().TWENTY_API_URL;
		if (!n(t) || t.trim().length === 0) throw new r(`Missing API url. Set the \`${e}\` environment variable or pass \`baseUrl\` to \`RestApiClient\`.`);
		return t.replace(/\/+$/, "");
	}
	resolveToken() {
		if (!n(this.authorizationToken)) {
			let e = i();
			this.authorizationToken = this.token ?? e.TWENTY_APP_ACCESS_TOKEN ?? e.TWENTY_API_KEY ?? null;
		}
		if (!n(this.authorizationToken) || this.authorizationToken.length === 0) throw new r(`Missing application access token. Set the \`${t}\` environment variable or pass \`token\` to \`RestApiClient\`.`);
		return this.authorizationToken;
	}
	async requestRefreshedAccessToken() {
		let e = globalThis.frontComponentHostCommunicationApi?.requestAccessTokenRefresh;
		return typeof e == "function" ? (this.refreshAccessTokenPromise ||= e().then((e) => typeof e != "string" || e.length === 0 ? null : (this.authorizationToken = e, i()[t] = e, e)).catch((e) => (console.error("Twenty REST client: token refresh failed", e), null)).finally(() => {
			this.refreshAccessTokenPromise = null;
		}), this.refreshAccessTokenPromise) : null;
	}
	sendRequest(e, t, i, a, o) {
		if (!n(this.fetchImplementation)) throw new r("Global `fetch` function is not available, pass a fetch implementation to `RestApiClient`.");
		let s = new Headers(this.defaultHeaders);
		n(o?.headers) && new Headers(o.headers).forEach((e, t) => s.set(t, e));
		let c = typeof FormData < "u" && i instanceof FormData, l = n(i) && !c && typeof i != "string";
		c ? s.delete("Content-Type") : l && s.set("Content-Type", "application/json"), s.set("Authorization", `Bearer ${a}`);
		let u = n(i) ? c || typeof i == "string" ? i : JSON.stringify(i) : void 0;
		return this.fetchImplementation.call(globalThis, e, {
			method: t,
			headers: s,
			body: u,
			signal: o?.signal
		});
	}
	async parseResponse(e, t) {
		let n = await e.text(), i;
		if (n.trim().length > 0) try {
			i = JSON.parse(n);
		} catch {
			i = n;
		}
		if (!e.ok) throw new r(`Request to ${t} failed with status ${e.status} ${e.statusText}`, {
			status: e.status,
			statusText: e.statusText,
			url: t,
			body: i
		});
		return i;
	}
	async execute(e, t, r, i) {
		let o = a(this.resolveBaseUrl(), t, i?.query), s = this.resolveToken(), c = await this.sendRequest(o, e, r, s, i);
		if (c.status === 401) {
			let t = await this.requestRefreshedAccessToken();
			n(t) && (c = await this.sendRequest(o, e, r, t, i));
		}
		return this.parseResponse(c, o);
	}
};
//#endregion
export { o as RestApiClient, r as RestApiClientError };
