// twenty-sdk/dist/logic-function/get-function-input-schema-GNk3NRLJ-DSc3lI49.mjs
import { t as e } from "./isDefined-Dtu5EYqP-CMNeFg4z.mjs";
import { ScriptTarget as t, SyntaxKind as n, createSourceFile as r } from "typescript";
//#region ../twenty-shared/dist/get-function-input-schema-GNk3NRLJ.mjs
var i = "TwentyRecord", a = (t) => {
	let r = t.typeArguments?.[0];
	if (e(r) && r.kind === n.LiteralType && r.literal.kind === n.StringLiteral) return r.literal.text;
}, o = (e) => e.type === "record" ? {
	type: "records",
	objectUniversalIdentifier: e.objectUniversalIdentifier
} : {
	type: "array",
	items: e
}, s = (t) => {
	switch (t.kind) {
		case n.NumberKeyword: return { type: "number" };
		case n.StringKeyword: return { type: "string" };
		case n.BooleanKeyword: return { type: "boolean" };
		case n.ArrayType: return o(s(t.elementType));
		case n.TypeReference: {
			let r = t, c = r.typeName.kind === n.Identifier ? r.typeName.text : void 0;
			if (c === "Array" || c === "ReadonlyArray") {
				let t = r.typeArguments?.[0];
				return o(e(t) ? s(t) : {});
			}
			if (c === i) {
				let t = a(r);
				if (e(t)) return {
					type: "record",
					objectUniversalIdentifier: t
				};
			}
			return {};
		}
		case n.ObjectKeyword: return { type: "object" };
		case n.TypeLiteral: {
			let n = {};
			return t.members.forEach((t) => {
				if (e(t.name) && e(t.type)) {
					let e = t.name.text;
					n[e] = s(t.type);
				}
			}), {
				type: "object",
				properties: n
			};
		}
		case n.UnionType: {
			let e = t, r = [], i = !0;
			return e.types.forEach((e) => {
				if (e.kind === n.LiteralType) {
					let t = e.literal;
					t.kind === n.StringLiteral ? r.push(t.text) : i = !1;
				} else i = !1;
			}), i ? {
				type: "string",
				enum: r
			} : {};
		}
		default: return {};
	}
}, c = (t, n) => t.parameters.reduce((t, n) => {
	let r = n.type;
	return e(r) ? [...t, s(r)] : [...t, {}];
}, n), l = (t) => t.kind === n.FunctionDeclaration ? [t] : t.kind === n.VariableStatement ? t.declarationList.declarations.filter((t) => e(t.initializer) && t.initializer.kind === n.ArrowFunction).map((e) => e.initializer) : [], u = (e) => {
	let i = r("temp.ts", e, t.ESNext, !0), a = [];
	return i.forEachChild((e) => {
		(e.kind === n.FunctionDeclaration || e.kind === n.VariableStatement) && l(e).forEach((e) => {
			a = c(e, a);
		});
	}), a;
};
//#endregion
export { u as getFunctionInputSchema };
