// twenty-sdk/dist/logic-function/index.mjs
import { t as e } from "./isDefined-Dtu5EYqP-CMNeFg4z.mjs";
import { isNonEmptyString as t, isObject as n } from "@sniptt/guards";
import "uuid";
//#region ../twenty-shared/dist/is-record-object-schema-CwzshFdt.mjs
var r = (e) => (e?.type === "record" || e?.type === "object") && t(e.objectUniversalIdentifier), i = (e, n) => {
	let r = t(e) ? n?.(e) : void 0;
	return `Id of the ${t(r) ? r : "linked"} record`;
}, a = (t, o) => {
	if (r(t)) return {
		type: "string",
		description: i(t.objectUniversalIdentifier, o)
	};
	if (t.type === "records") return {
		type: "array",
		items: {
			type: "string",
			description: i(t.objectUniversalIdentifier, o)
		}
	};
	let { objectUniversalIdentifier: s, multiline: c, label: l, items: u, properties: d, additionalProperties: f, ...p } = t, m = { ...p };
	return e(u) && (m.items = a(u, o)), e(d) && (m.properties = Object.fromEntries(Object.entries(d).map(([e, t]) => [e, a(t, o)]))), e(f) && (m.additionalProperties = n(f) ? a(f, o) : f), m;
}, o = (e) => {
	let n = { type: "unknown" };
	switch (e.type) {
		case "string":
			n.type = "string";
			break;
		case "number":
		case "integer":
			n.type = "number";
			break;
		case "boolean":
			n.type = "boolean";
			break;
		case "array":
			n.type = "array", e.items && (n.items = o(e.items));
			break;
		case "object":
			n.type = "object", e.properties && (n.properties = Object.fromEntries(Object.entries(e.properties).map(([e, t]) => [e, o(t)])));
			break;
		case "record":
			n.type = "record";
			break;
		case "records":
			n.type = "records";
			break;
		default: n.type = "unknown";
	}
	return Array.isArray(e.enum) && (n.enum = e.enum.filter((e) => typeof e == "string")), e.multiline === !0 && (n.multiline = !0), t(e.label) && (n.label = e.label), t(e.objectUniversalIdentifier) && (n.objectUniversalIdentifier = e.objectUniversalIdentifier), n;
}, s = (e) => [o(e)];
s({
	type: "object",
	properties: {
		a: { type: "string" },
		b: { type: "number" }
	}
});
//#endregion
//#region ../twenty-shared/dist/logic-function.mjs
var c = (t) => !e(t) || t === null ? "unknown" : typeof t == "string" ? "string" : typeof t == "number" ? "number" : typeof t == "boolean" ? "boolean" : Array.isArray(t) ? "array" : "unknown", l = (e) => e ? Object.entries(e).reduce((e, [t, r]) => (n(r) && !Array.isArray(r) ? e[t] = {
	isLeaf: !1,
	type: "object",
	label: t,
	value: l(r)
} : e[t] = {
	isLeaf: !0,
	value: r,
	type: c(r),
	label: t
}, e), {}) : {}, u = (e, t) => e ? `${e}.${t}` : t, d = (t, n, r = "") => {
	let i = [];
	for (let [a, o] of Object.entries(n)) {
		let n = u(r, a), s = t[a];
		if (!e(s)) {
			i.push(`Missing key "${n}" in declared output schema.`);
			continue;
		}
		if (o.isLeaf !== s.isLeaf) {
			i.push(`Type mismatch at "${n}": expected ${o.isLeaf ? o.type : "object"} but declared ${s.isLeaf ? s.type : "object"}.`);
			continue;
		}
		if (!o.isLeaf && !s.isLeaf) {
			i.push(...d(s.value, o.value, n));
			continue;
		}
		o.isLeaf && s.isLeaf && o.type !== "unknown" && s.type !== "unknown" && o.type !== s.type && i.push(`Type mismatch at "${n}": expected ${o.type} but declared ${s.type}.`);
	}
	return i;
}, f = [
	"string",
	"number",
	"boolean",
	"array",
	"unknown"
], p = (e) => f.includes(e), m = (e, t) => {
	let r = t.label ?? e;
	return t.type === "record" ? {
		isLeaf: !0,
		type: "string",
		label: r,
		value: null
	} : t.type === "records" ? {
		isLeaf: !0,
		type: "array",
		label: r,
		value: null
	} : t.type === "object" ? {
		isLeaf: !1,
		type: "object",
		label: r,
		value: n(t.properties) ? h(t.properties) : {}
	} : {
		isLeaf: !0,
		type: p(t.type) ? t.type : "unknown",
		label: r,
		value: null
	};
}, h = (e) => Object.entries(e).reduce((e, [t, n]) => (e[t] = m(t, n), e), {}), g = class extends Error {
	constructor(e) {
		super(`App connection ${e} requires the user to reconnect (authFailedAt is set). Surface a "Reconnect" prompt in your UI.`), this.name = "AppConnectionAuthFailedError", this.connectionId = e;
	}
}, _ = "TWENTY_API_URL", v = "TWENTY_APP_ACCESS_TOKEN", y = async ({ query: e, variables: t, caller: n }) => {
	let r = process.env[_], i = process.env[v];
	if (!r || !i) throw Error(`${n}() requires the app runtime env vars ${_} and ${v}.`);
	let a = await fetch(`${r}/metadata`, {
		method: "POST",
		headers: {
			"Content-Type": "application/json",
			Authorization: `Bearer ${i}`
		},
		body: JSON.stringify({
			query: e,
			variables: t
		})
	});
	if (!a.ok) throw Error(`${n}() failed: HTTP ${a.status} ${a.statusText}`);
	let o = await a.json();
	if (o.errors && o.errors.length > 0) throw Error(`${n}() failed: ${o.errors.map((e) => e.message).join(", ")}`);
	if (!o.data) throw Error(`${n}() failed: response contained no data.`);
	return o.data;
}, b = "\n  query GetAppConnection($id: ID!) {\n    appConnection(id: $id) {\n      id\n      providerName\n      name\n      handle\n      visibility\n      userWorkspaceId\n      accessToken\n      scopes\n      authFailedAt\n    }\n  }\n", x = async (e) => {
	let { appConnection: t } = await y({
		query: b,
		variables: { id: e },
		caller: "getConnection"
	});
	if (t.authFailedAt !== null) throw new g(t.id);
	return t;
}, S = "\n  query ListAppConnections($filter: ListAppConnectionsInput) {\n    appConnections(filter: $filter) {\n      id\n      providerName\n      name\n      handle\n      visibility\n      userWorkspaceId\n      accessToken\n      scopes\n      authFailedAt\n    }\n  }\n", C = async (e = {}) => {
	let { appConnections: t } = await y({
		query: S,
		variables: { filter: e },
		caller: "listConnections"
	});
	return t;
}, w = (e, t) => {
	if (t.userWorkspaceId !== null) {
		let n = e.find((e) => e.visibility === "user" && e.userWorkspaceId === t.userWorkspaceId);
		if (n) return n;
	}
	return e.find((e) => e.visibility === "workspace") ?? null;
}, T = "\n  mutation RunAgent($input: RunAgentInput!) {\n    runAgent(input: $input) {\n      result\n      error\n      success\n    }\n  }\n", E = async (e) => {
	let { runAgent: t } = await y({
		query: T,
		variables: { input: e },
		caller: "runAgent"
	});
	return t;
}, D = class {
	constructor(e, t) {
		this.__twentyHttpResponse = !0, this.body = e, this.status = t?.status, this.headers = t?.headers;
	}
};
//#endregion
export { g as AppConnectionAuthFailedError, D as Response, w as findConnectionForRequest, x as getConnection, s as jsonSchemaToInputSchema, C as listConnections, E as runAgent };
