// twenty-sdk/dist/cli.mjs
#!/usr/bin/env node
import { $ as e, A as t, B as n, D as r, E as i, F as a, H as o, I as s, K as c, L as l, M as u, N as d, P as f, Q as p, R as m, S as h, U as g, V as _, W as v, X as y, Y as b, Z as x, _ as ee, _t as te, a as ne, at as re, b as ie, bt as ae, c as oe, ct as S, d as se, dt as ce, et as le, f as ue, ft as de, g as fe, gt as pe, h as me, ht as C, it as he, j as ge, l as _e, lt as ve, m as ye, mt as be, n as xe, nt as Se, o as Ce, ot as we, p as Te, pt as Ee, rt as De, s as Oe, st as ke, t as Ae, tt as je, u as Me, ut as Ne, v as Pe, vt as Fe, w as Ie, x as Le, y as Re, yt as w, z as ze } from "./login-rU5te3B6.mjs";
import Be, { dirname as Ve, join as T, relative as He } from "path";
import { readFile as Ue, writeFile as We } from "node:fs/promises";
import { v4 as E } from "uuid";
import Ge, { useCallback as Ke, useEffect as qe, useReducer as Je, useRef as Ye } from "react";
import Xe from "node:path";
import Ze from "crypto";
import { execSync as Qe, spawn as $e, spawnSync as et } from "node:child_process";
import D from "chalk";
import tt from "inquirer";
import nt from "chokidar";
import { Fragment as rt, jsx as O, jsxs as k } from "react/jsx-runtime";
import { Command as it, CommanderError as at } from "commander";
import { inspect as ot } from "util";
//#region src/cli/utilities/file/file-path.ts
var A = (e) => e && !e?.startsWith("/") ? T(w, e) : e, st = async (e) => {
	let t = T(e, "package.json");
	if (await re(t)) try {
		let e = await Ue(t, "utf-8"), n = JSON.parse(e);
		return n.dependencies?.["twenty-sdk"] ?? n.devDependencies?.["twenty-sdk"] ?? void 0;
	} catch {
		return;
	}
}, ct = (e) => {
	let t = e.match(/(\d+)\./);
	if (t) return parseInt(t[1], 10);
	let n = e.match(/^[~^>=<]*(\d+)$/);
	return n ? parseInt(n[1], 10) : null;
}, lt = async (e) => {
	let t = f.version, n = await st(e);
	if (!t || !n) return;
	let r = ct(t), i = ct(n);
	r === null || i === null || r !== i && (console.warn(D.yellow(`⚠ Version mismatch: your app requires twenty-sdk@${i}.x but the CLI running is v${t}. Major version mismatches may cause unexpected behavior.\n  Update with: npm install -g twenty-sdk@${i}`)), console.warn(""));
}, ut = class {
	async execute(e) {
		e.private ? await this.executePrivate(e) : await this.executeNpm(e);
	}
	async executeNpm(e) {
		let t = e.appPath ?? w;
		await lt(t), console.log(D.blue("Publishing to npm...")), console.log(D.gray(`App path: ${t}\n`));
		let n = await a({
			appPath: t,
			npmTag: e.tag,
			onProgress: (e) => console.log(D.gray(e))
		});
		n.success || (console.error(D.red(n.error.message)), process.exit(1)), console.log(D.green("✓ Published to npm successfully"));
	}
	async executePrivate(e) {
		let t = e.appPath ?? w;
		await lt(t);
		let n = e.remote ?? _.getActiveRemote();
		console.log(D.blue(`Deploying application on ${n}...`)), console.log(D.gray(`App path: ${t}\n`));
		let r = (e) => console.log(D.gray(e)), i = await g({
			appPath: t,
			tarball: !0,
			onProgress: r
		});
		i.success || (console.error(D.red(i.error.message)), process.exit(1));
		let a = await s({
			tarballPath: i.data.tarballPath,
			remote: e.remote,
			onProgress: r
		});
		a.success || (console.error(D.red(a.error.message)), process.exit(1));
		let o = await we(Be.join(t, "package.json")).catch(() => void 0), c = o?.name ?? a.data.name, l = o?.version ?? "unknown";
		console.log(D.green(`\n✓ Published ${c} v${l} to ${n}\n`)), console.log("  To install deployed application: `yarn twenty app:install`");
	}
}, dt = class {
	async execute(e) {
		let t = e.appPath ?? w;
		await lt(t), await ee();
		let n = e.remote ?? _.getActiveRemote();
		console.log(D.blue(`Installing application on ${n}...`)), console.log(D.gray(`App path: ${t}\n`));
		let r = await ge({
			appPath: t,
			remote: e.remote
		});
		r.success || (console.error(D.red(r.error.message)), process.exit(1)), console.log(D.green("✓ Application installed"));
	}
}, ft = class {
	async execute({ appPath: e = w, askForConfirmation: t }) {
		let n = _.getActiveRemote();
		console.log(D.blue(`🚀 Uninstall Twenty Application on ${n}`)), console.log(D.gray(`📁 App Path: ${e}\n`)), t && !await this.confirmationPrompt() && (console.error(D.red("⛔️ Aborting uninstall")), process.exit(1));
		let r = await fe({ appPath: e });
		return r.success ? (console.log(D.green("✅ Application uninstalled successfully")), {
			success: !0,
			data: void 0
		}) : (console.error(D.red("❌ Uninstall failed:"), r.error.message), {
			success: !1,
			error: r.error.message
		});
	}
	async confirmationPrompt() {
		let { confirmation: e } = await tt.prompt([{
			type: "confirm",
			name: "confirmation",
			message: "Are you sure you want to uninstall this application?",
			default: !1
		}]);
		return e;
	}
}, pt = (e) => {
	let t = new ut(), n = new dt(), r = new ft();
	e.command("app:publish [appPath]").description("Build and publish to npm (default) or server registry").option("--private", "Push to a Twenty server's registry instead of npm").option("-r, --remote <name>", "Publish to a specific remote (with --private)").option("--tag <tag>", "npm dist-tag (e.g. beta, next)").action(async (e, n) => {
		await t.execute({
			appPath: A(e),
			private: n.private,
			remote: n.remote,
			tag: n.tag
		});
	}), e.command("app:install [appPath]").description("Install a deployed app on the server").option("-r, --remote <name>", "Install on a specific remote").action(async (e, t) => {
		await n.execute({
			appPath: A(e),
			remote: t.remote
		});
	}), e.command("app:uninstall [appPath]").description("Uninstall app from server").option("-y, --yes", "Skip confirmation prompt").action(async (e, t) => {
		try {
			let n = await r.execute({
				appPath: A(e),
				askForConfirmation: !t?.yes
			});
			process.exit(+!n.success);
		} catch {
			process.exit(1);
		}
	});
}, mt = (/* @__PURE__ */ ae(((e, t) => {
	(function() {
		var n, r = "4.18.1", i = 200, a = "Unsupported core-js use. Try https://npms.io/search?q=ponyfill.", o = "Expected a function", s = "Invalid `variable` option passed into `_.template`", c = "Invalid `imports` option passed into `_.template`", l = "__lodash_hash_undefined__", u = 500, d = "__lodash_placeholder__", f = 1, p = 2, m = 4, h = 1, g = 2, _ = 1, v = 2, y = 4, b = 8, x = 16, ee = 32, te = 64, ne = 128, re = 256, ie = 512, ae = 30, oe = "...", S = 800, se = 16, ce = 1, le = 2, ue = 3, de = Infinity, fe = 9007199254740991, pe = 17976931348623157e292, me = NaN, C = 4294967295, he = C - 1, ge = C >>> 1, _e = [
			["ary", ne],
			["bind", _],
			["bindKey", v],
			["curry", b],
			["curryRight", x],
			["flip", ie],
			["partial", ee],
			["partialRight", te],
			["rearg", re]
		], ve = "[object Arguments]", ye = "[object Array]", be = "[object AsyncFunction]", xe = "[object Boolean]", Se = "[object Date]", Ce = "[object DOMException]", we = "[object Error]", Te = "[object Function]", Ee = "[object GeneratorFunction]", De = "[object Map]", Oe = "[object Number]", ke = "[object Null]", Ae = "[object Object]", je = "[object Promise]", Me = "[object Proxy]", Ne = "[object RegExp]", Pe = "[object Set]", Fe = "[object String]", Ie = "[object Symbol]", Le = "[object Undefined]", Re = "[object WeakMap]", w = "[object WeakSet]", ze = "[object ArrayBuffer]", Be = "[object DataView]", Ve = "[object Float32Array]", T = "[object Float64Array]", He = "[object Int8Array]", Ue = "[object Int16Array]", We = "[object Int32Array]", E = "[object Uint8Array]", Ge = "[object Uint8ClampedArray]", Ke = "[object Uint16Array]", qe = "[object Uint32Array]", Je = /\b__p \+= '';/g, Ye = /\b(__p \+=) '' \+/g, Xe = /(__e\(.*?\)|\b__t\)) \+\n'';/g, Ze = /&(?:amp|lt|gt|quot|#39);/g, Qe = /[&<>"']/g, $e = RegExp(Ze.source), et = RegExp(Qe.source), D = /<%-([\s\S]+?)%>/g, tt = /<%([\s\S]+?)%>/g, nt = /<%=([\s\S]+?)%>/g, rt = /\.|\[(?:[^[\]]*|(["'])(?:(?!\1)[^\\]|\\.)*?\1)\]/, O = /^\w*$/, k = /[^.[\]]+|\[(?:(-?\d+(?:\.\d+)?)|(["'])((?:(?!\2)[^\\]|\\.)*?)\2)\]|(?=(?:\.|\[\])(?:\.|\[\]|$))/g, it = /[\\^$.*+?()[\]{}|]/g, at = RegExp(it.source), ot = /^\s+/, A = /\s/, st = /\{(?:\n\/\* \[wrapped with .+\] \*\/)?\n?/, ct = /\{\n\/\* \[wrapped with (.+)\] \*/, lt = /,? & /, ut = /[^\x00-\x2f\x3a-\x40\x5b-\x60\x7b-\x7f]+/g, dt = /[()=,{}\[\]\/\s]/, ft = /\\(\\)?/g, pt = /\$\{([^\\}]*(?:\\.[^\\}]*)*)\}/g, mt = /\w*$/, ht = /^[-+]0x[0-9a-f]+$/i, gt = /^0b[01]+$/i, _t = /^\[object .+?Constructor\]$/, vt = /^0o[0-7]+$/i, yt = /^(?:0|[1-9]\d*)$/, bt = /[\xc0-\xd6\xd8-\xf6\xf8-\xff\u0100-\u017f]/g, xt = /($^)/, St = /['\n\r\u2028\u2029\\]/g, Ct = "\\ud800-\\udfff", j = "\\u0300-\\u036f\\ufe20-\\ufe2f\\u20d0-\\u20ff", wt = "\\u2700-\\u27bf", M = "a-z\\xdf-\\xf6\\xf8-\\xff", Tt = "\\xac\\xb1\\xd7\\xf7", Et = "\\x00-\\x2f\\x3a-\\x40\\x5b-\\x60\\x7b-\\xbf", Dt = "\\u2000-\\u206f", Ot = " \\t\\x0b\\f\\xa0\\ufeff\\n\\r\\u2028\\u2029\\u1680\\u180e\\u2000\\u2001\\u2002\\u2003\\u2004\\u2005\\u2006\\u2007\\u2008\\u2009\\u200a\\u202f\\u205f\\u3000", kt = "A-Z\\xc0-\\xd6\\xd8-\\xde", At = "\\ufe0e\\ufe0f", jt = Tt + Et + Dt + Ot, Mt = "['’]", N = "[" + Ct + "]", Nt = "[" + jt + "]", Pt = "[" + j + "]", Ft = "\\d+", It = "[" + wt + "]", Lt = "[" + M + "]", Rt = "[^" + Ct + jt + Ft + wt + M + kt + "]", zt = "\\ud83c[\\udffb-\\udfff]", Bt = "(?:" + Pt + "|" + zt + ")", Vt = "[^" + Ct + "]", Ht = "(?:\\ud83c[\\udde6-\\uddff]){2}", Ut = "[\\ud800-\\udbff][\\udc00-\\udfff]", Wt = "[" + kt + "]", Gt = "\\u200d", Kt = "(?:" + Lt + "|" + Rt + ")", qt = "(?:" + Wt + "|" + Rt + ")", Jt = "(?:" + Mt + "(?:d|ll|m|re|s|t|ve))?", Yt = "(?:" + Mt + "(?:D|LL|M|RE|S|T|VE))?", Xt = Bt + "?", Zt = "[" + At + "]?", Qt = "(?:" + Gt + "(?:" + [
			Vt,
			Ht,
			Ut
		].join("|") + ")" + Zt + Xt + ")*", $t = "\\d*(?:1st|2nd|3rd|(?![123])\\dth)(?=\\b|[A-Z_])", en = "\\d*(?:1ST|2ND|3RD|(?![123])\\dTH)(?=\\b|[a-z_])", tn = Zt + Xt + Qt, nn = "(?:" + [
			It,
			Ht,
			Ut
		].join("|") + ")" + tn, rn = "(?:" + [
			Vt + Pt + "?",
			Pt,
			Ht,
			Ut,
			N
		].join("|") + ")", an = RegExp(Mt, "g"), on = RegExp(Pt, "g"), sn = RegExp(zt + "(?=" + zt + ")|" + rn + tn, "g"), cn = RegExp([
			Wt + "?" + Lt + "+" + Jt + "(?=" + [
				Nt,
				Wt,
				"$"
			].join("|") + ")",
			qt + "+" + Yt + "(?=" + [
				Nt,
				Wt + Kt,
				"$"
			].join("|") + ")",
			Wt + "?" + Kt + "+" + Jt,
			Wt + "+" + Yt,
			en,
			$t,
			Ft,
			nn
		].join("|"), "g"), ln = RegExp("[" + Gt + Ct + j + At + "]"), P = /[a-z][A-Z]|[A-Z]{2}[a-z]|[0-9][a-zA-Z]|[a-zA-Z][0-9]|[^a-zA-Z0-9 ]/, un = /* @__PURE__ */ "Array.Buffer.DataView.Date.Error.Float32Array.Float64Array.Function.Int8Array.Int16Array.Int32Array.Map.Math.Object.Promise.RegExp.Set.String.Symbol.TypeError.Uint8Array.Uint8ClampedArray.Uint16Array.Uint32Array.WeakMap._.clearTimeout.isFinite.parseInt.setTimeout".split("."), dn = -1, F = {};
		F[Ve] = F[T] = F[He] = F[Ue] = F[We] = F[E] = F[Ge] = F[Ke] = F[qe] = !0, F[ve] = F[ye] = F[ze] = F[xe] = F[Be] = F[Se] = F[we] = F[Te] = F[De] = F[Oe] = F[Ae] = F[Ne] = F[Pe] = F[Fe] = F[Re] = !1;
		var I = {};
		I[ve] = I[ye] = I[ze] = I[Be] = I[xe] = I[Se] = I[Ve] = I[T] = I[He] = I[Ue] = I[We] = I[De] = I[Oe] = I[Ae] = I[Ne] = I[Pe] = I[Fe] = I[Ie] = I[E] = I[Ge] = I[Ke] = I[qe] = !0, I[we] = I[Te] = I[Re] = !1;
		var L = {
			À: "A",
			Á: "A",
			Â: "A",
			Ã: "A",
			Ä: "A",
			Å: "A",
			à: "a",
			á: "a",
			â: "a",
			ã: "a",
			ä: "a",
			å: "a",
			Ç: "C",
			ç: "c",
			Ð: "D",
			ð: "d",
			È: "E",
			É: "E",
			Ê: "E",
			Ë: "E",
			è: "e",
			é: "e",
			ê: "e",
			ë: "e",
			Ì: "I",
			Í: "I",
			Î: "I",
			Ï: "I",
			ì: "i",
			í: "i",
			î: "i",
			ï: "i",
			Ñ: "N",
			ñ: "n",
			Ò: "O",
			Ó: "O",
			Ô: "O",
			Õ: "O",
			Ö: "O",
			Ø: "O",
			ò: "o",
			ó: "o",
			ô: "o",
			õ: "o",
			ö: "o",
			ø: "o",
			Ù: "U",
			Ú: "U",
			Û: "U",
			Ü: "U",
			ù: "u",
			ú: "u",
			û: "u",
			ü: "u",
			Ý: "Y",
			ý: "y",
			ÿ: "y",
			Æ: "Ae",
			æ: "ae",
			Þ: "Th",
			þ: "th",
			ß: "ss",
			Ā: "A",
			Ă: "A",
			Ą: "A",
			ā: "a",
			ă: "a",
			ą: "a",
			Ć: "C",
			Ĉ: "C",
			Ċ: "C",
			Č: "C",
			ć: "c",
			ĉ: "c",
			ċ: "c",
			č: "c",
			Ď: "D",
			Đ: "D",
			ď: "d",
			đ: "d",
			Ē: "E",
			Ĕ: "E",
			Ė: "E",
			Ę: "E",
			Ě: "E",
			ē: "e",
			ĕ: "e",
			ė: "e",
			ę: "e",
			ě: "e",
			Ĝ: "G",
			Ğ: "G",
			Ġ: "G",
			Ģ: "G",
			ĝ: "g",
			ğ: "g",
			ġ: "g",
			ģ: "g",
			Ĥ: "H",
			Ħ: "H",
			ĥ: "h",
			ħ: "h",
			Ĩ: "I",
			Ī: "I",
			Ĭ: "I",
			Į: "I",
			İ: "I",
			ĩ: "i",
			ī: "i",
			ĭ: "i",
			į: "i",
			ı: "i",
			Ĵ: "J",
			ĵ: "j",
			Ķ: "K",
			ķ: "k",
			ĸ: "k",
			Ĺ: "L",
			Ļ: "L",
			Ľ: "L",
			Ŀ: "L",
			Ł: "L",
			ĺ: "l",
			ļ: "l",
			ľ: "l",
			ŀ: "l",
			ł: "l",
			Ń: "N",
			Ņ: "N",
			Ň: "N",
			Ŋ: "N",
			ń: "n",
			ņ: "n",
			ň: "n",
			ŋ: "n",
			Ō: "O",
			Ŏ: "O",
			Ő: "O",
			ō: "o",
			ŏ: "o",
			ő: "o",
			Ŕ: "R",
			Ŗ: "R",
			Ř: "R",
			ŕ: "r",
			ŗ: "r",
			ř: "r",
			Ś: "S",
			Ŝ: "S",
			Ş: "S",
			Š: "S",
			ś: "s",
			ŝ: "s",
			ş: "s",
			š: "s",
			Ţ: "T",
			Ť: "T",
			Ŧ: "T",
			ţ: "t",
			ť: "t",
			ŧ: "t",
			Ũ: "U",
			Ū: "U",
			Ŭ: "U",
			Ů: "U",
			Ű: "U",
			Ų: "U",
			ũ: "u",
			ū: "u",
			ŭ: "u",
			ů: "u",
			ű: "u",
			ų: "u",
			Ŵ: "W",
			ŵ: "w",
			Ŷ: "Y",
			ŷ: "y",
			Ÿ: "Y",
			Ź: "Z",
			Ż: "Z",
			Ž: "Z",
			ź: "z",
			ż: "z",
			ž: "z",
			Ĳ: "IJ",
			ĳ: "ij",
			Œ: "Oe",
			œ: "oe",
			ŉ: "'n",
			ſ: "s"
		}, fn = {
			"&": "&amp;",
			"<": "&lt;",
			">": "&gt;",
			"\"": "&quot;",
			"'": "&#39;"
		}, pn = {
			"&amp;": "&",
			"&lt;": "<",
			"&gt;": ">",
			"&quot;": "\"",
			"&#39;": "'"
		}, mn = {
			"\\": "\\",
			"'": "'",
			"\n": "n",
			"\r": "r",
			"\u2028": "u2028",
			"\u2029": "u2029"
		}, hn = parseFloat, gn = parseInt, _n = typeof global == "object" && global && global.Object === Object && global, vn = typeof self == "object" && self && self.Object === Object && self, R = _n || vn || Function("return this")(), yn = typeof e == "object" && e && !e.nodeType && e, bn = yn && typeof t == "object" && t && !t.nodeType && t, xn = bn && bn.exports === yn, Sn = xn && _n.process, z = function() {
			try {
				return bn && bn.require && bn.require("util").types || Sn && Sn.binding && Sn.binding("util");
			} catch {}
		}(), Cn = z && z.isArrayBuffer, wn = z && z.isDate, Tn = z && z.isMap, En = z && z.isRegExp, Dn = z && z.isSet, On = z && z.isTypedArray;
		function kn(e, t, n) {
			switch (n.length) {
				case 0: return e.call(t);
				case 1: return e.call(t, n[0]);
				case 2: return e.call(t, n[0], n[1]);
				case 3: return e.call(t, n[0], n[1], n[2]);
			}
			return e.apply(t, n);
		}
		function An(e, t, n, r) {
			for (var i = -1, a = e == null ? 0 : e.length; ++i < a;) {
				var o = e[i];
				t(r, o, n(o), e);
			}
			return r;
		}
		function jn(e, t) {
			for (var n = -1, r = e == null ? 0 : e.length; ++n < r && t(e[n], n, e) !== !1;);
			return e;
		}
		function Mn(e, t) {
			for (var n = e == null ? 0 : e.length; n-- && t(e[n], n, e) !== !1;);
			return e;
		}
		function Nn(e, t) {
			for (var n = -1, r = e == null ? 0 : e.length; ++n < r;) if (!t(e[n], n, e)) return !1;
			return !0;
		}
		function Pn(e, t) {
			for (var n = -1, r = e == null ? 0 : e.length, i = 0, a = []; ++n < r;) {
				var o = e[n];
				t(o, n, e) && (a[i++] = o);
			}
			return a;
		}
		function Fn(e, t) {
			return !!(e != null && e.length) && Kn(e, t, 0) > -1;
		}
		function In(e, t, n) {
			for (var r = -1, i = e == null ? 0 : e.length; ++r < i;) if (n(t, e[r])) return !0;
			return !1;
		}
		function B(e, t) {
			for (var n = -1, r = e == null ? 0 : e.length, i = Array(r); ++n < r;) i[n] = t(e[n], n, e);
			return i;
		}
		function Ln(e, t) {
			for (var n = -1, r = t.length, i = e.length; ++n < r;) e[i + n] = t[n];
			return e;
		}
		function Rn(e, t, n, r) {
			var i = -1, a = e == null ? 0 : e.length;
			for (r && a && (n = e[++i]); ++i < a;) n = t(n, e[i], i, e);
			return n;
		}
		function zn(e, t, n, r) {
			var i = e == null ? 0 : e.length;
			for (r && i && (n = e[--i]); i--;) n = t(n, e[i], i, e);
			return n;
		}
		function Bn(e, t) {
			for (var n = -1, r = e == null ? 0 : e.length; ++n < r;) if (t(e[n], n, e)) return !0;
			return !1;
		}
		var Vn = Xn("length");
		function Hn(e) {
			return e.split("");
		}
		function Un(e) {
			return e.match(ut) || [];
		}
		function Wn(e, t, n) {
			var r;
			return n(e, function(e, n, i) {
				if (t(e, n, i)) return r = n, !1;
			}), r;
		}
		function Gn(e, t, n, r) {
			for (var i = e.length, a = n + (r ? 1 : -1); r ? a-- : ++a < i;) if (t(e[a], a, e)) return a;
			return -1;
		}
		function Kn(e, t, n) {
			return t === t ? Sr(e, t, n) : Gn(e, Jn, n);
		}
		function qn(e, t, n, r) {
			for (var i = n - 1, a = e.length; ++i < a;) if (r(e[i], t)) return i;
			return -1;
		}
		function Jn(e) {
			return e !== e;
		}
		function Yn(e, t) {
			var n = e == null ? 0 : e.length;
			return n ? er(e, t) / n : me;
		}
		function Xn(e) {
			return function(t) {
				return t == null ? n : t[e];
			};
		}
		function Zn(e) {
			return function(t) {
				return e == null ? n : e[t];
			};
		}
		function Qn(e, t, n, r, i) {
			return i(e, function(e, i, a) {
				n = r ? (r = !1, e) : t(n, e, i, a);
			}), n;
		}
		function $n(e, t) {
			var n = e.length;
			for (e.sort(t); n--;) e[n] = e[n].value;
			return e;
		}
		function er(e, t) {
			for (var r, i = -1, a = e.length; ++i < a;) {
				var o = t(e[i]);
				o !== n && (r = r === n ? o : r + o);
			}
			return r;
		}
		function tr(e, t) {
			for (var n = -1, r = Array(e); ++n < e;) r[n] = t(n);
			return r;
		}
		function nr(e, t) {
			return B(t, function(t) {
				return [t, e[t]];
			});
		}
		function rr(e) {
			return e && e.slice(0, Er(e) + 1).replace(ot, "");
		}
		function ir(e) {
			return function(t) {
				return e(t);
			};
		}
		function ar(e, t) {
			return B(t, function(t) {
				return e[t];
			});
		}
		function or(e, t) {
			return e.has(t);
		}
		function sr(e, t) {
			for (var n = -1, r = e.length; ++n < r && Kn(t, e[n], 0) > -1;);
			return n;
		}
		function cr(e, t) {
			for (var n = e.length; n-- && Kn(t, e[n], 0) > -1;);
			return n;
		}
		function lr(e, t) {
			for (var n = e.length, r = 0; n--;) e[n] === t && ++r;
			return r;
		}
		var ur = Zn(L), dr = Zn(fn);
		function fr(e) {
			return "\\" + mn[e];
		}
		function pr(e, t) {
			return e == null ? n : e[t];
		}
		function mr(e) {
			return ln.test(e);
		}
		function hr(e) {
			return P.test(e);
		}
		function gr(e) {
			for (var t, n = []; !(t = e.next()).done;) n.push(t.value);
			return n;
		}
		function _r(e) {
			var t = -1, n = Array(e.size);
			return e.forEach(function(e, r) {
				n[++t] = [r, e];
			}), n;
		}
		function vr(e, t) {
			return function(n) {
				return e(t(n));
			};
		}
		function yr(e, t) {
			for (var n = -1, r = e.length, i = 0, a = []; ++n < r;) {
				var o = e[n];
				(o === t || o === d) && (e[n] = d, a[i++] = n);
			}
			return a;
		}
		function br(e) {
			var t = -1, n = Array(e.size);
			return e.forEach(function(e) {
				n[++t] = e;
			}), n;
		}
		function xr(e) {
			var t = -1, n = Array(e.size);
			return e.forEach(function(e) {
				n[++t] = [e, e];
			}), n;
		}
		function Sr(e, t, n) {
			for (var r = n - 1, i = e.length; ++r < i;) if (e[r] === t) return r;
			return -1;
		}
		function Cr(e, t, n) {
			for (var r = n + 1; r--;) if (e[r] === t) return r;
			return r;
		}
		function wr(e) {
			return mr(e) ? Or(e) : Vn(e);
		}
		function Tr(e) {
			return mr(e) ? kr(e) : Hn(e);
		}
		function Er(e) {
			for (var t = e.length; t-- && A.test(e.charAt(t)););
			return t;
		}
		var Dr = Zn(pn);
		function Or(e) {
			for (var t = sn.lastIndex = 0; sn.test(e);) ++t;
			return t;
		}
		function kr(e) {
			return e.match(sn) || [];
		}
		function Ar(e) {
			return e.match(cn) || [];
		}
		var jr = (function e(t) {
			t = t == null ? R : jr.defaults(R.Object(), t, jr.pick(R, un));
			var A = t.Array, ut = t.Date, Ct = t.Error, j = t.Function, wt = t.Math, M = t.Object, Tt = t.RegExp, Et = t.String, Dt = t.TypeError, Ot = A.prototype, kt = j.prototype, At = M.prototype, jt = t["__core-js_shared__"], Mt = kt.toString, N = At.hasOwnProperty, Nt = 0, Pt = function() {
				var e = /[^.]+$/.exec(jt && jt.keys && jt.keys.IE_PROTO || "");
				return e ? "Symbol(src)_1." + e : "";
			}(), Ft = At.toString, It = Mt.call(M), Lt = R._, Rt = Tt("^" + Mt.call(N).replace(it, "\\$&").replace(/hasOwnProperty|(function).*?(?=\\\()| for .+?(?=\\\])/g, "$1.*?") + "$"), zt = xn ? t.Buffer : n, Bt = t.Symbol, Vt = t.Uint8Array, Ht = zt ? zt.allocUnsafe : n, Ut = vr(M.getPrototypeOf, M), Wt = M.create, Gt = At.propertyIsEnumerable, Kt = Ot.splice, qt = Bt ? Bt.isConcatSpreadable : n, Jt = Bt ? Bt.iterator : n, Yt = Bt ? Bt.toStringTag : n, Xt = function() {
				try {
					var e = ns(M, "defineProperty");
					return e({}, "", {}), e;
				} catch {}
			}(), Zt = t.clearTimeout !== R.clearTimeout && t.clearTimeout, Qt = ut && ut.now !== R.Date.now && ut.now, $t = t.setTimeout !== R.setTimeout && t.setTimeout, en = wt.ceil, tn = wt.floor, nn = M.getOwnPropertySymbols, rn = zt ? zt.isBuffer : n, sn = t.isFinite, cn = Ot.join, ln = vr(M.keys, M), P = wt.max, L = wt.min, fn = ut.now, pn = t.parseInt, mn = wt.random, _n = Ot.reverse, vn = ns(t, "DataView"), yn = ns(t, "Map"), bn = ns(t, "Promise"), Sn = ns(t, "Set"), z = ns(t, "WeakMap"), Vn = ns(M, "create"), Hn = z && new z(), Zn = {}, Sr = Bs(vn), Or = Bs(yn), kr = Bs(bn), Mr = Bs(Sn), Nr = Bs(z), Pr = Bt ? Bt.prototype : n, Fr = Pr ? Pr.valueOf : n, Ir = Pr ? Pr.toString : n;
			function V(e) {
				if (X(e) && !q(e) && !(e instanceof H)) {
					if (e instanceof zr) return e;
					if (N.call(e, "__wrapped__")) return Hs(e);
				}
				return new zr(e);
			}
			var Lr = function() {
				function e() {}
				return function(t) {
					if (!Y(t)) return {};
					if (Wt) return Wt(t);
					e.prototype = t;
					var r = new e();
					return e.prototype = n, r;
				};
			}();
			function Rr() {}
			function zr(e, t) {
				this.__wrapped__ = e, this.__actions__ = [], this.__chain__ = !!t, this.__index__ = 0, this.__values__ = n;
			}
			V.templateSettings = {
				escape: D,
				evaluate: tt,
				interpolate: nt,
				variable: "",
				imports: { _: V }
			}, V.prototype = Rr.prototype, V.prototype.constructor = V, zr.prototype = Lr(Rr.prototype), zr.prototype.constructor = zr;
			function H(e) {
				this.__wrapped__ = e, this.__actions__ = [], this.__dir__ = 1, this.__filtered__ = !1, this.__iteratees__ = [], this.__takeCount__ = C, this.__views__ = [];
			}
			function Br() {
				var e = new H(this.__wrapped__);
				return e.__actions__ = mo(this.__actions__), e.__dir__ = this.__dir__, e.__filtered__ = this.__filtered__, e.__iteratees__ = mo(this.__iteratees__), e.__takeCount__ = this.__takeCount__, e.__views__ = mo(this.__views__), e;
			}
			function Vr() {
				if (this.__filtered__) {
					var e = new H(this);
					e.__dir__ = -1, e.__filtered__ = !0;
				} else e = this.clone(), e.__dir__ *= -1;
				return e;
			}
			function Hr() {
				var e = this.__wrapped__.value(), t = this.__dir__, n = q(e), r = t < 0, i = n ? e.length : 0, a = os(0, i, this.__views__), o = a.start, s = a.end, c = s - o, l = r ? s : o - 1, u = this.__iteratees__, d = u.length, f = 0, p = L(c, this.__takeCount__);
				if (!n || !r && i == c && p == c) return Ja(e, this.__actions__);
				var m = [];
				outer: for (; c-- && f < p;) {
					l += t;
					for (var h = -1, g = e[l]; ++h < d;) {
						var _ = u[h], v = _.iteratee, y = _.type, b = v(g);
						if (y == le) g = b;
						else if (!b) {
							if (y == ce) continue outer;
							break outer;
						}
					}
					m[f++] = g;
				}
				return m;
			}
			H.prototype = Lr(Rr.prototype), H.prototype.constructor = H;
			function Ur(e) {
				var t = -1, n = e == null ? 0 : e.length;
				for (this.clear(); ++t < n;) {
					var r = e[t];
					this.set(r[0], r[1]);
				}
			}
			function Wr() {
				this.__data__ = Vn ? Vn(null) : {}, this.size = 0;
			}
			function Gr(e) {
				var t = this.has(e) && delete this.__data__[e];
				return this.size -= +!!t, t;
			}
			function Kr(e) {
				var t = this.__data__;
				if (Vn) {
					var r = t[e];
					return r === l ? n : r;
				}
				return N.call(t, e) ? t[e] : n;
			}
			function qr(e) {
				var t = this.__data__;
				return Vn ? t[e] !== n : N.call(t, e);
			}
			function Jr(e, t) {
				var r = this.__data__;
				return this.size += +!this.has(e), r[e] = Vn && t === n ? l : t, this;
			}
			Ur.prototype.clear = Wr, Ur.prototype.delete = Gr, Ur.prototype.get = Kr, Ur.prototype.has = qr, Ur.prototype.set = Jr;
			function Yr(e) {
				var t = -1, n = e == null ? 0 : e.length;
				for (this.clear(); ++t < n;) {
					var r = e[t];
					this.set(r[0], r[1]);
				}
			}
			function Xr() {
				this.__data__ = [], this.size = 0;
			}
			function Zr(e) {
				var t = this.__data__, n = Si(t, e);
				return n < 0 ? !1 : (n == t.length - 1 ? t.pop() : Kt.call(t, n, 1), --this.size, !0);
			}
			function Qr(e) {
				var t = this.__data__, r = Si(t, e);
				return r < 0 ? n : t[r][1];
			}
			function $r(e) {
				return Si(this.__data__, e) > -1;
			}
			function ei(e, t) {
				var n = this.__data__, r = Si(n, e);
				return r < 0 ? (++this.size, n.push([e, t])) : n[r][1] = t, this;
			}
			Yr.prototype.clear = Xr, Yr.prototype.delete = Zr, Yr.prototype.get = Qr, Yr.prototype.has = $r, Yr.prototype.set = ei;
			function ti(e) {
				var t = -1, n = e == null ? 0 : e.length;
				for (this.clear(); ++t < n;) {
					var r = e[t];
					this.set(r[0], r[1]);
				}
			}
			function ni() {
				this.size = 0, this.__data__ = {
					hash: new Ur(),
					map: new (yn || Yr)(),
					string: new Ur()
				};
			}
			function ri(e) {
				var t = es(this, e).delete(e);
				return this.size -= +!!t, t;
			}
			function ii(e) {
				return es(this, e).get(e);
			}
			function ai(e) {
				return es(this, e).has(e);
			}
			function oi(e, t) {
				var n = es(this, e), r = n.size;
				return n.set(e, t), this.size += n.size == r ? 0 : 1, this;
			}
			ti.prototype.clear = ni, ti.prototype.delete = ri, ti.prototype.get = ii, ti.prototype.has = ai, ti.prototype.set = oi;
			function si(e) {
				var t = -1, n = e == null ? 0 : e.length;
				for (this.__data__ = new ti(); ++t < n;) this.add(e[t]);
			}
			function ci(e) {
				return this.__data__.set(e, l), this;
			}
			function li(e) {
				return this.__data__.has(e);
			}
			si.prototype.add = si.prototype.push = ci, si.prototype.has = li;
			function ui(e) {
				var t = this.__data__ = new Yr(e);
				this.size = t.size;
			}
			function di() {
				this.__data__ = new Yr(), this.size = 0;
			}
			function fi(e) {
				var t = this.__data__, n = t.delete(e);
				return this.size = t.size, n;
			}
			function pi(e) {
				return this.__data__.get(e);
			}
			function mi(e) {
				return this.__data__.has(e);
			}
			function hi(e, t) {
				var n = this.__data__;
				if (n instanceof Yr) {
					var r = n.__data__;
					if (!yn || r.length < i - 1) return r.push([e, t]), this.size = ++n.size, this;
					n = this.__data__ = new ti(r);
				}
				return n.set(e, t), this.size = n.size, this;
			}
			ui.prototype.clear = di, ui.prototype.delete = fi, ui.prototype.get = pi, ui.prototype.has = mi, ui.prototype.set = hi;
			function gi(e, t) {
				var n = q(e), r = !n && hu(e), i = !n && !r && yu(e), a = !n && !r && !i && Wu(e), o = n || r || i || a, s = o ? tr(e.length, Et) : [], c = s.length;
				for (var l in e) (t || N.call(e, l)) && !(o && (l == "length" || i && (l == "offset" || l == "parent") || a && (l == "buffer" || l == "byteLength" || l == "byteOffset") || ms(l, c))) && s.push(l);
				return s;
			}
			function _i(e) {
				var t = e.length;
				return t ? e[Oa(0, t - 1)] : n;
			}
			function vi(e, t) {
				return Ls(mo(e), Oi(t, 0, e.length));
			}
			function yi(e) {
				return Ls(mo(e));
			}
			function bi(e, t, r) {
				(r !== n && !fu(e[t], r) || r === n && !(t in e)) && Ei(e, t, r);
			}
			function xi(e, t, r) {
				var i = e[t];
				(!(N.call(e, t) && fu(i, r)) || r === n && !(t in e)) && Ei(e, t, r);
			}
			function Si(e, t) {
				for (var n = e.length; n--;) if (fu(e[n][0], t)) return n;
				return -1;
			}
			function Ci(e, t, n, r) {
				return Pi(e, function(e, i, a) {
					t(r, e, n(e), a);
				}), r;
			}
			function wi(e, t) {
				return e && ho(t, $(t), e);
			}
			function Ti(e, t) {
				return e && ho(t, wd(t), e);
			}
			function Ei(e, t, n) {
				t == "__proto__" && Xt ? Xt(e, t, {
					configurable: !0,
					enumerable: !0,
					value: n,
					writable: !0
				}) : e[t] = n;
			}
			function Di(e, t) {
				for (var r = -1, i = t.length, a = A(i), o = e == null; ++r < i;) a[r] = o ? n : vd(e, t[r]);
				return a;
			}
			function Oi(e, t, r) {
				return e === e && (r !== n && (e = e <= r ? e : r), t !== n && (e = e >= t ? e : t)), e;
			}
			function ki(e, t, r, i, a, o) {
				var s, c = t & f, l = t & p, u = t & m;
				if (r && (s = a ? r(e, i, a, o) : r(e)), s !== n) return s;
				if (!Y(e)) return e;
				var d = q(e);
				if (d) {
					if (s = ls(e), !c) return mo(e, s);
				} else {
					var h = K(e), g = h == Te || h == Ee;
					if (yu(e)) return ro(e, c);
					if (h == Ae || h == ve || g && !a) {
						if (s = l || g ? {} : us(e), !c) return l ? _o(e, Ti(s, e)) : go(e, wi(s, e));
					} else {
						if (!I[h]) return a ? e : {};
						s = ds(e, h, c);
					}
				}
				o ||= new ui();
				var _ = o.get(e);
				if (_) return _;
				o.set(e, s), Vu(e) ? e.forEach(function(n) {
					s.add(ki(n, t, r, n, e, o));
				}) : Au(e) && e.forEach(function(n, i) {
					s.set(i, ki(n, t, r, i, e, o));
				});
				var v = d ? n : (u ? l ? Xo : Yo : l ? wd : $)(e);
				return jn(v || e, function(n, i) {
					v && (i = n, n = e[i]), xi(s, i, ki(n, t, r, i, e, o));
				}), s;
			}
			function Ai(e) {
				var t = $(e);
				return function(n) {
					return ji(n, e, t);
				};
			}
			function ji(e, t, r) {
				var i = r.length;
				if (e == null) return !i;
				for (e = M(e); i--;) {
					var a = r[i], o = t[a], s = e[a];
					if (s === n && !(a in e) || !o(s)) return !1;
				}
				return !0;
			}
			function Mi(e, t, r) {
				if (typeof e != "function") throw new Dt(o);
				return Ns(function() {
					e.apply(n, r);
				}, t);
			}
			function Ni(e, t, n, r) {
				var a = -1, o = Fn, s = !0, c = e.length, l = [], u = t.length;
				if (!c) return l;
				n && (t = B(t, ir(n))), r ? (o = In, s = !1) : t.length >= i && (o = or, s = !1, t = new si(t));
				outer: for (; ++a < c;) {
					var d = e[a], f = n == null ? d : n(d);
					if (d = r || d !== 0 ? d : 0, s && f === f) {
						for (var p = u; p--;) if (t[p] === f) continue outer;
						l.push(d);
					} else o(t, f, r) || l.push(d);
				}
				return l;
			}
			var Pi = bo(Hi), Fi = bo(Ui, !0);
			function Ii(e, t) {
				var n = !0;
				return Pi(e, function(e, r, i) {
					return n = !!t(e, r, i), n;
				}), n;
			}
			function Li(e, t, r) {
				for (var i = -1, a = e.length; ++i < a;) {
					var o = e[i], s = t(o);
					if (s != null && (c === n ? s === s && !Uu(s) : r(s, c))) var c = s, l = o;
				}
				return l;
			}
			function Ri(e, t, r, i) {
				var a = e.length;
				for (r = Z(r), r < 0 && (r = -r > a ? 0 : a + r), i = i === n || i > a ? a : Z(i), i < 0 && (i += a), i = r > i ? 0 : Qu(i); r < i;) e[r++] = t;
				return e;
			}
			function zi(e, t) {
				var n = [];
				return Pi(e, function(e, r, i) {
					t(e, r, i) && n.push(e);
				}), n;
			}
			function U(e, t, n, r, i) {
				var a = -1, o = e.length;
				for (n ||= ps, i ||= []; ++a < o;) {
					var s = e[a];
					t > 0 && n(s) ? t > 1 ? U(s, t - 1, n, r, i) : Ln(i, s) : r || (i[i.length] = s);
				}
				return i;
			}
			var Bi = xo(), Vi = xo(!0);
			function Hi(e, t) {
				return e && Bi(e, t, $);
			}
			function Ui(e, t) {
				return e && Vi(e, t, $);
			}
			function Wi(e, t) {
				return Pn(t, function(t) {
					return Du(e[t]);
				});
			}
			function Gi(e, t) {
				t = $a(t, e);
				for (var r = 0, i = t.length; e != null && r < i;) e = e[zs(t[r++])];
				return r && r == i ? e : n;
			}
			function Ki(e, t, n) {
				var r = t(e);
				return q(e) ? r : Ln(r, n(e));
			}
			function qi(e) {
				return e == null ? e === n ? Le : ke : Yt && Yt in M(e) ? rs(e) : Ds(e);
			}
			function Ji(e, t) {
				return e > t;
			}
			function Yi(e, t) {
				return e != null && N.call(e, t);
			}
			function Xi(e, t) {
				return e != null && t in M(e);
			}
			function Zi(e, t, n) {
				return e >= L(t, n) && e < P(t, n);
			}
			function Qi(e, t, r) {
				for (var i = r ? In : Fn, a = e[0].length, o = e.length, s = o, c = A(o), l = Infinity, u = []; s--;) {
					var d = e[s];
					s && t && (d = B(d, ir(t))), l = L(d.length, l), c[s] = !r && (t || a >= 120 && d.length >= 120) ? new si(s && d) : n;
				}
				d = e[0];
				var f = -1, p = c[0];
				outer: for (; ++f < a && u.length < l;) {
					var m = d[f], h = t ? t(m) : m;
					if (m = r || m !== 0 ? m : 0, !(p ? or(p, h) : i(u, h, r))) {
						for (s = o; --s;) {
							var g = c[s];
							if (!(g ? or(g, h) : i(e[s], h, r))) continue outer;
						}
						p && p.push(h), u.push(m);
					}
				}
				return u;
			}
			function $i(e, t, n, r) {
				return Hi(e, function(e, i, a) {
					t(r, n(e), i, a);
				}), r;
			}
			function ea(e, t, r) {
				t = $a(t, e), e = ks(e, t);
				var i = e == null ? e : e[zs(pc(t))];
				return i == null ? n : kn(i, e, r);
			}
			function ta(e) {
				return X(e) && qi(e) == ve;
			}
			function na(e) {
				return X(e) && qi(e) == ze;
			}
			function ra(e) {
				return X(e) && qi(e) == Se;
			}
			function ia(e, t, n, r, i) {
				return e === t ? !0 : e == null || t == null || !X(e) && !X(t) ? e !== e && t !== t : aa(e, t, n, r, ia, i);
			}
			function aa(e, t, n, r, i, a) {
				var o = q(e), s = q(t), c = o ? ye : K(e), l = s ? ye : K(t);
				c = c == ve ? Ae : c, l = l == ve ? Ae : l;
				var u = c == Ae, d = l == Ae, f = c == l;
				if (f && yu(e)) {
					if (!yu(t)) return !1;
					o = !0, u = !1;
				}
				if (f && !u) return a ||= new ui(), o || Wu(e) ? Go(e, t, n, r, i, a) : Ko(e, t, c, n, r, i, a);
				if (!(n & h)) {
					var p = u && N.call(e, "__wrapped__"), m = d && N.call(t, "__wrapped__");
					if (p || m) {
						var g = p ? e.value() : e, _ = m ? t.value() : t;
						return a ||= new ui(), i(g, _, n, r, a);
					}
				}
				return f ? (a ||= new ui(), qo(e, t, n, r, i, a)) : !1;
			}
			function oa(e) {
				return X(e) && K(e) == De;
			}
			function sa(e, t, r, i) {
				var a = r.length, o = a, s = !i;
				if (e == null) return !o;
				for (e = M(e); a--;) {
					var c = r[a];
					if (s && c[2] ? c[1] !== e[c[0]] : !(c[0] in e)) return !1;
				}
				for (; ++a < o;) {
					c = r[a];
					var l = c[0], u = e[l], d = c[1];
					if (s && c[2]) {
						if (u === n && !(l in e)) return !1;
					} else {
						var f = new ui();
						if (i) var p = i(u, d, l, e, t, f);
						if (!(p === n ? ia(d, u, h | g, i, f) : p)) return !1;
					}
				}
				return !0;
			}
			function ca(e) {
				return !Y(e) || ys(e) ? !1 : (Du(e) ? Rt : _t).test(Bs(e));
			}
			function la(e) {
				return X(e) && qi(e) == Ne;
			}
			function ua(e) {
				return X(e) && K(e) == Pe;
			}
			function da(e) {
				return X(e) && ku(e.length) && !!F[qi(e)];
			}
			function fa(e) {
				return typeof e == "function" ? e : e == null ? Mf : typeof e == "object" ? q(e) ? va(e[0], e[1]) : _a(e) : Gf(e);
			}
			function pa(e) {
				if (!xs(e)) return ln(e);
				var t = [];
				for (var n in M(e)) N.call(e, n) && n != "constructor" && t.push(n);
				return t;
			}
			function ma(e) {
				if (!Y(e)) return Es(e);
				var t = xs(e), n = [];
				for (var r in e) r == "constructor" && (t || !N.call(e, r)) || n.push(r);
				return n;
			}
			function ha(e, t) {
				return e < t;
			}
			function ga(e, t) {
				var n = -1, r = _u(e) ? A(e.length) : [];
				return Pi(e, function(e, i, a) {
					r[++n] = t(e, i, a);
				}), r;
			}
			function _a(e) {
				var t = ts(e);
				return t.length == 1 && t[0][2] ? Cs(t[0][0], t[0][1]) : function(n) {
					return n === e || sa(n, e, t);
				};
			}
			function va(e, t) {
				return gs(e) && Ss(t) ? Cs(zs(e), t) : function(r) {
					var i = vd(r, e);
					return i === n && i === t ? bd(r, e) : ia(t, i, h | g);
				};
			}
			function ya(e, t, r, i, a) {
				e !== t && Bi(t, function(o, s) {
					if (a ||= new ui(), Y(o)) ba(e, t, s, r, ya, i, a);
					else {
						var c = i ? i(js(e, s), o, s + "", e, t, a) : n;
						c === n && (c = o), bi(e, s, c);
					}
				}, wd);
			}
			function ba(e, t, r, i, a, o, s) {
				var c = js(e, r), l = js(t, r), u = s.get(l);
				if (u) {
					bi(e, r, u);
					return;
				}
				var d = o ? o(c, l, r + "", e, t, s) : n, f = d === n;
				if (f) {
					var p = q(l), m = !p && yu(l), h = !p && !m && Wu(l);
					d = l, p || m || h ? q(c) ? d = c : J(c) ? d = mo(c) : m ? (f = !1, d = ro(l, !0)) : h ? (f = !1, d = co(l, !0)) : d = [] : Ru(l) || hu(l) ? (d = c, hu(c) ? d = ed(c) : (!Y(c) || Du(c)) && (d = us(l))) : f = !1;
				}
				f && (s.set(l, d), a(d, l, i, o, s), s.delete(l)), bi(e, r, d);
			}
			function xa(e, t) {
				var r = e.length;
				if (r) return t += t < 0 ? r : 0, ms(t, r) ? e[t] : n;
			}
			function Sa(e, t, n) {
				t = t.length ? B(t, function(e) {
					return q(e) ? function(t) {
						return Gi(t, e.length === 1 ? e[0] : e);
					} : e;
				}) : [Mf];
				var r = -1;
				return t = B(t, ir(G())), $n(ga(e, function(e, n, i) {
					return {
						criteria: B(t, function(t) {
							return t(e);
						}),
						index: ++r,
						value: e
					};
				}), function(e, t) {
					return uo(e, t, n);
				});
			}
			function Ca(e, t) {
				return wa(e, t, function(t, n) {
					return bd(e, n);
				});
			}
			function wa(e, t, n) {
				for (var r = -1, i = t.length, a = {}; ++r < i;) {
					var o = t[r], s = Gi(e, o);
					n(s, o) && Na(a, $a(o, e), s);
				}
				return a;
			}
			function Ta(e) {
				return function(t) {
					return Gi(t, e);
				};
			}
			function Ea(e, t, n, r) {
				var i = r ? qn : Kn, a = -1, o = t.length, s = e;
				for (e === t && (t = mo(t)), n && (s = B(e, ir(n))); ++a < o;) for (var c = 0, l = t[a], u = n ? n(l) : l; (c = i(s, u, c, r)) > -1;) s !== e && Kt.call(s, c, 1), Kt.call(e, c, 1);
				return e;
			}
			function Da(e, t) {
				for (var n = e ? t.length : 0, r = n - 1; n--;) {
					var i = t[n];
					if (n == r || i !== a) {
						var a = i;
						ms(i) ? Kt.call(e, i, 1) : Ga(e, i);
					}
				}
				return e;
			}
			function Oa(e, t) {
				return e + tn(mn() * (t - e + 1));
			}
			function ka(e, t, n, r) {
				for (var i = -1, a = P(en((t - e) / (n || 1)), 0), o = A(a); a--;) o[r ? a : ++i] = e, e += n;
				return o;
			}
			function Aa(e, t) {
				var n = "";
				if (!e || t < 1 || t > fe) return n;
				do
					t % 2 && (n += e), t = tn(t / 2), t && (e += e);
				while (t);
				return n;
			}
			function W(e, t) {
				return Ps(Os(e, t, Mf), e + "");
			}
			function ja(e) {
				return _i(Hd(e));
			}
			function Ma(e, t) {
				var n = Hd(e);
				return Ls(n, Oi(t, 0, n.length));
			}
			function Na(e, t, r, i) {
				if (!Y(e)) return e;
				t = $a(t, e);
				for (var a = -1, o = t.length, s = o - 1, c = e; c != null && ++a < o;) {
					var l = zs(t[a]), u = r;
					if (l === "__proto__" || l === "constructor" || l === "prototype") return e;
					if (a != s) {
						var d = c[l];
						u = i ? i(d, l, c) : n, u === n && (u = Y(d) ? d : ms(t[a + 1]) ? [] : {});
					}
					xi(c, l, u), c = c[l];
				}
				return e;
			}
			var Pa = Hn ? function(e, t) {
				return Hn.set(e, t), e;
			} : Mf, Fa = Xt ? function(e, t) {
				return Xt(e, "toString", {
					configurable: !0,
					enumerable: !1,
					value: Of(t),
					writable: !0
				});
			} : Mf;
			function Ia(e) {
				return Ls(Hd(e));
			}
			function La(e, t, n) {
				var r = -1, i = e.length;
				t < 0 && (t = -t > i ? 0 : i + t), n = n > i ? i : n, n < 0 && (n += i), i = t > n ? 0 : n - t >>> 0, t >>>= 0;
				for (var a = A(i); ++r < i;) a[r] = e[r + t];
				return a;
			}
			function Ra(e, t) {
				var n;
				return Pi(e, function(e, r, i) {
					return n = t(e, r, i), !n;
				}), !!n;
			}
			function za(e, t, n) {
				var r = 0, i = e == null ? r : e.length;
				if (typeof t == "number" && t === t && i <= ge) {
					for (; r < i;) {
						var a = r + i >>> 1, o = e[a];
						o !== null && !Uu(o) && (n ? o <= t : o < t) ? r = a + 1 : i = a;
					}
					return i;
				}
				return Ba(e, t, Mf, n);
			}
			function Ba(e, t, r, i) {
				var a = 0, o = e == null ? 0 : e.length;
				if (o === 0) return 0;
				t = r(t);
				for (var s = t !== t, c = t === null, l = Uu(t), u = t === n; a < o;) {
					var d = tn((a + o) / 2), f = r(e[d]), p = f !== n, m = f === null, h = f === f, g = Uu(f);
					if (s) var _ = i || h;
					else _ = u ? h && (i || p) : c ? h && p && (i || !m) : l ? h && p && !m && (i || !g) : m || g ? !1 : i ? f <= t : f < t;
					_ ? a = d + 1 : o = d;
				}
				return L(o, he);
			}
			function Va(e, t) {
				for (var n = -1, r = e.length, i = 0, a = []; ++n < r;) {
					var o = e[n], s = t ? t(o) : o;
					if (!n || !fu(s, c)) {
						var c = s;
						a[i++] = o === 0 ? 0 : o;
					}
				}
				return a;
			}
			function Ha(e) {
				return typeof e == "number" ? e : Uu(e) ? me : +e;
			}
			function Ua(e) {
				if (typeof e == "string") return e;
				if (q(e)) return B(e, Ua) + "";
				if (Uu(e)) return Ir ? Ir.call(e) : "";
				var t = e + "";
				return t == "0" && 1 / e == -de ? "-0" : t;
			}
			function Wa(e, t, n) {
				var r = -1, a = Fn, o = e.length, s = !0, c = [], l = c;
				if (n) s = !1, a = In;
				else if (o >= i) {
					var u = t ? null : zo(e);
					if (u) return br(u);
					s = !1, a = or, l = new si();
				} else l = t ? [] : c;
				outer: for (; ++r < o;) {
					var d = e[r], f = t ? t(d) : d;
					if (d = n || d !== 0 ? d : 0, s && f === f) {
						for (var p = l.length; p--;) if (l[p] === f) continue outer;
						t && l.push(f), c.push(d);
					} else a(l, f, n) || (l !== c && l.push(f), c.push(d));
				}
				return c;
			}
			function Ga(e, t) {
				t = $a(t, e);
				var n = -1, r = t.length;
				if (!r) return !0;
				for (; ++n < r;) {
					var i = zs(t[n]);
					if (i === "__proto__" && !N.call(e, "__proto__") || (i === "constructor" || i === "prototype") && n < r - 1) return !1;
				}
				var a = ks(e, t);
				return a == null || delete a[zs(pc(t))];
			}
			function Ka(e, t, n, r) {
				return Na(e, t, n(Gi(e, t)), r);
			}
			function qa(e, t, n, r) {
				for (var i = e.length, a = r ? i : -1; (r ? a-- : ++a < i) && t(e[a], a, e););
				return n ? La(e, r ? 0 : a, r ? a + 1 : i) : La(e, r ? a + 1 : 0, r ? i : a);
			}
			function Ja(e, t) {
				var n = e;
				return n instanceof H && (n = n.value()), Rn(t, function(e, t) {
					return t.func.apply(t.thisArg, Ln([e], t.args));
				}, n);
			}
			function Ya(e, t, n) {
				var r = e.length;
				if (r < 2) return r ? Wa(e[0]) : [];
				for (var i = -1, a = A(r); ++i < r;) for (var o = e[i], s = -1; ++s < r;) s != i && (a[i] = Ni(a[i] || o, e[s], t, n));
				return Wa(U(a, 1), t, n);
			}
			function Xa(e, t, r) {
				for (var i = -1, a = e.length, o = t.length, s = {}; ++i < a;) {
					var c = i < o ? t[i] : n;
					r(s, e[i], c);
				}
				return s;
			}
			function Za(e) {
				return J(e) ? e : [];
			}
			function Qa(e) {
				return typeof e == "function" ? e : Mf;
			}
			function $a(e, t) {
				return q(e) ? e : gs(e, t) ? [e] : Rs(Q(e));
			}
			var eo = W;
			function to(e, t, r) {
				var i = e.length;
				return r = r === n ? i : r, !t && r >= i ? e : La(e, t, r);
			}
			var no = Zt || function(e) {
				return R.clearTimeout(e);
			};
			function ro(e, t) {
				if (t) return e.slice();
				var n = e.length, r = Ht ? Ht(n) : new e.constructor(n);
				return e.copy(r), r;
			}
			function io(e) {
				var t = new e.constructor(e.byteLength);
				return new Vt(t).set(new Vt(e)), t;
			}
			function ao(e, t) {
				var n = t ? io(e.buffer) : e.buffer;
				return new e.constructor(n, e.byteOffset, e.byteLength);
			}
			function oo(e) {
				var t = new e.constructor(e.source, mt.exec(e));
				return t.lastIndex = e.lastIndex, t;
			}
			function so(e) {
				return Fr ? M(Fr.call(e)) : {};
			}
			function co(e, t) {
				var n = t ? io(e.buffer) : e.buffer;
				return new e.constructor(n, e.byteOffset, e.length);
			}
			function lo(e, t) {
				if (e !== t) {
					var r = e !== n, i = e === null, a = e === e, o = Uu(e), s = t !== n, c = t === null, l = t === t, u = Uu(t);
					if (!c && !u && !o && e > t || o && s && l && !c && !u || i && s && l || !r && l || !a) return 1;
					if (!i && !o && !u && e < t || u && r && a && !i && !o || c && r && a || !s && a || !l) return -1;
				}
				return 0;
			}
			function uo(e, t, n) {
				for (var r = -1, i = e.criteria, a = t.criteria, o = i.length, s = n.length; ++r < o;) {
					var c = lo(i[r], a[r]);
					if (c) return r >= s ? c : c * (n[r] == "desc" ? -1 : 1);
				}
				return e.index - t.index;
			}
			function fo(e, t, n, r) {
				for (var i = -1, a = e.length, o = n.length, s = -1, c = t.length, l = P(a - o, 0), u = A(c + l), d = !r; ++s < c;) u[s] = t[s];
				for (; ++i < o;) (d || i < a) && (u[n[i]] = e[i]);
				for (; l--;) u[s++] = e[i++];
				return u;
			}
			function po(e, t, n, r) {
				for (var i = -1, a = e.length, o = -1, s = n.length, c = -1, l = t.length, u = P(a - s, 0), d = A(u + l), f = !r; ++i < u;) d[i] = e[i];
				for (var p = i; ++c < l;) d[p + c] = t[c];
				for (; ++o < s;) (f || i < a) && (d[p + n[o]] = e[i++]);
				return d;
			}
			function mo(e, t) {
				var n = -1, r = e.length;
				for (t ||= A(r); ++n < r;) t[n] = e[n];
				return t;
			}
			function ho(e, t, r, i) {
				var a = !r;
				r ||= {};
				for (var o = -1, s = t.length; ++o < s;) {
					var c = t[o], l = i ? i(r[c], e[c], c, r, e) : n;
					l === n && (l = e[c]), a ? Ei(r, c, l) : xi(r, c, l);
				}
				return r;
			}
			function go(e, t) {
				return ho(e, is(e), t);
			}
			function _o(e, t) {
				return ho(e, as(e), t);
			}
			function vo(e, t) {
				return function(n, r) {
					var i = q(n) ? An : Ci, a = t ? t() : {};
					return i(n, e, G(r, 2), a);
				};
			}
			function yo(e) {
				return W(function(t, r) {
					var i = -1, a = r.length, o = a > 1 ? r[a - 1] : n, s = a > 2 ? r[2] : n;
					for (o = e.length > 3 && typeof o == "function" ? (a--, o) : n, s && hs(r[0], r[1], s) && (o = a < 3 ? n : o, a = 1), t = M(t); ++i < a;) {
						var c = r[i];
						c && e(t, c, i, o);
					}
					return t;
				});
			}
			function bo(e, t) {
				return function(n, r) {
					if (n == null) return n;
					if (!_u(n)) return e(n, r);
					for (var i = n.length, a = t ? i : -1, o = M(n); (t ? a-- : ++a < i) && r(o[a], a, o) !== !1;);
					return n;
				};
			}
			function xo(e) {
				return function(t, n, r) {
					for (var i = -1, a = M(t), o = r(t), s = o.length; s--;) {
						var c = o[e ? s : ++i];
						if (n(a[c], c, a) === !1) break;
					}
					return t;
				};
			}
			function So(e, t, n) {
				var r = t & _, i = To(e);
				function a() {
					return (this && this !== R && this instanceof a ? i : e).apply(r ? n : this, arguments);
				}
				return a;
			}
			function Co(e) {
				return function(t) {
					t = Q(t);
					var r = mr(t) ? Tr(t) : n, i = r ? r[0] : t.charAt(0), a = r ? to(r, 1).join("") : t.slice(1);
					return i[e]() + a;
				};
			}
			function wo(e) {
				return function(t) {
					return Rn(Cf(Yd(t).replace(an, "")), e, "");
				};
			}
			function To(e) {
				return function() {
					var t = arguments;
					switch (t.length) {
						case 0: return new e();
						case 1: return new e(t[0]);
						case 2: return new e(t[0], t[1]);
						case 3: return new e(t[0], t[1], t[2]);
						case 4: return new e(t[0], t[1], t[2], t[3]);
						case 5: return new e(t[0], t[1], t[2], t[3], t[4]);
						case 6: return new e(t[0], t[1], t[2], t[3], t[4], t[5]);
						case 7: return new e(t[0], t[1], t[2], t[3], t[4], t[5], t[6]);
					}
					var n = Lr(e.prototype), r = e.apply(n, t);
					return Y(r) ? r : n;
				};
			}
			function Eo(e, t, r) {
				var i = To(e);
				function a() {
					for (var o = arguments.length, s = A(o), c = o, l = $o(a); c--;) s[c] = arguments[c];
					var u = o < 3 && s[0] !== l && s[o - 1] !== l ? [] : yr(s, l);
					return o -= u.length, o < r ? Lo(e, t, ko, a.placeholder, n, s, u, n, n, r - o) : kn(this && this !== R && this instanceof a ? i : e, this, s);
				}
				return a;
			}
			function Do(e) {
				return function(t, r, i) {
					var a = M(t);
					if (!_u(t)) {
						var o = G(r, 3);
						t = $(t), r = function(e) {
							return o(a[e], e, a);
						};
					}
					var s = e(t, r, i);
					return s > -1 ? a[o ? t[s] : s] : n;
				};
			}
			function Oo(e) {
				return Jo(function(t) {
					var r = t.length, i = r, a = zr.prototype.thru;
					for (e && t.reverse(); i--;) {
						var s = t[i];
						if (typeof s != "function") throw new Dt(o);
						if (a && !c && Qo(s) == "wrapper") var c = new zr([], !0);
					}
					for (i = c ? i : r; ++i < r;) {
						s = t[i];
						var l = Qo(s), u = l == "wrapper" ? Zo(s) : n;
						c = u && vs(u[0]) && u[1] == (ne | b | ee | re) && !u[4].length && u[9] == 1 ? c[Qo(u[0])].apply(c, u[3]) : s.length == 1 && vs(s) ? c[l]() : c.thru(s);
					}
					return function() {
						var e = arguments, n = e[0];
						if (c && e.length == 1 && q(n)) return c.plant(n).value();
						for (var i = 0, a = r ? t[i].apply(this, e) : n; ++i < r;) a = t[i].call(this, a);
						return a;
					};
				});
			}
			function ko(e, t, r, i, a, o, s, c, l, u) {
				var d = t & ne, f = t & _, p = t & v, m = t & (b | x), h = t & ie, g = p ? n : To(e);
				function y() {
					for (var n = arguments.length, _ = A(n), v = n; v--;) _[v] = arguments[v];
					if (m) var b = $o(y), x = lr(_, b);
					if (i && (_ = fo(_, i, a, m)), o && (_ = po(_, o, s, m)), n -= x, m && n < u) {
						var ee = yr(_, b);
						return Lo(e, t, ko, y.placeholder, r, _, ee, c, l, u - n);
					}
					var te = f ? r : this, ne = p ? te[e] : e;
					return n = _.length, c ? _ = As(_, c) : h && n > 1 && _.reverse(), d && l < n && (_.length = l), this && this !== R && this instanceof y && (ne = g || To(ne)), ne.apply(te, _);
				}
				return y;
			}
			function Ao(e, t) {
				return function(n, r) {
					return $i(n, e, t(r), {});
				};
			}
			function jo(e, t) {
				return function(r, i) {
					var a;
					if (r === n && i === n) return t;
					if (r !== n && (a = r), i !== n) {
						if (a === n) return i;
						typeof r == "string" || typeof i == "string" ? (r = Ua(r), i = Ua(i)) : (r = Ha(r), i = Ha(i)), a = e(r, i);
					}
					return a;
				};
			}
			function Mo(e) {
				return Jo(function(t) {
					return t = B(t, ir(G())), W(function(n) {
						var r = this;
						return e(t, function(e) {
							return kn(e, r, n);
						});
					});
				});
			}
			function No(e, t) {
				t = t === n ? " " : Ua(t);
				var r = t.length;
				if (r < 2) return r ? Aa(t, e) : t;
				var i = Aa(t, en(e / wr(t)));
				return mr(t) ? to(Tr(i), 0, e).join("") : i.slice(0, e);
			}
			function Po(e, t, n, r) {
				var i = t & _, a = To(e);
				function o() {
					for (var t = -1, s = arguments.length, c = -1, l = r.length, u = A(l + s), d = this && this !== R && this instanceof o ? a : e; ++c < l;) u[c] = r[c];
					for (; s--;) u[c++] = arguments[++t];
					return kn(d, i ? n : this, u);
				}
				return o;
			}
			function Fo(e) {
				return function(t, r, i) {
					return i && typeof i != "number" && hs(t, r, i) && (r = i = n), t = Zu(t), r === n ? (r = t, t = 0) : r = Zu(r), i = i === n ? t < r ? 1 : -1 : Zu(i), ka(t, r, i, e);
				};
			}
			function Io(e) {
				return function(t, n) {
					return typeof t == "string" && typeof n == "string" || (t = $u(t), n = $u(n)), e(t, n);
				};
			}
			function Lo(e, t, r, i, a, o, s, c, l, u) {
				var d = t & b, f = d ? s : n, p = d ? n : s, m = d ? o : n, h = d ? n : o;
				t |= d ? ee : te, t &= ~(d ? te : ee), t & y || (t &= ~(_ | v));
				var g = [
					e,
					t,
					a,
					m,
					f,
					h,
					p,
					c,
					l,
					u
				], x = r.apply(n, g);
				return vs(e) && Ms(x, g), x.placeholder = i, Fs(x, e, t);
			}
			function Ro(e) {
				var t = wt[e];
				return function(e, n) {
					if (e = $u(e), n = n == null ? 0 : L(Z(n), 292), n && sn(e)) {
						var r = (Q(e) + "e").split("e");
						return r = (Q(t(r[0] + "e" + (+r[1] + n))) + "e").split("e"), +(r[0] + "e" + (+r[1] - n));
					}
					return t(e);
				};
			}
			var zo = Sn && 1 / br(new Sn([, -0]))[1] == de ? function(e) {
				return new Sn(e);
			} : Bf;
			function Bo(e) {
				return function(t) {
					var n = K(t);
					return n == De ? _r(t) : n == Pe ? xr(t) : nr(t, e(t));
				};
			}
			function Vo(e, t, r, i, a, s, c, l) {
				var u = t & v;
				if (!u && typeof e != "function") throw new Dt(o);
				var d = i ? i.length : 0;
				if (d || (t &= ~(ee | te), i = a = n), c = c === n ? c : P(Z(c), 0), l = l === n ? l : Z(l), d -= a ? a.length : 0, t & te) {
					var f = i, p = a;
					i = a = n;
				}
				var m = u ? n : Zo(e), h = [
					e,
					t,
					r,
					i,
					a,
					f,
					p,
					s,
					c,
					l
				];
				if (m && Ts(h, m), e = h[0], t = h[1], r = h[2], i = h[3], a = h[4], l = h[9] = h[9] === n ? u ? 0 : e.length : P(h[9] - d, 0), !l && t & (b | x) && (t &= ~(b | x)), !t || t == _) var g = So(e, t, r);
				else g = t == b || t == x ? Eo(e, t, l) : (t == ee || t == (_ | ee)) && !a.length ? Po(e, t, r, i) : ko.apply(n, h);
				return Fs((m ? Pa : Ms)(g, h), e, t);
			}
			function Ho(e, t, r, i) {
				return e === n || fu(e, At[r]) && !N.call(i, r) ? t : e;
			}
			function Uo(e, t, r, i, a, o) {
				return Y(e) && Y(t) && (o.set(t, e), ya(e, t, n, Uo, o), o.delete(t)), e;
			}
			function Wo(e) {
				return Ru(e) ? n : e;
			}
			function Go(e, t, r, i, a, o) {
				var s = r & h, c = e.length, l = t.length;
				if (c != l && !(s && l > c)) return !1;
				var u = o.get(e), d = o.get(t);
				if (u && d) return u == t && d == e;
				var f = -1, p = !0, m = r & g ? new si() : n;
				for (o.set(e, t), o.set(t, e); ++f < c;) {
					var _ = e[f], v = t[f];
					if (i) var y = s ? i(v, _, f, t, e, o) : i(_, v, f, e, t, o);
					if (y !== n) {
						if (y) continue;
						p = !1;
						break;
					}
					if (m) {
						if (!Bn(t, function(e, t) {
							if (!or(m, t) && (_ === e || a(_, e, r, i, o))) return m.push(t);
						})) {
							p = !1;
							break;
						}
					} else if (!(_ === v || a(_, v, r, i, o))) {
						p = !1;
						break;
					}
				}
				return o.delete(e), o.delete(t), p;
			}
			function Ko(e, t, n, r, i, a, o) {
				switch (n) {
					case Be:
						if (e.byteLength != t.byteLength || e.byteOffset != t.byteOffset) return !1;
						e = e.buffer, t = t.buffer;
					case ze: return !(e.byteLength != t.byteLength || !a(new Vt(e), new Vt(t)));
					case xe:
					case Se:
					case Oe: return fu(+e, +t);
					case we: return e.name == t.name && e.message == t.message;
					case Ne:
					case Fe: return e == t + "";
					case De: var s = _r;
					case Pe:
						var c = r & h;
						if (s ||= br, e.size != t.size && !c) return !1;
						var l = o.get(e);
						if (l) return l == t;
						r |= g, o.set(e, t);
						var u = Go(s(e), s(t), r, i, a, o);
						return o.delete(e), u;
					case Ie: if (Fr) return Fr.call(e) == Fr.call(t);
				}
				return !1;
			}
			function qo(e, t, r, i, a, o) {
				var s = r & h, c = Yo(e), l = c.length;
				if (l != Yo(t).length && !s) return !1;
				for (var u = l; u--;) {
					var d = c[u];
					if (!(s ? d in t : N.call(t, d))) return !1;
				}
				var f = o.get(e), p = o.get(t);
				if (f && p) return f == t && p == e;
				var m = !0;
				o.set(e, t), o.set(t, e);
				for (var g = s; ++u < l;) {
					d = c[u];
					var _ = e[d], v = t[d];
					if (i) var y = s ? i(v, _, d, t, e, o) : i(_, v, d, e, t, o);
					if (!(y === n ? _ === v || a(_, v, r, i, o) : y)) {
						m = !1;
						break;
					}
					g ||= d == "constructor";
				}
				if (m && !g) {
					var b = e.constructor, x = t.constructor;
					b != x && "constructor" in e && "constructor" in t && !(typeof b == "function" && b instanceof b && typeof x == "function" && x instanceof x) && (m = !1);
				}
				return o.delete(e), o.delete(t), m;
			}
			function Jo(e) {
				return Ps(Os(e, n, nc), e + "");
			}
			function Yo(e) {
				return Ki(e, $, is);
			}
			function Xo(e) {
				return Ki(e, wd, as);
			}
			var Zo = Hn ? function(e) {
				return Hn.get(e);
			} : Bf;
			function Qo(e) {
				for (var t = e.name + "", n = Zn[t], r = N.call(Zn, t) ? n.length : 0; r--;) {
					var i = n[r], a = i.func;
					if (a == null || a == e) return i.name;
				}
				return t;
			}
			function $o(e) {
				return (N.call(V, "placeholder") ? V : e).placeholder;
			}
			function G() {
				var e = V.iteratee || Nf;
				return e = e === Nf ? fa : e, arguments.length ? e(arguments[0], arguments[1]) : e;
			}
			function es(e, t) {
				var n = e.__data__;
				return _s(t) ? n[typeof t == "string" ? "string" : "hash"] : n.map;
			}
			function ts(e) {
				for (var t = $(e), n = t.length; n--;) {
					var r = t[n], i = e[r];
					t[n] = [
						r,
						i,
						Ss(i)
					];
				}
				return t;
			}
			function ns(e, t) {
				var r = pr(e, t);
				return ca(r) ? r : n;
			}
			function rs(e) {
				var t = N.call(e, Yt), r = e[Yt];
				try {
					e[Yt] = n;
					var i = !0;
				} catch {}
				var a = Ft.call(e);
				return i && (t ? e[Yt] = r : delete e[Yt]), a;
			}
			var is = nn ? function(e) {
				return e == null ? [] : (e = M(e), Pn(nn(e), function(t) {
					return Gt.call(e, t);
				}));
			} : Yf, as = nn ? function(e) {
				for (var t = []; e;) Ln(t, is(e)), e = Ut(e);
				return t;
			} : Yf, K = qi;
			(vn && K(new vn(/* @__PURE__ */ new ArrayBuffer(1))) != Be || yn && K(new yn()) != De || bn && K(bn.resolve()) != je || Sn && K(new Sn()) != Pe || z && K(new z()) != Re) && (K = function(e) {
				var t = qi(e), r = t == Ae ? e.constructor : n, i = r ? Bs(r) : "";
				if (i) switch (i) {
					case Sr: return Be;
					case Or: return De;
					case kr: return je;
					case Mr: return Pe;
					case Nr: return Re;
				}
				return t;
			});
			function os(e, t, n) {
				for (var r = -1, i = n.length; ++r < i;) {
					var a = n[r], o = a.size;
					switch (a.type) {
						case "drop":
							e += o;
							break;
						case "dropRight":
							t -= o;
							break;
						case "take":
							t = L(t, e + o);
							break;
						case "takeRight":
							e = P(e, t - o);
							break;
					}
				}
				return {
					start: e,
					end: t
				};
			}
			function ss(e) {
				var t = e.match(ct);
				return t ? t[1].split(lt) : [];
			}
			function cs(e, t, n) {
				t = $a(t, e);
				for (var r = -1, i = t.length, a = !1; ++r < i;) {
					var o = zs(t[r]);
					if (!(a = e != null && n(e, o))) break;
					e = e[o];
				}
				return a || ++r != i ? a : (i = e == null ? 0 : e.length, !!i && ku(i) && ms(o, i) && (q(e) || hu(e)));
			}
			function ls(e) {
				var t = e.length, n = new e.constructor(t);
				return t && typeof e[0] == "string" && N.call(e, "index") && (n.index = e.index, n.input = e.input), n;
			}
			function us(e) {
				return typeof e.constructor == "function" && !xs(e) ? Lr(Ut(e)) : {};
			}
			function ds(e, t, n) {
				var r = e.constructor;
				switch (t) {
					case ze: return io(e);
					case xe:
					case Se: return new r(+e);
					case Be: return ao(e, n);
					case Ve:
					case T:
					case He:
					case Ue:
					case We:
					case E:
					case Ge:
					case Ke:
					case qe: return co(e, n);
					case De: return new r();
					case Oe:
					case Fe: return new r(e);
					case Ne: return oo(e);
					case Pe: return new r();
					case Ie: return so(e);
				}
			}
			function fs(e, t) {
				var n = t.length;
				if (!n) return e;
				var r = n - 1;
				return t[r] = (n > 1 ? "& " : "") + t[r], t = t.join(n > 2 ? ", " : " "), e.replace(st, "{\n/* [wrapped with " + t + "] */\n");
			}
			function ps(e) {
				return q(e) || hu(e) || !!(qt && e && e[qt]);
			}
			function ms(e, t) {
				var n = typeof e;
				return t ??= fe, !!t && (n == "number" || n != "symbol" && yt.test(e)) && e > -1 && e % 1 == 0 && e < t;
			}
			function hs(e, t, n) {
				if (!Y(n)) return !1;
				var r = typeof t;
				return (r == "number" ? _u(n) && ms(t, n.length) : r == "string" && t in n) ? fu(n[t], e) : !1;
			}
			function gs(e, t) {
				if (q(e)) return !1;
				var n = typeof e;
				return n == "number" || n == "symbol" || n == "boolean" || e == null || Uu(e) ? !0 : O.test(e) || !rt.test(e) || t != null && e in M(t);
			}
			function _s(e) {
				var t = typeof e;
				return t == "string" || t == "number" || t == "symbol" || t == "boolean" ? e !== "__proto__" : e === null;
			}
			function vs(e) {
				var t = Qo(e), n = V[t];
				if (typeof n != "function" || !(t in H.prototype)) return !1;
				if (e === n) return !0;
				var r = Zo(n);
				return !!r && e === r[0];
			}
			function ys(e) {
				return !!Pt && Pt in e;
			}
			var bs = jt ? Du : Xf;
			function xs(e) {
				var t = e && e.constructor;
				return e === (typeof t == "function" && t.prototype || At);
			}
			function Ss(e) {
				return e === e && !Y(e);
			}
			function Cs(e, t) {
				return function(r) {
					return r == null ? !1 : r[e] === t && (t !== n || e in M(r));
				};
			}
			function ws(e) {
				var t = Jl(e, function(e) {
					return n.size === u && n.clear(), e;
				}), n = t.cache;
				return t;
			}
			function Ts(e, t) {
				var n = e[1], r = t[1], i = n | r, a = i < (_ | v | ne), o = r == ne && n == b || r == ne && n == re && e[7].length <= t[8] || r == (ne | re) && t[7].length <= t[8] && n == b;
				if (!(a || o)) return e;
				r & _ && (e[2] = t[2], i |= n & _ ? 0 : y);
				var s = t[3];
				if (s) {
					var c = e[3];
					e[3] = c ? fo(c, s, t[4]) : s, e[4] = c ? yr(e[3], d) : t[4];
				}
				return s = t[5], s && (c = e[5], e[5] = c ? po(c, s, t[6]) : s, e[6] = c ? yr(e[5], d) : t[6]), s = t[7], s && (e[7] = s), r & ne && (e[8] = e[8] == null ? t[8] : L(e[8], t[8])), e[9] ??= t[9], e[0] = t[0], e[1] = i, e;
			}
			function Es(e) {
				var t = [];
				if (e != null) for (var n in M(e)) t.push(n);
				return t;
			}
			function Ds(e) {
				return Ft.call(e);
			}
			function Os(e, t, r) {
				return t = P(t === n ? e.length - 1 : t, 0), function() {
					for (var n = arguments, i = -1, a = P(n.length - t, 0), o = A(a); ++i < a;) o[i] = n[t + i];
					i = -1;
					for (var s = A(t + 1); ++i < t;) s[i] = n[i];
					return s[t] = r(o), kn(e, this, s);
				};
			}
			function ks(e, t) {
				return t.length < 2 ? e : Gi(e, La(t, 0, -1));
			}
			function As(e, t) {
				for (var r = e.length, i = L(t.length, r), a = mo(e); i--;) {
					var o = t[i];
					e[i] = ms(o, r) ? a[o] : n;
				}
				return e;
			}
			function js(e, t) {
				if (!(t === "constructor" && typeof e[t] == "function") && t != "__proto__") return e[t];
			}
			var Ms = Is(Pa), Ns = $t || function(e, t) {
				return R.setTimeout(e, t);
			}, Ps = Is(Fa);
			function Fs(e, t, n) {
				var r = t + "";
				return Ps(e, fs(r, Vs(ss(r), n)));
			}
			function Is(e) {
				var t = 0, r = 0;
				return function() {
					var i = fn(), a = se - (i - r);
					if (r = i, a > 0) {
						if (++t >= S) return arguments[0];
					} else t = 0;
					return e.apply(n, arguments);
				};
			}
			function Ls(e, t) {
				var r = -1, i = e.length, a = i - 1;
				for (t = t === n ? i : t; ++r < t;) {
					var o = Oa(r, a), s = e[o];
					e[o] = e[r], e[r] = s;
				}
				return e.length = t, e;
			}
			var Rs = ws(function(e) {
				var t = [];
				return e.charCodeAt(0) === 46 && t.push(""), e.replace(k, function(e, n, r, i) {
					t.push(r ? i.replace(ft, "$1") : n || e);
				}), t;
			});
			function zs(e) {
				if (typeof e == "string" || Uu(e)) return e;
				var t = e + "";
				return t == "0" && 1 / e == -de ? "-0" : t;
			}
			function Bs(e) {
				if (e != null) {
					try {
						return Mt.call(e);
					} catch {}
					try {
						return e + "";
					} catch {}
				}
				return "";
			}
			function Vs(e, t) {
				return jn(_e, function(n) {
					var r = "_." + n[0];
					t & n[1] && !Fn(e, r) && e.push(r);
				}), e.sort();
			}
			function Hs(e) {
				if (e instanceof H) return e.clone();
				var t = new zr(e.__wrapped__, e.__chain__);
				return t.__actions__ = mo(e.__actions__), t.__index__ = e.__index__, t.__values__ = e.__values__, t;
			}
			function Us(e, t, r) {
				t = (r ? hs(e, t, r) : t === n) ? 1 : P(Z(t), 0);
				var i = e == null ? 0 : e.length;
				if (!i || t < 1) return [];
				for (var a = 0, o = 0, s = A(en(i / t)); a < i;) s[o++] = La(e, a, a += t);
				return s;
			}
			function Ws(e) {
				for (var t = -1, n = e == null ? 0 : e.length, r = 0, i = []; ++t < n;) {
					var a = e[t];
					a && (i[r++] = a);
				}
				return i;
			}
			function Gs() {
				var e = arguments.length;
				if (!e) return [];
				for (var t = A(e - 1), n = arguments[0], r = e; r--;) t[r - 1] = arguments[r];
				return Ln(q(n) ? mo(n) : [n], U(t, 1));
			}
			var Ks = W(function(e, t) {
				return J(e) ? Ni(e, U(t, 1, J, !0)) : [];
			}), qs = W(function(e, t) {
				var r = pc(t);
				return J(r) && (r = n), J(e) ? Ni(e, U(t, 1, J, !0), G(r, 2)) : [];
			}), Js = W(function(e, t) {
				var r = pc(t);
				return J(r) && (r = n), J(e) ? Ni(e, U(t, 1, J, !0), n, r) : [];
			});
			function Ys(e, t, r) {
				var i = e == null ? 0 : e.length;
				return i ? (t = r || t === n ? 1 : Z(t), La(e, t < 0 ? 0 : t, i)) : [];
			}
			function Xs(e, t, r) {
				var i = e == null ? 0 : e.length;
				return i ? (t = r || t === n ? 1 : Z(t), t = i - t, La(e, 0, t < 0 ? 0 : t)) : [];
			}
			function Zs(e, t) {
				return e && e.length ? qa(e, G(t, 3), !0, !0) : [];
			}
			function Qs(e, t) {
				return e && e.length ? qa(e, G(t, 3), !0) : [];
			}
			function $s(e, t, n, r) {
				var i = e == null ? 0 : e.length;
				return i ? (n && typeof n != "number" && hs(e, t, n) && (n = 0, r = i), Ri(e, t, n, r)) : [];
			}
			function ec(e, t, n) {
				var r = e == null ? 0 : e.length;
				if (!r) return -1;
				var i = n == null ? 0 : Z(n);
				return i < 0 && (i = P(r + i, 0)), Gn(e, G(t, 3), i);
			}
			function tc(e, t, r) {
				var i = e == null ? 0 : e.length;
				if (!i) return -1;
				var a = i - 1;
				return r !== n && (a = Z(r), a = r < 0 ? P(i + a, 0) : L(a, i - 1)), Gn(e, G(t, 3), a, !0);
			}
			function nc(e) {
				return e != null && e.length ? U(e, 1) : [];
			}
			function rc(e) {
				return e != null && e.length ? U(e, de) : [];
			}
			function ic(e, t) {
				return e != null && e.length ? (t = t === n ? 1 : Z(t), U(e, t)) : [];
			}
			function ac(e) {
				for (var t = -1, n = e == null ? 0 : e.length, r = {}; ++t < n;) {
					var i = e[t];
					Ei(r, i[0], i[1]);
				}
				return r;
			}
			function oc(e) {
				return e && e.length ? e[0] : n;
			}
			function sc(e, t, n) {
				var r = e == null ? 0 : e.length;
				if (!r) return -1;
				var i = n == null ? 0 : Z(n);
				return i < 0 && (i = P(r + i, 0)), Kn(e, t, i);
			}
			function cc(e) {
				return e != null && e.length ? La(e, 0, -1) : [];
			}
			var lc = W(function(e) {
				var t = B(e, Za);
				return t.length && t[0] === e[0] ? Qi(t) : [];
			}), uc = W(function(e) {
				var t = pc(e), r = B(e, Za);
				return t === pc(r) ? t = n : r.pop(), r.length && r[0] === e[0] ? Qi(r, G(t, 2)) : [];
			}), dc = W(function(e) {
				var t = pc(e), r = B(e, Za);
				return t = typeof t == "function" ? t : n, t && r.pop(), r.length && r[0] === e[0] ? Qi(r, n, t) : [];
			});
			function fc(e, t) {
				return e == null ? "" : cn.call(e, t);
			}
			function pc(e) {
				var t = e == null ? 0 : e.length;
				return t ? e[t - 1] : n;
			}
			function mc(e, t, r) {
				var i = e == null ? 0 : e.length;
				if (!i) return -1;
				var a = i;
				return r !== n && (a = Z(r), a = a < 0 ? P(i + a, 0) : L(a, i - 1)), t === t ? Cr(e, t, a) : Gn(e, Jn, a, !0);
			}
			function hc(e, t) {
				return e && e.length ? xa(e, Z(t)) : n;
			}
			var gc = W(_c);
			function _c(e, t) {
				return e && e.length && t && t.length ? Ea(e, t) : e;
			}
			function vc(e, t, n) {
				return e && e.length && t && t.length ? Ea(e, t, G(n, 2)) : e;
			}
			function yc(e, t, r) {
				return e && e.length && t && t.length ? Ea(e, t, n, r) : e;
			}
			var bc = Jo(function(e, t) {
				var n = e == null ? 0 : e.length, r = Di(e, t);
				return Da(e, B(t, function(e) {
					return ms(e, n) ? +e : e;
				}).sort(lo)), r;
			});
			function xc(e, t) {
				var n = [];
				if (!(e && e.length)) return n;
				var r = -1, i = [], a = e.length;
				for (t = G(t, 3); ++r < a;) {
					var o = e[r];
					t(o, r, e) && (n.push(o), i.push(r));
				}
				return Da(e, i), n;
			}
			function Sc(e) {
				return e == null ? e : _n.call(e);
			}
			function Cc(e, t, r) {
				var i = e == null ? 0 : e.length;
				return i ? (r && typeof r != "number" && hs(e, t, r) ? (t = 0, r = i) : (t = t == null ? 0 : Z(t), r = r === n ? i : Z(r)), La(e, t, r)) : [];
			}
			function wc(e, t) {
				return za(e, t);
			}
			function Tc(e, t, n) {
				return Ba(e, t, G(n, 2));
			}
			function Ec(e, t) {
				var n = e == null ? 0 : e.length;
				if (n) {
					var r = za(e, t);
					if (r < n && fu(e[r], t)) return r;
				}
				return -1;
			}
			function Dc(e, t) {
				return za(e, t, !0);
			}
			function Oc(e, t, n) {
				return Ba(e, t, G(n, 2), !0);
			}
			function kc(e, t) {
				if (e != null && e.length) {
					var n = za(e, t, !0) - 1;
					if (fu(e[n], t)) return n;
				}
				return -1;
			}
			function Ac(e) {
				return e && e.length ? Va(e) : [];
			}
			function jc(e, t) {
				return e && e.length ? Va(e, G(t, 2)) : [];
			}
			function Mc(e) {
				var t = e == null ? 0 : e.length;
				return t ? La(e, 1, t) : [];
			}
			function Nc(e, t, r) {
				return e && e.length ? (t = r || t === n ? 1 : Z(t), La(e, 0, t < 0 ? 0 : t)) : [];
			}
			function Pc(e, t, r) {
				var i = e == null ? 0 : e.length;
				return i ? (t = r || t === n ? 1 : Z(t), t = i - t, La(e, t < 0 ? 0 : t, i)) : [];
			}
			function Fc(e, t) {
				return e && e.length ? qa(e, G(t, 3), !1, !0) : [];
			}
			function Ic(e, t) {
				return e && e.length ? qa(e, G(t, 3)) : [];
			}
			var Lc = W(function(e) {
				return Wa(U(e, 1, J, !0));
			}), Rc = W(function(e) {
				var t = pc(e);
				return J(t) && (t = n), Wa(U(e, 1, J, !0), G(t, 2));
			}), zc = W(function(e) {
				var t = pc(e);
				return t = typeof t == "function" ? t : n, Wa(U(e, 1, J, !0), n, t);
			});
			function Bc(e) {
				return e && e.length ? Wa(e) : [];
			}
			function Vc(e, t) {
				return e && e.length ? Wa(e, G(t, 2)) : [];
			}
			function Hc(e, t) {
				return t = typeof t == "function" ? t : n, e && e.length ? Wa(e, n, t) : [];
			}
			function Uc(e) {
				if (!(e && e.length)) return [];
				var t = 0;
				return e = Pn(e, function(e) {
					if (J(e)) return t = P(e.length, t), !0;
				}), tr(t, function(t) {
					return B(e, Xn(t));
				});
			}
			function Wc(e, t) {
				if (!(e && e.length)) return [];
				var r = Uc(e);
				return t == null ? r : B(r, function(e) {
					return kn(t, n, e);
				});
			}
			var Gc = W(function(e, t) {
				return J(e) ? Ni(e, t) : [];
			}), Kc = W(function(e) {
				return Ya(Pn(e, J));
			}), qc = W(function(e) {
				var t = pc(e);
				return J(t) && (t = n), Ya(Pn(e, J), G(t, 2));
			}), Jc = W(function(e) {
				var t = pc(e);
				return t = typeof t == "function" ? t : n, Ya(Pn(e, J), n, t);
			}), Yc = W(Uc);
			function Xc(e, t) {
				return Xa(e || [], t || [], xi);
			}
			function Zc(e, t) {
				return Xa(e || [], t || [], Na);
			}
			var Qc = W(function(e) {
				var t = e.length, r = t > 1 ? e[t - 1] : n;
				return r = typeof r == "function" ? (e.pop(), r) : n, Wc(e, r);
			});
			function $c(e) {
				var t = V(e);
				return t.__chain__ = !0, t;
			}
			function el(e, t) {
				return t(e), e;
			}
			function tl(e, t) {
				return t(e);
			}
			var nl = Jo(function(e) {
				var t = e.length, r = t ? e[0] : 0, i = this.__wrapped__, a = function(t) {
					return Di(t, e);
				};
				return t > 1 || this.__actions__.length || !(i instanceof H) || !ms(r) ? this.thru(a) : (i = i.slice(r, +r + +!!t), i.__actions__.push({
					func: tl,
					args: [a],
					thisArg: n
				}), new zr(i, this.__chain__).thru(function(e) {
					return t && !e.length && e.push(n), e;
				}));
			});
			function rl() {
				return $c(this);
			}
			function il() {
				return new zr(this.value(), this.__chain__);
			}
			function al() {
				this.__values__ === n && (this.__values__ = Xu(this.value()));
				var e = this.__index__ >= this.__values__.length;
				return {
					done: e,
					value: e ? n : this.__values__[this.__index__++]
				};
			}
			function ol() {
				return this;
			}
			function sl(e) {
				for (var t, r = this; r instanceof Rr;) {
					var i = Hs(r);
					i.__index__ = 0, i.__values__ = n, t ? a.__wrapped__ = i : t = i;
					var a = i;
					r = r.__wrapped__;
				}
				return a.__wrapped__ = e, t;
			}
			function cl() {
				var e = this.__wrapped__;
				if (e instanceof H) {
					var t = e;
					return this.__actions__.length && (t = new H(this)), t = t.reverse(), t.__actions__.push({
						func: tl,
						args: [Sc],
						thisArg: n
					}), new zr(t, this.__chain__);
				}
				return this.thru(Sc);
			}
			function ll() {
				return Ja(this.__wrapped__, this.__actions__);
			}
			var ul = vo(function(e, t, n) {
				N.call(e, n) ? ++e[n] : Ei(e, n, 1);
			});
			function dl(e, t, r) {
				var i = q(e) ? Nn : Ii;
				return r && hs(e, t, r) && (t = n), i(e, G(t, 3));
			}
			function fl(e, t) {
				return (q(e) ? Pn : zi)(e, G(t, 3));
			}
			var pl = Do(ec), ml = Do(tc);
			function hl(e, t) {
				return U(wl(e, t), 1);
			}
			function gl(e, t) {
				return U(wl(e, t), de);
			}
			function _l(e, t, r) {
				return r = r === n ? 1 : Z(r), U(wl(e, t), r);
			}
			function vl(e, t) {
				return (q(e) ? jn : Pi)(e, G(t, 3));
			}
			function yl(e, t) {
				return (q(e) ? Mn : Fi)(e, G(t, 3));
			}
			var bl = vo(function(e, t, n) {
				N.call(e, n) ? e[n].push(t) : Ei(e, n, [t]);
			});
			function xl(e, t, n, r) {
				e = _u(e) ? e : Hd(e), n = n && !r ? Z(n) : 0;
				var i = e.length;
				return n < 0 && (n = P(i + n, 0)), Hu(e) ? n <= i && e.indexOf(t, n) > -1 : !!i && Kn(e, t, n) > -1;
			}
			var Sl = W(function(e, t, n) {
				var r = -1, i = typeof t == "function", a = _u(e) ? A(e.length) : [];
				return Pi(e, function(e) {
					a[++r] = i ? kn(t, e, n) : ea(e, t, n);
				}), a;
			}), Cl = vo(function(e, t, n) {
				Ei(e, n, t);
			});
			function wl(e, t) {
				return (q(e) ? B : ga)(e, G(t, 3));
			}
			function Tl(e, t, r, i) {
				return e == null ? [] : (q(t) || (t = t == null ? [] : [t]), r = i ? n : r, q(r) || (r = r == null ? [] : [r]), Sa(e, t, r));
			}
			var El = vo(function(e, t, n) {
				e[+!n].push(t);
			}, function() {
				return [[], []];
			});
			function Dl(e, t, n) {
				var r = q(e) ? Rn : Qn, i = arguments.length < 3;
				return r(e, G(t, 4), n, i, Pi);
			}
			function Ol(e, t, n) {
				var r = q(e) ? zn : Qn, i = arguments.length < 3;
				return r(e, G(t, 4), n, i, Fi);
			}
			function kl(e, t) {
				return (q(e) ? Pn : zi)(e, Yl(G(t, 3)));
			}
			function Al(e) {
				return (q(e) ? _i : ja)(e);
			}
			function jl(e, t, r) {
				return t = (r ? hs(e, t, r) : t === n) ? 1 : Z(t), (q(e) ? vi : Ma)(e, t);
			}
			function Ml(e) {
				return (q(e) ? yi : Ia)(e);
			}
			function Nl(e) {
				if (e == null) return 0;
				if (_u(e)) return Hu(e) ? wr(e) : e.length;
				var t = K(e);
				return t == De || t == Pe ? e.size : pa(e).length;
			}
			function Pl(e, t, r) {
				var i = q(e) ? Bn : Ra;
				return r && hs(e, t, r) && (t = n), i(e, G(t, 3));
			}
			var Fl = W(function(e, t) {
				if (e == null) return [];
				var n = t.length;
				return n > 1 && hs(e, t[0], t[1]) ? t = [] : n > 2 && hs(t[0], t[1], t[2]) && (t = [t[0]]), Sa(e, U(t, 1), []);
			}), Il = Qt || function() {
				return R.Date.now();
			};
			function Ll(e, t) {
				if (typeof t != "function") throw new Dt(o);
				return e = Z(e), function() {
					if (--e < 1) return t.apply(this, arguments);
				};
			}
			function Rl(e, t, r) {
				return t = r ? n : t, t = e && t == null ? e.length : t, Vo(e, ne, n, n, n, n, t);
			}
			function zl(e, t) {
				var r;
				if (typeof t != "function") throw new Dt(o);
				return e = Z(e), function() {
					return --e > 0 && (r = t.apply(this, arguments)), e <= 1 && (t = n), r;
				};
			}
			var Bl = W(function(e, t, n) {
				var r = _;
				if (n.length) {
					var i = yr(n, $o(Bl));
					r |= ee;
				}
				return Vo(e, r, t, n, i);
			}), Vl = W(function(e, t, n) {
				var r = _ | v;
				if (n.length) {
					var i = yr(n, $o(Vl));
					r |= ee;
				}
				return Vo(t, r, e, n, i);
			});
			function Hl(e, t, r) {
				t = r ? n : t;
				var i = Vo(e, b, n, n, n, n, n, t);
				return i.placeholder = Hl.placeholder, i;
			}
			function Ul(e, t, r) {
				t = r ? n : t;
				var i = Vo(e, x, n, n, n, n, n, t);
				return i.placeholder = Ul.placeholder, i;
			}
			function Wl(e, t, r) {
				var i, a, s, c, l, u, d = 0, f = !1, p = !1, m = !0;
				if (typeof e != "function") throw new Dt(o);
				t = $u(t) || 0, Y(r) && (f = !!r.leading, p = "maxWait" in r, s = p ? P($u(r.maxWait) || 0, t) : s, m = "trailing" in r ? !!r.trailing : m);
				function h(t) {
					var r = i, o = a;
					return i = a = n, d = t, c = e.apply(o, r), c;
				}
				function g(e) {
					return d = e, l = Ns(y, t), f ? h(e) : c;
				}
				function _(e) {
					var n = e - u, r = e - d, i = t - n;
					return p ? L(i, s - r) : i;
				}
				function v(e) {
					var r = e - u, i = e - d;
					return u === n || r >= t || r < 0 || p && i >= s;
				}
				function y() {
					var e = Il();
					if (v(e)) return b(e);
					l = Ns(y, _(e));
				}
				function b(e) {
					return l = n, m && i ? h(e) : (i = a = n, c);
				}
				function x() {
					l !== n && no(l), d = 0, i = u = a = l = n;
				}
				function ee() {
					return l === n ? c : b(Il());
				}
				function te() {
					var e = Il(), r = v(e);
					if (i = arguments, a = this, u = e, r) {
						if (l === n) return g(u);
						if (p) return no(l), l = Ns(y, t), h(u);
					}
					return l === n && (l = Ns(y, t)), c;
				}
				return te.cancel = x, te.flush = ee, te;
			}
			var Gl = W(function(e, t) {
				return Mi(e, 1, t);
			}), Kl = W(function(e, t, n) {
				return Mi(e, $u(t) || 0, n);
			});
			function ql(e) {
				return Vo(e, ie);
			}
			function Jl(e, t) {
				if (typeof e != "function" || t != null && typeof t != "function") throw new Dt(o);
				var n = function() {
					var r = arguments, i = t ? t.apply(this, r) : r[0], a = n.cache;
					if (a.has(i)) return a.get(i);
					var o = e.apply(this, r);
					return n.cache = a.set(i, o) || a, o;
				};
				return n.cache = new (Jl.Cache || ti)(), n;
			}
			Jl.Cache = ti;
			function Yl(e) {
				if (typeof e != "function") throw new Dt(o);
				return function() {
					var t = arguments;
					switch (t.length) {
						case 0: return !e.call(this);
						case 1: return !e.call(this, t[0]);
						case 2: return !e.call(this, t[0], t[1]);
						case 3: return !e.call(this, t[0], t[1], t[2]);
					}
					return !e.apply(this, t);
				};
			}
			function Xl(e) {
				return zl(2, e);
			}
			var Zl = eo(function(e, t) {
				t = t.length == 1 && q(t[0]) ? B(t[0], ir(G())) : B(U(t, 1), ir(G()));
				var n = t.length;
				return W(function(r) {
					for (var i = -1, a = L(r.length, n); ++i < a;) r[i] = t[i].call(this, r[i]);
					return kn(e, this, r);
				});
			}), Ql = W(function(e, t) {
				return Vo(e, ee, n, t, yr(t, $o(Ql)));
			}), $l = W(function(e, t) {
				return Vo(e, te, n, t, yr(t, $o($l)));
			}), eu = Jo(function(e, t) {
				return Vo(e, re, n, n, n, t);
			});
			function tu(e, t) {
				if (typeof e != "function") throw new Dt(o);
				return t = t === n ? t : Z(t), W(e, t);
			}
			function nu(e, t) {
				if (typeof e != "function") throw new Dt(o);
				return t = t == null ? 0 : P(Z(t), 0), W(function(n) {
					var r = n[t], i = to(n, 0, t);
					return r && Ln(i, r), kn(e, this, i);
				});
			}
			function ru(e, t, n) {
				var r = !0, i = !0;
				if (typeof e != "function") throw new Dt(o);
				return Y(n) && (r = "leading" in n ? !!n.leading : r, i = "trailing" in n ? !!n.trailing : i), Wl(e, t, {
					leading: r,
					maxWait: t,
					trailing: i
				});
			}
			function iu(e) {
				return Rl(e, 1);
			}
			function au(e, t) {
				return Ql(Qa(t), e);
			}
			function ou() {
				if (!arguments.length) return [];
				var e = arguments[0];
				return q(e) ? e : [e];
			}
			function su(e) {
				return ki(e, m);
			}
			function cu(e, t) {
				return t = typeof t == "function" ? t : n, ki(e, m, t);
			}
			function lu(e) {
				return ki(e, f | m);
			}
			function uu(e, t) {
				return t = typeof t == "function" ? t : n, ki(e, f | m, t);
			}
			function du(e, t) {
				return t == null || ji(e, t, $(t));
			}
			function fu(e, t) {
				return e === t || e !== e && t !== t;
			}
			var pu = Io(Ji), mu = Io(function(e, t) {
				return e >= t;
			}), hu = ta(function() {
				return arguments;
			}()) ? ta : function(e) {
				return X(e) && N.call(e, "callee") && !Gt.call(e, "callee");
			}, q = A.isArray, gu = Cn ? ir(Cn) : na;
			function _u(e) {
				return e != null && ku(e.length) && !Du(e);
			}
			function J(e) {
				return X(e) && _u(e);
			}
			function vu(e) {
				return e === !0 || e === !1 || X(e) && qi(e) == xe;
			}
			var yu = rn || Xf, bu = wn ? ir(wn) : ra;
			function xu(e) {
				return X(e) && e.nodeType === 1 && !Ru(e);
			}
			function Su(e) {
				if (e == null) return !0;
				if (_u(e) && (q(e) || typeof e == "string" || typeof e.splice == "function" || yu(e) || Wu(e) || hu(e))) return !e.length;
				var t = K(e);
				if (t == De || t == Pe) return !e.size;
				if (xs(e)) return !pa(e).length;
				for (var n in e) if (N.call(e, n)) return !1;
				return !0;
			}
			function Cu(e, t) {
				return ia(e, t);
			}
			function wu(e, t, r) {
				r = typeof r == "function" ? r : n;
				var i = r ? r(e, t) : n;
				return i === n ? ia(e, t, n, r) : !!i;
			}
			function Tu(e) {
				if (!X(e)) return !1;
				var t = qi(e);
				return t == we || t == Ce || typeof e.message == "string" && typeof e.name == "string" && !Ru(e);
			}
			function Eu(e) {
				return typeof e == "number" && sn(e);
			}
			function Du(e) {
				if (!Y(e)) return !1;
				var t = qi(e);
				return t == Te || t == Ee || t == be || t == Me;
			}
			function Ou(e) {
				return typeof e == "number" && e == Z(e);
			}
			function ku(e) {
				return typeof e == "number" && e > -1 && e % 1 == 0 && e <= fe;
			}
			function Y(e) {
				var t = typeof e;
				return e != null && (t == "object" || t == "function");
			}
			function X(e) {
				return typeof e == "object" && !!e;
			}
			var Au = Tn ? ir(Tn) : oa;
			function ju(e, t) {
				return e === t || sa(e, t, ts(t));
			}
			function Mu(e, t, r) {
				return r = typeof r == "function" ? r : n, sa(e, t, ts(t), r);
			}
			function Nu(e) {
				return Lu(e) && e != +e;
			}
			function Pu(e) {
				if (bs(e)) throw new Ct(a);
				return ca(e);
			}
			function Fu(e) {
				return e === null;
			}
			function Iu(e) {
				return e == null;
			}
			function Lu(e) {
				return typeof e == "number" || X(e) && qi(e) == Oe;
			}
			function Ru(e) {
				if (!X(e) || qi(e) != Ae) return !1;
				var t = Ut(e);
				if (t === null) return !0;
				var n = N.call(t, "constructor") && t.constructor;
				return typeof n == "function" && n instanceof n && Mt.call(n) == It;
			}
			var zu = En ? ir(En) : la;
			function Bu(e) {
				return Ou(e) && e >= -fe && e <= fe;
			}
			var Vu = Dn ? ir(Dn) : ua;
			function Hu(e) {
				return typeof e == "string" || !q(e) && X(e) && qi(e) == Fe;
			}
			function Uu(e) {
				return typeof e == "symbol" || X(e) && qi(e) == Ie;
			}
			var Wu = On ? ir(On) : da;
			function Gu(e) {
				return e === n;
			}
			function Ku(e) {
				return X(e) && K(e) == Re;
			}
			function qu(e) {
				return X(e) && qi(e) == w;
			}
			var Ju = Io(ha), Yu = Io(function(e, t) {
				return e <= t;
			});
			function Xu(e) {
				if (!e) return [];
				if (_u(e)) return Hu(e) ? Tr(e) : mo(e);
				if (Jt && e[Jt]) return gr(e[Jt]());
				var t = K(e);
				return (t == De ? _r : t == Pe ? br : Hd)(e);
			}
			function Zu(e) {
				return e ? (e = $u(e), e === de || e === -de ? (e < 0 ? -1 : 1) * pe : e === e ? e : 0) : e === 0 ? e : 0;
			}
			function Z(e) {
				var t = Zu(e), n = t % 1;
				return t === t ? n ? t - n : t : 0;
			}
			function Qu(e) {
				return e ? Oi(Z(e), 0, C) : 0;
			}
			function $u(e) {
				if (typeof e == "number") return e;
				if (Uu(e)) return me;
				if (Y(e)) {
					var t = typeof e.valueOf == "function" ? e.valueOf() : e;
					e = Y(t) ? t + "" : t;
				}
				if (typeof e != "string") return e === 0 ? e : +e;
				e = rr(e);
				var n = gt.test(e);
				return n || vt.test(e) ? gn(e.slice(2), n ? 2 : 8) : ht.test(e) ? me : +e;
			}
			function ed(e) {
				return ho(e, wd(e));
			}
			function td(e) {
				return e ? Oi(Z(e), -fe, fe) : e === 0 ? e : 0;
			}
			function Q(e) {
				return e == null ? "" : Ua(e);
			}
			var nd = yo(function(e, t) {
				if (xs(t) || _u(t)) {
					ho(t, $(t), e);
					return;
				}
				for (var n in t) N.call(t, n) && xi(e, n, t[n]);
			}), rd = yo(function(e, t) {
				ho(t, wd(t), e);
			}), id = yo(function(e, t, n, r) {
				ho(t, wd(t), e, r);
			}), ad = yo(function(e, t, n, r) {
				ho(t, $(t), e, r);
			}), od = Jo(Di);
			function sd(e, t) {
				var n = Lr(e);
				return t == null ? n : wi(n, t);
			}
			var cd = W(function(e, t) {
				e = M(e);
				var r = -1, i = t.length, a = i > 2 ? t[2] : n;
				for (a && hs(t[0], t[1], a) && (i = 1); ++r < i;) for (var o = t[r], s = wd(o), c = -1, l = s.length; ++c < l;) {
					var u = s[c], d = e[u];
					(d === n || fu(d, At[u]) && !N.call(e, u)) && (e[u] = o[u]);
				}
				return e;
			}), ld = W(function(e) {
				return e.push(n, Uo), kn(Od, n, e);
			});
			function ud(e, t) {
				return Wn(e, G(t, 3), Hi);
			}
			function dd(e, t) {
				return Wn(e, G(t, 3), Ui);
			}
			function fd(e, t) {
				return e == null ? e : Bi(e, G(t, 3), wd);
			}
			function pd(e, t) {
				return e == null ? e : Vi(e, G(t, 3), wd);
			}
			function md(e, t) {
				return e && Hi(e, G(t, 3));
			}
			function hd(e, t) {
				return e && Ui(e, G(t, 3));
			}
			function gd(e) {
				return e == null ? [] : Wi(e, $(e));
			}
			function _d(e) {
				return e == null ? [] : Wi(e, wd(e));
			}
			function vd(e, t, r) {
				var i = e == null ? n : Gi(e, t);
				return i === n ? r : i;
			}
			function yd(e, t) {
				return e != null && cs(e, t, Yi);
			}
			function bd(e, t) {
				return e != null && cs(e, t, Xi);
			}
			var xd = Ao(function(e, t, n) {
				t != null && typeof t.toString != "function" && (t = Ft.call(t)), e[t] = n;
			}, Of(Mf)), Sd = Ao(function(e, t, n) {
				t != null && typeof t.toString != "function" && (t = Ft.call(t)), N.call(e, t) ? e[t].push(n) : e[t] = [n];
			}, G), Cd = W(ea);
			function $(e) {
				return _u(e) ? gi(e) : pa(e);
			}
			function wd(e) {
				return _u(e) ? gi(e, !0) : ma(e);
			}
			function Td(e, t) {
				var n = {};
				return t = G(t, 3), Hi(e, function(e, r, i) {
					Ei(n, t(e, r, i), e);
				}), n;
			}
			function Ed(e, t) {
				var n = {};
				return t = G(t, 3), Hi(e, function(e, r, i) {
					Ei(n, r, t(e, r, i));
				}), n;
			}
			var Dd = yo(function(e, t, n) {
				ya(e, t, n);
			}), Od = yo(function(e, t, n, r) {
				ya(e, t, n, r);
			}), kd = Jo(function(e, t) {
				var n = {};
				if (e == null) return n;
				var r = !1;
				t = B(t, function(t) {
					return t = $a(t, e), r ||= t.length > 1, t;
				}), ho(e, Xo(e), n), r && (n = ki(n, f | p | m, Wo));
				for (var i = t.length; i--;) Ga(n, t[i]);
				return n;
			});
			function Ad(e, t) {
				return Md(e, Yl(G(t)));
			}
			var jd = Jo(function(e, t) {
				return e == null ? {} : Ca(e, t);
			});
			function Md(e, t) {
				if (e == null) return {};
				var n = B(Xo(e), function(e) {
					return [e];
				});
				return t = G(t), wa(e, n, function(e, n) {
					return t(e, n[0]);
				});
			}
			function Nd(e, t, r) {
				t = $a(t, e);
				var i = -1, a = t.length;
				for (a || (a = 1, e = n); ++i < a;) {
					var o = e == null ? n : e[zs(t[i])];
					o === n && (i = a, o = r), e = Du(o) ? o.call(e) : o;
				}
				return e;
			}
			function Pd(e, t, n) {
				return e == null ? e : Na(e, t, n);
			}
			function Fd(e, t, r, i) {
				return i = typeof i == "function" ? i : n, e == null ? e : Na(e, t, r, i);
			}
			var Id = Bo($), Ld = Bo(wd);
			function Rd(e, t, n) {
				var r = q(e), i = r || yu(e) || Wu(e);
				if (t = G(t, 4), n == null) {
					var a = e && e.constructor;
					n = i ? r ? new a() : [] : Y(e) && Du(a) ? Lr(Ut(e)) : {};
				}
				return (i ? jn : Hi)(e, function(e, r, i) {
					return t(n, e, r, i);
				}), n;
			}
			function zd(e, t) {
				return e == null ? !0 : Ga(e, t);
			}
			function Bd(e, t, n) {
				return e == null ? e : Ka(e, t, Qa(n));
			}
			function Vd(e, t, r, i) {
				return i = typeof i == "function" ? i : n, e == null ? e : Ka(e, t, Qa(r), i);
			}
			function Hd(e) {
				return e == null ? [] : ar(e, $(e));
			}
			function Ud(e) {
				return e == null ? [] : ar(e, wd(e));
			}
			function Wd(e, t, r) {
				return r === n && (r = t, t = n), r !== n && (r = $u(r), r = r === r ? r : 0), t !== n && (t = $u(t), t = t === t ? t : 0), Oi($u(e), t, r);
			}
			function Gd(e, t, r) {
				return t = Zu(t), r === n ? (r = t, t = 0) : r = Zu(r), e = $u(e), Zi(e, t, r);
			}
			function Kd(e, t, r) {
				if (r && typeof r != "boolean" && hs(e, t, r) && (t = r = n), r === n && (typeof t == "boolean" ? (r = t, t = n) : typeof e == "boolean" && (r = e, e = n)), e === n && t === n ? (e = 0, t = 1) : (e = Zu(e), t === n ? (t = e, e = 0) : t = Zu(t)), e > t) {
					var i = e;
					e = t, t = i;
				}
				if (r || e % 1 || t % 1) {
					var a = mn();
					return L(e + a * (t - e + hn("1e-" + ((a + "").length - 1))), t);
				}
				return Oa(e, t);
			}
			var qd = wo(function(e, t, n) {
				return t = t.toLowerCase(), e + (n ? Jd(t) : t);
			});
			function Jd(e) {
				return Sf(Q(e).toLowerCase());
			}
			function Yd(e) {
				return e = Q(e), e && e.replace(bt, ur).replace(on, "");
			}
			function Xd(e, t, r) {
				e = Q(e), t = Ua(t);
				var i = e.length;
				r = r === n ? i : Oi(Z(r), 0, i);
				var a = r;
				return r -= t.length, r >= 0 && e.slice(r, a) == t;
			}
			function Zd(e) {
				return e = Q(e), e && et.test(e) ? e.replace(Qe, dr) : e;
			}
			function Qd(e) {
				return e = Q(e), e && at.test(e) ? e.replace(it, "\\$&") : e;
			}
			var $d = wo(function(e, t, n) {
				return e + (n ? "-" : "") + t.toLowerCase();
			}), ef = wo(function(e, t, n) {
				return e + (n ? " " : "") + t.toLowerCase();
			}), tf = Co("toLowerCase");
			function nf(e, t, n) {
				e = Q(e), t = Z(t);
				var r = t ? wr(e) : 0;
				if (!t || r >= t) return e;
				var i = (t - r) / 2;
				return No(tn(i), n) + e + No(en(i), n);
			}
			function rf(e, t, n) {
				e = Q(e), t = Z(t);
				var r = t ? wr(e) : 0;
				return t && r < t ? e + No(t - r, n) : e;
			}
			function af(e, t, n) {
				e = Q(e), t = Z(t);
				var r = t ? wr(e) : 0;
				return t && r < t ? No(t - r, n) + e : e;
			}
			function of(e, t, n) {
				return n || t == null ? t = 0 : t &&= +t, pn(Q(e).replace(ot, ""), t || 0);
			}
			function sf(e, t, r) {
				return t = (r ? hs(e, t, r) : t === n) ? 1 : Z(t), Aa(Q(e), t);
			}
			function cf() {
				var e = arguments, t = Q(e[0]);
				return e.length < 3 ? t : t.replace(e[1], e[2]);
			}
			var lf = wo(function(e, t, n) {
				return e + (n ? "_" : "") + t.toLowerCase();
			});
			function uf(e, t, r) {
				return r && typeof r != "number" && hs(e, t, r) && (t = r = n), r = r === n ? C : r >>> 0, r ? (e = Q(e), e && (typeof t == "string" || t != null && !zu(t)) && (t = Ua(t), !t && mr(e)) ? to(Tr(e), 0, r) : e.split(t, r)) : [];
			}
			var df = wo(function(e, t, n) {
				return e + (n ? " " : "") + Sf(t);
			});
			function ff(e, t, n) {
				return e = Q(e), n = n == null ? 0 : Oi(Z(n), 0, e.length), t = Ua(t), e.slice(n, n + t.length) == t;
			}
			function pf(e, t, r) {
				var i = V.templateSettings;
				r && hs(e, t, r) && (t = n), e = Q(e), t = ad({}, t, i, Ho);
				var a = ad({}, t.imports, i.imports, Ho), o = $(a), l = ar(a, o);
				jn(o, function(e) {
					if (dt.test(e)) throw new Ct(c);
				});
				var u, d, f = 0, p = t.interpolate || xt, m = "__p += '", h = Tt((t.escape || xt).source + "|" + p.source + "|" + (p === nt ? pt : xt).source + "|" + (t.evaluate || xt).source + "|$", "g"), g = "//# sourceURL=" + (N.call(t, "sourceURL") ? (t.sourceURL + "").replace(/\s/g, " ") : "lodash.templateSources[" + ++dn + "]") + "\n";
				e.replace(h, function(t, n, r, i, a, o) {
					return r ||= i, m += e.slice(f, o).replace(St, fr), n && (u = !0, m += "' +\n__e(" + n + ") +\n'"), a && (d = !0, m += "';\n" + a + ";\n__p += '"), r && (m += "' +\n((__t = (" + r + ")) == null ? '' : __t) +\n'"), f = o + t.length, t;
				}), m += "';\n";
				var _ = N.call(t, "variable") && t.variable;
				if (!_) m = "with (obj) {\n" + m + "\n}\n";
				else if (dt.test(_)) throw new Ct(s);
				m = (d ? m.replace(Je, "") : m).replace(Ye, "$1").replace(Xe, "$1;"), m = "function(" + (_ || "obj") + ") {\n" + (_ ? "" : "obj || (obj = {});\n") + "var __t, __p = ''" + (u ? ", __e = _.escape" : "") + (d ? ", __j = Array.prototype.join;\nfunction print() { __p += __j.call(arguments, '') }\n" : ";\n") + m + "return __p\n}";
				var v = wf(function() {
					return j(o, g + "return " + m).apply(n, l);
				});
				if (v.source = m, Tu(v)) throw v;
				return v;
			}
			function mf(e) {
				return Q(e).toLowerCase();
			}
			function hf(e) {
				return Q(e).toUpperCase();
			}
			function gf(e, t, r) {
				if (e = Q(e), e && (r || t === n)) return rr(e);
				if (!e || !(t = Ua(t))) return e;
				var i = Tr(e), a = Tr(t);
				return to(i, sr(i, a), cr(i, a) + 1).join("");
			}
			function _f(e, t, r) {
				if (e = Q(e), e && (r || t === n)) return e.slice(0, Er(e) + 1);
				if (!e || !(t = Ua(t))) return e;
				var i = Tr(e);
				return to(i, 0, cr(i, Tr(t)) + 1).join("");
			}
			function vf(e, t, r) {
				if (e = Q(e), e && (r || t === n)) return e.replace(ot, "");
				if (!e || !(t = Ua(t))) return e;
				var i = Tr(e);
				return to(i, sr(i, Tr(t))).join("");
			}
			function yf(e, t) {
				var r = ae, i = oe;
				if (Y(t)) {
					var a = "separator" in t ? t.separator : a;
					r = "length" in t ? Z(t.length) : r, i = "omission" in t ? Ua(t.omission) : i;
				}
				e = Q(e);
				var o = e.length;
				if (mr(e)) {
					var s = Tr(e);
					o = s.length;
				}
				if (r >= o) return e;
				var c = r - wr(i);
				if (c < 1) return i;
				var l = s ? to(s, 0, c).join("") : e.slice(0, c);
				if (a === n) return l + i;
				if (s && (c += l.length - c), zu(a)) {
					if (e.slice(c).search(a)) {
						var u, d = l;
						for (a.global || (a = Tt(a.source, Q(mt.exec(a)) + "g")), a.lastIndex = 0; u = a.exec(d);) var f = u.index;
						l = l.slice(0, f === n ? c : f);
					}
				} else if (e.indexOf(Ua(a), c) != c) {
					var p = l.lastIndexOf(a);
					p > -1 && (l = l.slice(0, p));
				}
				return l + i;
			}
			function bf(e) {
				return e = Q(e), e && $e.test(e) ? e.replace(Ze, Dr) : e;
			}
			var xf = wo(function(e, t, n) {
				return e + (n ? " " : "") + t.toUpperCase();
			}), Sf = Co("toUpperCase");
			function Cf(e, t, r) {
				return e = Q(e), t = r ? n : t, t === n ? hr(e) ? Ar(e) : Un(e) : e.match(t) || [];
			}
			var wf = W(function(e, t) {
				try {
					return kn(e, n, t);
				} catch (e) {
					return Tu(e) ? e : new Ct(e);
				}
			}), Tf = Jo(function(e, t) {
				return jn(t, function(t) {
					t = zs(t), Ei(e, t, Bl(e[t], e));
				}), e;
			});
			function Ef(e) {
				var t = e == null ? 0 : e.length, n = G();
				return e = t ? B(e, function(e) {
					if (typeof e[1] != "function") throw new Dt(o);
					return [n(e[0]), e[1]];
				}) : [], W(function(n) {
					for (var r = -1; ++r < t;) {
						var i = e[r];
						if (kn(i[0], this, n)) return kn(i[1], this, n);
					}
				});
			}
			function Df(e) {
				return Ai(ki(e, f));
			}
			function Of(e) {
				return function() {
					return e;
				};
			}
			function kf(e, t) {
				return e == null || e !== e ? t : e;
			}
			var Af = Oo(), jf = Oo(!0);
			function Mf(e) {
				return e;
			}
			function Nf(e) {
				return fa(typeof e == "function" ? e : ki(e, f));
			}
			function Pf(e) {
				return _a(ki(e, f));
			}
			function Ff(e, t) {
				return va(e, ki(t, f));
			}
			var If = W(function(e, t) {
				return function(n) {
					return ea(n, e, t);
				};
			}), Lf = W(function(e, t) {
				return function(n) {
					return ea(e, n, t);
				};
			});
			function Rf(e, t, n) {
				var r = $(t), i = Wi(t, r);
				n == null && !(Y(t) && (i.length || !r.length)) && (n = t, t = e, e = this, i = Wi(t, $(t)));
				var a = !(Y(n) && "chain" in n) || !!n.chain, o = Du(e);
				return jn(i, function(n) {
					var r = t[n];
					e[n] = r, o && (e.prototype[n] = function() {
						var t = this.__chain__;
						if (a || t) {
							var n = e(this.__wrapped__);
							return (n.__actions__ = mo(this.__actions__)).push({
								func: r,
								args: arguments,
								thisArg: e
							}), n.__chain__ = t, n;
						}
						return r.apply(e, Ln([this.value()], arguments));
					});
				}), e;
			}
			function zf() {
				return R._ === this && (R._ = Lt), this;
			}
			function Bf() {}
			function Vf(e) {
				return e = Z(e), W(function(t) {
					return xa(t, e);
				});
			}
			var Hf = Mo(B), Uf = Mo(Nn), Wf = Mo(Bn);
			function Gf(e) {
				return gs(e) ? Xn(zs(e)) : Ta(e);
			}
			function Kf(e) {
				return function(t) {
					return e == null ? n : Gi(e, t);
				};
			}
			var qf = Fo(), Jf = Fo(!0);
			function Yf() {
				return [];
			}
			function Xf() {
				return !1;
			}
			function Zf() {
				return {};
			}
			function Qf() {
				return "";
			}
			function $f() {
				return !0;
			}
			function ep(e, t) {
				if (e = Z(e), e < 1 || e > fe) return [];
				var n = C, r = L(e, C);
				t = G(t), e -= C;
				for (var i = tr(r, t); ++n < e;) t(n);
				return i;
			}
			function tp(e) {
				return q(e) ? B(e, zs) : Uu(e) ? [e] : mo(Rs(Q(e)));
			}
			function np(e) {
				var t = ++Nt;
				return Q(e) + t;
			}
			var rp = jo(function(e, t) {
				return e + t;
			}, 0), ip = Ro("ceil"), ap = jo(function(e, t) {
				return e / t;
			}, 1), op = Ro("floor");
			function sp(e) {
				return e && e.length ? Li(e, Mf, Ji) : n;
			}
			function cp(e, t) {
				return e && e.length ? Li(e, G(t, 2), Ji) : n;
			}
			function lp(e) {
				return Yn(e, Mf);
			}
			function up(e, t) {
				return Yn(e, G(t, 2));
			}
			function dp(e) {
				return e && e.length ? Li(e, Mf, ha) : n;
			}
			function fp(e, t) {
				return e && e.length ? Li(e, G(t, 2), ha) : n;
			}
			var pp = jo(function(e, t) {
				return e * t;
			}, 1), mp = Ro("round"), hp = jo(function(e, t) {
				return e - t;
			}, 0);
			function gp(e) {
				return e && e.length ? er(e, Mf) : 0;
			}
			function _p(e, t) {
				return e && e.length ? er(e, G(t, 2)) : 0;
			}
			return V.after = Ll, V.ary = Rl, V.assign = nd, V.assignIn = rd, V.assignInWith = id, V.assignWith = ad, V.at = od, V.before = zl, V.bind = Bl, V.bindAll = Tf, V.bindKey = Vl, V.castArray = ou, V.chain = $c, V.chunk = Us, V.compact = Ws, V.concat = Gs, V.cond = Ef, V.conforms = Df, V.constant = Of, V.countBy = ul, V.create = sd, V.curry = Hl, V.curryRight = Ul, V.debounce = Wl, V.defaults = cd, V.defaultsDeep = ld, V.defer = Gl, V.delay = Kl, V.difference = Ks, V.differenceBy = qs, V.differenceWith = Js, V.drop = Ys, V.dropRight = Xs, V.dropRightWhile = Zs, V.dropWhile = Qs, V.fill = $s, V.filter = fl, V.flatMap = hl, V.flatMapDeep = gl, V.flatMapDepth = _l, V.flatten = nc, V.flattenDeep = rc, V.flattenDepth = ic, V.flip = ql, V.flow = Af, V.flowRight = jf, V.fromPairs = ac, V.functions = gd, V.functionsIn = _d, V.groupBy = bl, V.initial = cc, V.intersection = lc, V.intersectionBy = uc, V.intersectionWith = dc, V.invert = xd, V.invertBy = Sd, V.invokeMap = Sl, V.iteratee = Nf, V.keyBy = Cl, V.keys = $, V.keysIn = wd, V.map = wl, V.mapKeys = Td, V.mapValues = Ed, V.matches = Pf, V.matchesProperty = Ff, V.memoize = Jl, V.merge = Dd, V.mergeWith = Od, V.method = If, V.methodOf = Lf, V.mixin = Rf, V.negate = Yl, V.nthArg = Vf, V.omit = kd, V.omitBy = Ad, V.once = Xl, V.orderBy = Tl, V.over = Hf, V.overArgs = Zl, V.overEvery = Uf, V.overSome = Wf, V.partial = Ql, V.partialRight = $l, V.partition = El, V.pick = jd, V.pickBy = Md, V.property = Gf, V.propertyOf = Kf, V.pull = gc, V.pullAll = _c, V.pullAllBy = vc, V.pullAllWith = yc, V.pullAt = bc, V.range = qf, V.rangeRight = Jf, V.rearg = eu, V.reject = kl, V.remove = xc, V.rest = tu, V.reverse = Sc, V.sampleSize = jl, V.set = Pd, V.setWith = Fd, V.shuffle = Ml, V.slice = Cc, V.sortBy = Fl, V.sortedUniq = Ac, V.sortedUniqBy = jc, V.split = uf, V.spread = nu, V.tail = Mc, V.take = Nc, V.takeRight = Pc, V.takeRightWhile = Fc, V.takeWhile = Ic, V.tap = el, V.throttle = ru, V.thru = tl, V.toArray = Xu, V.toPairs = Id, V.toPairsIn = Ld, V.toPath = tp, V.toPlainObject = ed, V.transform = Rd, V.unary = iu, V.union = Lc, V.unionBy = Rc, V.unionWith = zc, V.uniq = Bc, V.uniqBy = Vc, V.uniqWith = Hc, V.unset = zd, V.unzip = Uc, V.unzipWith = Wc, V.update = Bd, V.updateWith = Vd, V.values = Hd, V.valuesIn = Ud, V.without = Gc, V.words = Cf, V.wrap = au, V.xor = Kc, V.xorBy = qc, V.xorWith = Jc, V.zip = Yc, V.zipObject = Xc, V.zipObjectDeep = Zc, V.zipWith = Qc, V.entries = Id, V.entriesIn = Ld, V.extend = rd, V.extendWith = id, Rf(V, V), V.add = rp, V.attempt = wf, V.camelCase = qd, V.capitalize = Jd, V.ceil = ip, V.clamp = Wd, V.clone = su, V.cloneDeep = lu, V.cloneDeepWith = uu, V.cloneWith = cu, V.conformsTo = du, V.deburr = Yd, V.defaultTo = kf, V.divide = ap, V.endsWith = Xd, V.eq = fu, V.escape = Zd, V.escapeRegExp = Qd, V.every = dl, V.find = pl, V.findIndex = ec, V.findKey = ud, V.findLast = ml, V.findLastIndex = tc, V.findLastKey = dd, V.floor = op, V.forEach = vl, V.forEachRight = yl, V.forIn = fd, V.forInRight = pd, V.forOwn = md, V.forOwnRight = hd, V.get = vd, V.gt = pu, V.gte = mu, V.has = yd, V.hasIn = bd, V.head = oc, V.identity = Mf, V.includes = xl, V.indexOf = sc, V.inRange = Gd, V.invoke = Cd, V.isArguments = hu, V.isArray = q, V.isArrayBuffer = gu, V.isArrayLike = _u, V.isArrayLikeObject = J, V.isBoolean = vu, V.isBuffer = yu, V.isDate = bu, V.isElement = xu, V.isEmpty = Su, V.isEqual = Cu, V.isEqualWith = wu, V.isError = Tu, V.isFinite = Eu, V.isFunction = Du, V.isInteger = Ou, V.isLength = ku, V.isMap = Au, V.isMatch = ju, V.isMatchWith = Mu, V.isNaN = Nu, V.isNative = Pu, V.isNil = Iu, V.isNull = Fu, V.isNumber = Lu, V.isObject = Y, V.isObjectLike = X, V.isPlainObject = Ru, V.isRegExp = zu, V.isSafeInteger = Bu, V.isSet = Vu, V.isString = Hu, V.isSymbol = Uu, V.isTypedArray = Wu, V.isUndefined = Gu, V.isWeakMap = Ku, V.isWeakSet = qu, V.join = fc, V.kebabCase = $d, V.last = pc, V.lastIndexOf = mc, V.lowerCase = ef, V.lowerFirst = tf, V.lt = Ju, V.lte = Yu, V.max = sp, V.maxBy = cp, V.mean = lp, V.meanBy = up, V.min = dp, V.minBy = fp, V.stubArray = Yf, V.stubFalse = Xf, V.stubObject = Zf, V.stubString = Qf, V.stubTrue = $f, V.multiply = pp, V.nth = hc, V.noConflict = zf, V.noop = Bf, V.now = Il, V.pad = nf, V.padEnd = rf, V.padStart = af, V.parseInt = of, V.random = Kd, V.reduce = Dl, V.reduceRight = Ol, V.repeat = sf, V.replace = cf, V.result = Nd, V.round = mp, V.runInContext = e, V.sample = Al, V.size = Nl, V.snakeCase = lf, V.some = Pl, V.sortedIndex = wc, V.sortedIndexBy = Tc, V.sortedIndexOf = Ec, V.sortedLastIndex = Dc, V.sortedLastIndexBy = Oc, V.sortedLastIndexOf = kc, V.startCase = df, V.startsWith = ff, V.subtract = hp, V.sum = gp, V.sumBy = _p, V.template = pf, V.times = ep, V.toFinite = Zu, V.toInteger = Z, V.toLength = Qu, V.toLower = mf, V.toNumber = $u, V.toSafeInteger = td, V.toString = Q, V.toUpper = hf, V.trim = gf, V.trimEnd = _f, V.trimStart = vf, V.truncate = yf, V.unescape = bf, V.uniqueId = np, V.upperCase = xf, V.upperFirst = Sf, V.each = vl, V.eachRight = yl, V.first = oc, Rf(V, function() {
				var e = {};
				return Hi(V, function(t, n) {
					N.call(V.prototype, n) || (e[n] = t);
				}), e;
			}(), { chain: !1 }), V.VERSION = r, jn([
				"bind",
				"bindKey",
				"curry",
				"curryRight",
				"partial",
				"partialRight"
			], function(e) {
				V[e].placeholder = V;
			}), jn(["drop", "take"], function(e, t) {
				H.prototype[e] = function(r) {
					r = r === n ? 1 : P(Z(r), 0);
					var i = this.__filtered__ && !t ? new H(this) : this.clone();
					return i.__filtered__ ? i.__takeCount__ = L(r, i.__takeCount__) : i.__views__.push({
						size: L(r, C),
						type: e + (i.__dir__ < 0 ? "Right" : "")
					}), i;
				}, H.prototype[e + "Right"] = function(t) {
					return this.reverse()[e](t).reverse();
				};
			}), jn([
				"filter",
				"map",
				"takeWhile"
			], function(e, t) {
				var n = t + 1, r = n == ce || n == ue;
				H.prototype[e] = function(e) {
					var t = this.clone();
					return t.__iteratees__.push({
						iteratee: G(e, 3),
						type: n
					}), t.__filtered__ = t.__filtered__ || r, t;
				};
			}), jn(["head", "last"], function(e, t) {
				var n = "take" + (t ? "Right" : "");
				H.prototype[e] = function() {
					return this[n](1).value()[0];
				};
			}), jn(["initial", "tail"], function(e, t) {
				var n = "drop" + (t ? "" : "Right");
				H.prototype[e] = function() {
					return this.__filtered__ ? new H(this) : this[n](1);
				};
			}), H.prototype.compact = function() {
				return this.filter(Mf);
			}, H.prototype.find = function(e) {
				return this.filter(e).head();
			}, H.prototype.findLast = function(e) {
				return this.reverse().find(e);
			}, H.prototype.invokeMap = W(function(e, t) {
				return typeof e == "function" ? new H(this) : this.map(function(n) {
					return ea(n, e, t);
				});
			}), H.prototype.reject = function(e) {
				return this.filter(Yl(G(e)));
			}, H.prototype.slice = function(e, t) {
				e = Z(e);
				var r = this;
				return r.__filtered__ && (e > 0 || t < 0) ? new H(r) : (e < 0 ? r = r.takeRight(-e) : e && (r = r.drop(e)), t !== n && (t = Z(t), r = t < 0 ? r.dropRight(-t) : r.take(t - e)), r);
			}, H.prototype.takeRightWhile = function(e) {
				return this.reverse().takeWhile(e).reverse();
			}, H.prototype.toArray = function() {
				return this.take(C);
			}, Hi(H.prototype, function(e, t) {
				var r = /^(?:filter|find|map|reject)|While$/.test(t), i = /^(?:head|last)$/.test(t), a = V[i ? "take" + (t == "last" ? "Right" : "") : t], o = i || /^find/.test(t);
				a && (V.prototype[t] = function() {
					var t = this.__wrapped__, s = i ? [1] : arguments, c = t instanceof H, l = s[0], u = c || q(t), d = function(e) {
						var t = a.apply(V, Ln([e], s));
						return i && f ? t[0] : t;
					};
					u && r && typeof l == "function" && l.length != 1 && (c = u = !1);
					var f = this.__chain__, p = !!this.__actions__.length, m = o && !f, h = c && !p;
					if (!o && u) {
						t = h ? t : new H(this);
						var g = e.apply(t, s);
						return g.__actions__.push({
							func: tl,
							args: [d],
							thisArg: n
						}), new zr(g, f);
					}
					return m && h ? e.apply(this, s) : (g = this.thru(d), m ? i ? g.value()[0] : g.value() : g);
				});
			}), jn([
				"pop",
				"push",
				"shift",
				"sort",
				"splice",
				"unshift"
			], function(e) {
				var t = Ot[e], n = /^(?:push|sort|unshift)$/.test(e) ? "tap" : "thru", r = /^(?:pop|shift)$/.test(e);
				V.prototype[e] = function() {
					var e = arguments;
					if (r && !this.__chain__) {
						var i = this.value();
						return t.apply(q(i) ? i : [], e);
					}
					return this[n](function(n) {
						return t.apply(q(n) ? n : [], e);
					});
				};
			}), Hi(H.prototype, function(e, t) {
				var n = V[t];
				if (n) {
					var r = n.name + "";
					N.call(Zn, r) || (Zn[r] = []), Zn[r].push({
						name: t,
						func: n
					});
				}
			}), Zn[ko(n, v).name] = [{
				name: "wrapper",
				func: n
			}], H.prototype.clone = Br, H.prototype.reverse = Vr, H.prototype.value = Hr, V.prototype.at = nl, V.prototype.chain = rl, V.prototype.commit = il, V.prototype.next = al, V.prototype.plant = sl, V.prototype.reverse = cl, V.prototype.toJSON = V.prototype.valueOf = V.prototype.value = ll, V.prototype.first = V.prototype.head, Jt && (V.prototype[Jt] = ol), V;
		})();
		typeof define == "function" && typeof define.amd == "object" && define.amd ? (R._ = jr, define(function() {
			return jr;
		})) : bn ? ((bn.exports = jr)._ = jr, yn._ = jr) : R._ = jr;
	}).call(e);
})))(), ht = (e) => {
	let t = (0, mt.startCase)(e).toLowerCase();
	return t.charAt(0).toUpperCase() + t.slice(1);
}, gt = [
	"src/application.config.ts",
	"src/application-config.ts",
	"src/applicationConfig.ts"
], _t = /serverVariables\s*:\s*\{/, vt = /defineApplication\s*\(\s*\{/, yt = async ({ projectRoot: e, variables: t }) => {
	let n = await bt(e);
	if (n === null) return { status: "skipped-no-config" };
	let r = await Ue(n, "utf8"), i = t.filter((e) => !RegExp(`\\b${e.name}\\s*:`).test(r));
	if (i.length === 0) return { status: "skipped-existing" };
	let a = xt(i);
	if (_t.test(r)) return await We(n, r.replace(_t, (e) => `${e}\n${a}`), "utf8"), {
		status: "appended",
		file: n
	};
	if (!vt.test(r)) return { status: "skipped-no-app-call" };
	let o = r.replace(/\n(\s*)\}\s*\)\s*;?\s*$/, (e, t) => `\n${t}${St(i, t)}\n${t}});\n`);
	return o === r ? { status: "skipped-no-app-call" } : (await We(n, o, "utf8"), {
		status: "created",
		file: n
	});
}, bt = async (e) => {
	for (let t of gt) {
		let n = `${e}/${t}`;
		if (await re(n)) return n;
	}
	return null;
}, xt = (e) => e.map(({ name: e, description: t, isSecret: n }) => `    ${e}: {\n      description: ${JSON.stringify(t)},\n      isSecret: ${n},\n      isRequired: true,\n    },`).join("\n"), St = (e, t) => `serverVariables: {\n${e.map(({ name: e, description: n, isSecret: r }) => `${t}  ${e}: {\n${t}    description: ${JSON.stringify(n)},\n${t}    isSecret: ${r},\n${t}    isRequired: true,\n${t}  },`).join("\n")}\n${t}},`, Ct = ({ data: e }) => {
	let t = E(), n = e.description ? `\n  description: '${e.description}',` : "";
	if (e.type === C.RELATION || e.type === C.MORPH_RELATION) {
		let r = e.onDelete && e.onDelete !== "None", i = `import { defineField, FieldType, RelationType${r ? ", OnDeleteAction" : ""} } from 'twenty-sdk/define';`, a = r ? `, onDelete: OnDeleteAction.${e.onDelete}` : "", o = e.type === C.MORPH_RELATION ? `\n  morphId: '${E()}',` : "";
		return `${i}

export default defineField({
  universalIdentifier: '${t}',
  name: '${e.name}',
  label: '${e.label}',
  type: FieldType.${e.type},
  objectUniversalIdentifier: '${e.objectUniversalIdentifier}',
  relationTargetObjectMetadataUniversalIdentifier: '${e.relationTargetObjectMetadataUniversalIdentifier}',
  relationTargetFieldMetadataUniversalIdentifier: '${e.relationTargetFieldMetadataUniversalIdentifier}',
  universalSettings: { relationType: RelationType.${e.relationType}${a} },${o}${n}
});
`;
	}
	return `import { defineField, FieldType } from 'twenty-sdk/define';

export default defineField({
  universalIdentifier: '${t}',
  name: '${e.name}',
  label: '${e.label}',
  type: FieldType.${e.type},
  objectUniversalIdentifier: '${e.objectUniversalIdentifier}',${n}
});
`;
}, j = (e) => e.replace(/([a-z])([A-Z])/g, "$1-$2").replace(/([A-Z])([A-Z][a-z])/g, "$1-$2").replace(/([a-zA-Z])(\d)/g, "$1-$2").replace(/(\d)([a-zA-Z])/g, "$1-$2").replace(/[\s_]+/g, "-").replace(/[^a-zA-Z0-9-]+/g, "-").toLowerCase(), wt = ({ name: e, universalIdentifier: t = E() }) => {
	let n = j(e);
	return `import { defineFrontComponent } from 'twenty-sdk/define';

// React component - implement your UI here
const Component = () => {
  return (
    <div style={{ padding: '20px', fontFamily: 'sans-serif' }}>
      <h1>My new component!</h1>
      <p>This is your front component: ${n}</p>
    </div>
  );
};

export default defineFrontComponent({
  universalIdentifier: '${t}',
  name: '${n}',
  description: 'Add a description for your front component',
  component: Component,
});
`;
}, M = ({ name: e, universalIdentifier: t = E() }) => {
	let n = j(e);
	return `import { defineLogicFunction } from 'twenty-sdk/define';

// Logic function handler - rename and implement your logic
const handler = async (params: {
  a: string;
  b: number;
}): Promise<{ message: string }> => {
  const { a, b } = params;

  // Replace with your own logic
  const message = \`Hello, input: \${a} and \${b}\`;

  return { message };
};

export default defineLogicFunction({
  universalIdentifier: '${t}',
  name: '${n}',
  description: 'Add a description for your logic function',
  timeoutSeconds: 5,
  handler,
    // Add your trigger here
    // Route trigger example:
    // httpRouteTriggerSettings: {
    //   path: '/${n}',
    //   httpMethod: 'POST',
    //   isAuthRequired: true,
    // },
    // Cron trigger example:
    // cronTriggerSettings: {
    //   pattern: '0 0 * * *', // Daily at midnight
    // },
    // Database event trigger example:
    // databaseEventTriggerSettings: {
    //   eventName: 'objectName.created',
    // },
});
`;
}, Tt = ({ name: e, universalIdentifier: t = E(), type: n, viewUniversalIdentifier: r, targetObjectUniversalIdentifier: i }) => {
	let a = j(e), o;
	return o = n === "OBJECT" && i ? `  type: NavigationMenuItemType.OBJECT,
  targetObjectUniversalIdentifier: '${i}',` : n === "VIEW" && r ? `  type: NavigationMenuItemType.VIEW,
  viewUniversalIdentifier: '${r}',` : n === "LINK" ? "  type: NavigationMenuItemType.LINK,\n  link: 'https://example.com'," : n === "FOLDER" ? "  type: NavigationMenuItemType.FOLDER," : "  type: NavigationMenuItemType.VIEW,\n  // viewUniversalIdentifier: '...',", `import { defineNavigationMenuItem, NavigationMenuItemType } from 'twenty-sdk/define';

export default defineNavigationMenuItem({
  universalIdentifier: '${t}',
  name: '${a}',
  icon: 'IconList',
  position: 0,
${o}
});
`;
}, Et = ({ data: e, universalIdentifier: t = E(), nameFieldUniversalIdentifier: n = E() }) => `import { defineObject, FieldType } from 'twenty-sdk/define';

export const NAME_FIELD_UNIVERSAL_IDENTIFIER =
  '${n}';

export default defineObject({
  universalIdentifier: '${t}',
  nameSingular: '${e.nameSingular}',
  namePlural: '${e.namePlural}',
  labelSingular: '${e.labelSingular}',
  labelPlural: '${e.labelPlural}',
  icon: 'IconBox',
  labelIdentifierFieldMetadataUniversalIdentifier: NAME_FIELD_UNIVERSAL_IDENTIFIER,
  fields: [
    {
      universalIdentifier: NAME_FIELD_UNIVERSAL_IDENTIFIER,
      type: FieldType.TEXT,
      name: 'name',
      label: 'Name',
      description: 'Name of the ${e.nameSingular}',
      icon: 'IconAbc',
    },
  ],
});
`, Dt = ({ name: e }) => `import { defineCommandMenuItem } from 'twenty-sdk/define';

export default defineCommandMenuItem({
  universalIdentifier: '${E()}',
  frontComponentUniversalIdentifier:
    'replace-with-existing-front-component-uuid',
  label: '${e}',
  availabilityType: 'GLOBAL',
});
`, Ot = ({ name: e }) => `import { definePageLayout } from 'twenty-sdk/define';

export default definePageLayout({
  universalIdentifier: '${E()}',
  name: '${e}',
  tabs: [
    {
      universalIdentifier: '${E()}',
      title: 'Overview',
      widgets: [],
    },
  ],
});
`, kt = ({ name: e }) => `import { definePageLayoutTab, PageLayoutTabLayoutMode } from 'twenty-sdk/define';

export default definePageLayoutTab({
  universalIdentifier: '${E()}',
  pageLayoutUniversalIdentifier: 'replace-with-existing-page-layout-uuid',
  title: '${e}',
  position: 1000,
  icon: 'IconLayout',
  layoutMode: PageLayoutTabLayoutMode.CANVAS,
  widgets: [],
});
`, At = ({ objectLabelSingular: e, objectUniversalIdentifier: t, fieldsWidgetViewUniversalIdentifier: n }) => `import { definePageLayout, PageLayoutTabLayoutMode } from 'twenty-sdk/define';

export default definePageLayout({
  universalIdentifier: '${E()}',
  name: 'Default ${e} Layout',
  type: 'RECORD_PAGE',
  objectUniversalIdentifier: '${t}',
  tabs: [
    {
      universalIdentifier: '${E()}',
      title: 'Home',
      position: 10,
      icon: 'IconHome',
      layoutMode: PageLayoutTabLayoutMode.VERTICAL_LIST,
      widgets: [
        {
          universalIdentifier: '${E()}',
          title: 'Fields',
          type: 'FIELDS',
          configuration: {
            configurationType: 'FIELDS',
            viewUniversalIdentifier: '${n}',
          },
        },
      ],
    },
    {
      universalIdentifier: '${E()}',
      title: 'Timeline',
      position: 20,
      icon: 'IconTimelineEvent',
      layoutMode: PageLayoutTabLayoutMode.CANVAS,
      widgets: [
        {
          universalIdentifier: '${E()}',
          title: 'Timeline',
          type: 'TIMELINE',
          configuration: {
            configurationType: 'TIMELINE',
          },
        },
      ],
    },
  ],
});
`, jt = ({ name: e, universalIdentifier: t = E() }) => {
	let n = j(e);
	return `import { defineRole } from 'twenty-sdk/define';

export const ${n.toUpperCase().replace(/-/g, "_")}_ROLE_UNIVERSAL_IDENTIFIER =
  '${t}';

export default defineRole({
  universalIdentifier: ${n.toUpperCase().replace(/-/g, "_")}_ROLE_UNIVERSAL_IDENTIFIER,
  label: '${e}',
  description: 'Add a description for your role',
  canReadAllObjectRecords: true,
  canUpdateAllObjectRecords: true,
  canSoftDeleteAllObjectRecords: true,
  canDestroyAllObjectRecords: false,
});
`;
}, Mt = ({ name: e, universalIdentifier: t = E() }) => {
	let n = j(e);
	return `import { defineAgent } from 'twenty-sdk/define';

export const ${n.toUpperCase().replace(/-/g, "_")}_AGENT_UNIVERSAL_IDENTIFIER =
  '${t}';

export default defineAgent({
  universalIdentifier: ${n.toUpperCase().replace(/-/g, "_")}_AGENT_UNIVERSAL_IDENTIFIER,
  name: '${n}',
  label: '${e}',
  description: 'Add a description for your agent',
  prompt: 'Add the agent system prompt here',
  responseFormat: {type: 'text'},
});
`;
}, N = ({ name: e, universalIdentifier: t = E() }) => {
	let n = j(e), r = n.toUpperCase().replace(/-/g, "_");
	return `import { defineConnectionProvider } from 'twenty-sdk/define';

export const ${r}_CONNECTION_PROVIDER_UNIVERSAL_IDENTIFIER =
  '${t}';

export default defineConnectionProvider({
  universalIdentifier: ${r}_CONNECTION_PROVIDER_UNIVERSAL_IDENTIFIER,
  name: '${n}',
  displayName: '${e.replace(/\\/g, "\\\\").replace(/'/g, "\\'")}',
  type: 'oauth',
  oauth: {
    // Replace with the OAuth provider's endpoints.
    authorizationEndpoint: 'https://example.com/oauth/authorize',
    tokenEndpoint: 'https://example.com/oauth/access_token',
    scopes: [],
    // Names of serverVariables declared on this app's defineApplication.
    clientIdVariable: '${r}_CLIENT_ID',
    clientSecretVariable: '${r}_CLIENT_SECRET',
  },
});
`;
}, Nt = ({ name: e, universalIdentifier: t = E() }) => {
	let n = j(e);
	return `import { defineSkill } from 'twenty-sdk/define';

export const ${n.toUpperCase().replace(/-/g, "_")}_SKILL_UNIVERSAL_IDENTIFIER =
  '${t}';

export default defineSkill({
  universalIdentifier: ${n.toUpperCase().replace(/-/g, "_")}_SKILL_UNIVERSAL_IDENTIFIER,
  name: '${n}',
  label: '${e}',
  description: 'Add a description for your skill',
  content: 'Add the skill content here',
});
`;
}, Pt = ({ field: e, index: t, objectUniversalIdentifier: n }) => {
	let r = e.universalIdentifier ?? E(), i = e.position ?? t, a = e.isVisible ?? !0, o = e.size ?? 200;
	return `  {
    universalIdentifier: '${r}',
${"defaultFieldName" in e ? `    fieldMetadataUniversalIdentifier: generateDefaultFieldUniversalIdentifier({
      objectUniversalIdentifier: '${n}',
      fieldName: '${e.defaultFieldName}',
    })` : `    fieldMetadataUniversalIdentifier: '${e.fieldMetadataUniversalIdentifier}'`},
    position: ${i},
    isVisible: ${a},
    size: ${o},
  }`;
}, Ft = ({ name: e, universalIdentifier: t = E(), objectUniversalIdentifier: n = "fill-later", fields: r = [], type: i }) => {
	let a = j(e), o = r.some((e) => "defaultFieldName" in e), s = r.length > 0 ? `  fields: [
${r.map((e, t) => Pt({
		field: e,
		index: t,
		objectUniversalIdentifier: n
	})).join(",\n")},
  ],` : "  // fields: [\n  //   {\n  //     universalIdentifier: '...',\n  //     fieldMetadataUniversalIdentifier: '...',\n  //     position: 0,\n  //     isVisible: true,\n  //   },\n  // ],", c = i === void 0 ? "" : `  type: '${i}',\n`;
	return `${o ? "import {\n  defineView,\n  generateDefaultFieldUniversalIdentifier,\n} from 'twenty-sdk/define';" : "import { defineView } from 'twenty-sdk/define';"}

export default defineView({
  universalIdentifier: '${t}',
  name: '${a}',
  objectUniversalIdentifier: '${n}',
${c}  icon: 'IconList',
  position: 0,
${s}
  // filters: [
  //   {
  //     universalIdentifier: '...',
  //     fieldMetadataUniversalIdentifier: '...',
  //     operand: 'Contains',
  //     value: '',
  //   },
  // ],
  // sorts: [
  //   {
  //     universalIdentifier: '...',
  //     fieldMetadataUniversalIdentifier: '...',
  //     direction: 'DESC',
  //   },
  // ],
});
`;
}, It = ({ universalIdentifier: e = E() }) => `import {
  defineViewField,
  STANDARD_OBJECT_UNIVERSAL_IDENTIFIERS,
} from 'twenty-sdk/define';

export default defineViewField({
  universalIdentifier: '${e}',
  // The universalIdentifier of the existing view to add this column to
  viewUniversalIdentifier: 'STANDARD_OBJECT_UNIVERSAL_IDENTIFIERS.<fill-later>.views.<fill-later>.universalIdentifier',
  // The universalIdentifier of the field to display in that view
  fieldMetadataUniversalIdentifier: '<fill-later>',
  position: 0,
  isVisible: true,
  size: 150,
});
`, Lt = "src/application.config.ts", Rt = "src", zt = class {
	async execute(e, t) {
		try {
			let n = e ?? await this.getEntity(), r = this.getFolderName(n), i = t ? T(w, t) : T(w, Rt, r);
			await he(i);
			let { name: a, file: o } = await this.getEntityData(n), s = T(i, this.getFileName(a, n));
			if (await re(s)) {
				let { overwrite: e } = await this.handleFileExist();
				if (!e) return;
			}
			await We(s, o), console.log(D.green(`✓ Created ${r}:`), D.cyan(He(w, s))), n === S.Object && await this.promptAndCreateObjectCompanions(a, t), n === S.ConnectionProvider && await this.registerConnectionProviderServerVariables(a);
		} catch (e) {
			console.error(D.red("Add new entity failed:"), e instanceof Error ? e.message : e), process.exit(1);
		}
	}
	async getEntityData(e) {
		switch (e) {
			case S.Object: {
				let e = await this.getObjectData(), t = e.nameSingular, n = E(), r = E();
				return this.lastObjectUniversalIdentifier = n, this.lastNameFieldUniversalIdentifier = r, this.lastObjectLabelSingular = e.labelSingular, {
					name: t,
					file: Et({
						data: e,
						name: t,
						universalIdentifier: n,
						nameFieldUniversalIdentifier: r
					})
				};
			}
			case S.Field: {
				let e = await this.getFieldData(), t = e.name;
				return {
					name: t,
					file: Ct({
						data: e,
						name: t
					})
				};
			}
			case S.LogicFunction: {
				let t = await this.getEntityName(e);
				return {
					name: t,
					file: M({ name: t })
				};
			}
			case S.FrontComponent: {
				let t = await this.getEntityName(e);
				return {
					name: t,
					file: wt({ name: t })
				};
			}
			case S.Role: {
				let t = await this.getEntityName(e);
				return {
					name: t,
					file: jt({ name: t })
				};
			}
			case S.Skill: {
				let t = await this.getEntityName(e);
				return {
					name: t,
					file: Nt({ name: t })
				};
			}
			case S.Agent: {
				let t = await this.getEntityName(e);
				return {
					name: t,
					file: Mt({ name: t })
				};
			}
			case S.ConnectionProvider: {
				let t = await this.getEntityName(e);
				return {
					name: t,
					file: N({ name: t })
				};
			}
			case S.View: {
				let e = (await this.getViewData()).name;
				return {
					name: e,
					file: Ft({ name: e })
				};
			}
			case S.ViewField: return {
				name: await this.getEntityName(e),
				file: It({})
			};
			case S.NavigationMenuItem: {
				let t = await this.getEntityName(e);
				return {
					name: t,
					file: Tt({ name: t })
				};
			}
			case S.PageLayout: {
				let t = await this.getEntityName(e);
				return {
					name: t,
					file: Ot({ name: t })
				};
			}
			case S.PageLayoutTab: {
				let t = await this.getEntityName(e);
				return {
					name: t,
					file: kt({ name: t })
				};
			}
			case S.CommandMenuItem: {
				let t = await this.getEntityName(e);
				return {
					name: t,
					file: Dt({ name: t })
				};
			}
			default: ce(e);
		}
	}
	async promptAndCreateObjectCompanions(e, t) {
		let { createCompanions: n } = await tt.prompt([{
			type: "confirm",
			name: "createCompanions",
			message: "Also create a view, navigation menu item, and record page layout for this object? (recommended)",
			default: !0
		}]);
		if (!n || !this.lastObjectUniversalIdentifier) return;
		let r = E(), i = E(), a = Ft({
			name: `all-${j(e)}`,
			universalIdentifier: r,
			objectUniversalIdentifier: this.lastObjectUniversalIdentifier,
			fields: this.lastNameFieldUniversalIdentifier ? [{
				fieldMetadataUniversalIdentifier: this.lastNameFieldUniversalIdentifier,
				position: 0,
				isVisible: !0,
				size: 200
			}] : []
		}), o = t ? T(w, t) : T(w, Rt, this.getFolderName(S.View));
		await he(o);
		let s = T(o, `all-${j(e)}.ts`);
		if (await re(s)) {
			let { overwrite: e } = await this.handleFileExist();
			if (!e) return;
		}
		await We(s, a), console.log(D.green("✓ Created view:"), D.cyan(He(w, s)));
		let c = this.buildRecordPageFieldsViewFields(this.lastNameFieldUniversalIdentifier), l = Ft({
			name: `${j(e)}-record-page-fields`,
			universalIdentifier: i,
			objectUniversalIdentifier: this.lastObjectUniversalIdentifier,
			type: Ee.FIELDS_WIDGET,
			fields: c
		}), u = T(o, `${j(e)}-record-page-fields.ts`);
		if (await re(u)) {
			let { overwrite: e } = await this.handleFileExist();
			if (!e) return;
		}
		await We(u, l), console.log(D.green("✓ Created record-page-fields view:"), D.cyan(He(w, u)));
		let d = Tt({
			name: e,
			type: "OBJECT",
			targetObjectUniversalIdentifier: this.lastObjectUniversalIdentifier
		}), f = t ? T(w, t) : T(w, Rt, this.getFolderName(S.NavigationMenuItem));
		await he(f);
		let p = T(f, `${j(e)}.ts`);
		if (await re(p)) {
			let { overwrite: e } = await this.handleFileExist();
			if (!e) return;
		}
		await We(p, d), console.log(D.green("✓ Created navigation menu item:"), D.cyan(He(w, p)));
		let m = At({
			objectLabelSingular: this.lastObjectLabelSingular ?? e,
			objectUniversalIdentifier: this.lastObjectUniversalIdentifier,
			fieldsWidgetViewUniversalIdentifier: i
		}), h = t ? T(w, t) : T(w, Rt, this.getFolderName(S.PageLayout));
		await he(h);
		let g = T(h, `${j(e)}-record-page-layout.ts`);
		if (await re(g)) {
			let { overwrite: e } = await this.handleFileExist();
			if (!e) return;
		}
		await We(g, m), console.log(D.green("✓ Created record page layout:"), D.cyan(He(w, g)));
	}
	async registerConnectionProviderServerVariables(e) {
		let t = j(e).toUpperCase().replace(/-/g, "_"), n = [{
			name: `${t}_CLIENT_ID`,
			description: "OAuth client ID issued by the third-party provider. Filled in once by the server admin on the application registration.",
			isSecret: !1
		}, {
			name: `${t}_CLIENT_SECRET`,
			description: "OAuth client secret issued by the third-party provider. Stored encrypted; never echoed in API responses.",
			isSecret: !0
		}], r = await yt({
			projectRoot: w,
			variables: n
		});
		switch (r.status) {
			case "appended":
			case "created":
				console.log(D.green(`✓ Added ${n.map((e) => e.name).join(" + ")} to defineApplication.serverVariables:`), D.cyan(He(w, r.file)));
				break;
			case "skipped-existing":
				console.log(D.dim(`  (${n.map((e) => e.name).join(" / ")} already declared on defineApplication.serverVariables)`));
				break;
			case "skipped-no-config":
			case "skipped-no-app-call":
				console.log(D.yellow(`! Couldn't auto-update ${Lt}. Add these manually to defineApplication.serverVariables:\n` + n.map((e) => `    ${e.name}: { description: '...', isSecret: ${e.isSecret}, isRequired: true },`).join("\n")));
				break;
		}
	}
	buildRecordPageFieldsViewFields(e) {
		let t = [
			"createdAt",
			"updatedAt",
			"createdBy",
			"updatedBy"
		].map((e) => ({
			defaultFieldName: e,
			isVisible: !0,
			size: 200
		}));
		return (e ? [{
			fieldMetadataUniversalIdentifier: e,
			isVisible: !0,
			size: 200
		}, ...t] : t).map((e, t) => ({
			...e,
			position: t
		}));
	}
	async getEntity() {
		let { entity: e } = await tt.prompt([{
			type: "select",
			name: "entity",
			message: "What entity do you want to create?",
			default: "",
			choices: Object.values(S)
		}]);
		return e;
	}
	async handleFileExist() {
		return await tt.prompt([{
			type: "confirm",
			name: "overwrite",
			message: "File already exists. Do you want to overwrite it?",
			default: !1
		}]);
	}
	async getEntityName(e) {
		let { name: t } = await tt.prompt([{
			type: "input",
			name: "name",
			message: `Enter a name for your new ${e}:`,
			default: "",
			validate: (t) => t.length === 0 ? `${e} name is required` : !0
		}]);
		return t;
	}
	async getFieldData() {
		let e = (e) => e === C.RELATION || e === C.MORPH_RELATION;
		return tt.prompt([
			{
				type: "input",
				name: "name",
				message: "Enter a name for your field:",
				default: "",
				validate: (e) => !e || e.trim().length === 0 ? "Please enter a non empty string" : !0
			},
			{
				type: "input",
				name: "label",
				message: "Enter a label for your field:",
				default: (e) => ht(e.name),
				validate: (e) => !e || e.trim().length === 0 ? "Please enter a non empty string" : !0
			},
			{
				type: "select",
				name: "type",
				message: "Select the field type:",
				choices: Object.values(C),
				default: C.TEXT
			},
			{
				type: "input",
				name: "objectUniversalIdentifier",
				message: "Enter the universalIdentifier of the object this field belongs to:",
				default: "fill-later",
				validate: (e) => !e || e.trim().length === 0 ? "Please enter a non empty string" : !0
			},
			{
				type: "input",
				name: "description",
				message: "Enter a description for your field (optional):",
				default: ""
			},
			{
				type: "input",
				name: "relationTargetObjectMetadataUniversalIdentifier",
				message: "Enter the universalIdentifier of the target object of the relation:",
				default: "fill-later",
				when: (t) => e(t.type),
				validate: (e) => !e || e.trim().length === 0 ? "Please enter a non empty string" : !0
			},
			{
				type: "input",
				name: "relationTargetFieldMetadataUniversalIdentifier",
				message: "Enter the universalIdentifier of the target field on the related object:",
				default: "fill-later",
				when: (t) => e(t.type),
				validate: (e) => !e || e.trim().length === 0 ? "Please enter a non empty string" : !0
			},
			{
				type: "select",
				name: "relationType",
				message: "Select the relation type:",
				choices: Object.values(te),
				default: te.ONE_TO_MANY,
				when: (t) => e(t.type)
			},
			{
				type: "select",
				name: "onDelete",
				message: "Select the onDelete behavior:",
				choices: Object.values(pe),
				default: pe.CASCADE,
				when: (t) => e(t.type)
			}
		]);
	}
	async getObjectData() {
		return tt.prompt([
			{
				type: "input",
				name: "nameSingular",
				message: "Enter a name singular for your object (eg: company):",
				default: "",
				validate: (e) => !e || e.trim().length === 0 ? "Please enter a non empty string" : !0
			},
			{
				type: "input",
				name: "namePlural",
				message: "Enter a name plural for your object (eg: companies):",
				default: "",
				validate: (e, t) => e.trim() === t?.nameSingular.trim() ? "Name plural must be different from name singular" : !e || e.trim().length === 0 ? "Please enter a non empty string" : !0
			},
			{
				type: "input",
				name: "labelSingular",
				message: "Enter a label singular for your object:",
				default: (e) => ht(e.nameSingular),
				validate: (e) => !e || e.trim().length === 0 ? "Please enter a non empty string" : !0
			},
			{
				type: "input",
				name: "labelPlural",
				message: "Enter a label plural for your object:",
				default: (e) => ht(e.namePlural),
				validate: (e) => !e || e.trim().length === 0 ? "Please enter a non empty string" : !0
			}
		]);
	}
	async getViewData() {
		return tt.prompt([{
			type: "input",
			name: "name",
			message: "Enter a name for your view:",
			default: "",
			validate: (e) => !e || e.trim().length === 0 ? "Please enter a non empty string" : !0
		}, {
			type: "input",
			name: "objectUniversalIdentifier",
			message: "Enter the universalIdentifier of the object this view belongs to:",
			default: "fill-later",
			validate: (e) => !e || e.trim().length === 0 ? "Please enter a non empty string" : !0
		}]);
	}
	getFolderName(e) {
		return `${j(e)}s`;
	}
	getFileName(e, t) {
		switch (t) {
			case S.FrontComponent: return `${j(e)}.tsx`;
			default: return `${j(e)}.ts`;
		}
	}
}, Bt = class {
	async execute(e) {
		let t = e.appPath ?? w;
		await lt(t), console.log(D.blue("Building application...")), console.log(D.gray(`App path: ${t}\n`));
		let n = await g({
			appPath: t,
			tarball: e.tarball,
			onProgress: (e) => console.log(D.gray(e))
		});
		n.success || (console.error(D.red(n.error.message)), process.exit(1)), console.log(D.green(`✓ Build succeeded (${n.data.fileCount} file${n.data.fileCount === 1 ? "" : "s"})`)), console.log(D.gray(`Output: ${n.data.outputDir}`)), n.data.tarballPath && console.log(D.gray(`Tarball: ${n.data.tarballPath}`));
	}
}, Vt = class {
	async execute({ appPath: e = w, postInstall: t = !1, preInstall: n = !1, functionUniversalIdentifier: r, functionName: i, payload: a = "{}" }) {
		let o;
		try {
			o = JSON.parse(a);
		} catch {
			console.error(D.red("Invalid JSON payload. Please provide valid JSON.")), process.exit(1);
		}
		let s = t ? "post install" : n ? "pre install" : r ?? i, l = _.getActiveRemote();
		console.log(D.blue(`🚀 Executing function "${s}" on ${l}...`)), console.log(D.gray(`   Payload: ${JSON.stringify(o)}\n`));
		let u = await me(t ? {
			appPath: e,
			postInstall: !0,
			payload: o
		} : n ? {
			appPath: e,
			preInstall: !0,
			payload: o
		} : r ? {
			appPath: e,
			functionUniversalIdentifier: r,
			payload: o
		} : {
			appPath: e,
			functionName: i,
			payload: o
		});
		if (!u.success) {
			switch (u.error.code) {
				case v.MANIFEST_NOT_FOUND:
					console.error(D.red("Failed to build manifest."));
					break;
				case c.FETCH_FUNCTIONS_FAILED:
					console.error(D.red("Failed to fetch functions:"), u.error.message);
					break;
				case c.FUNCTION_NOT_FOUND: {
					console.error(D.red(u.error.message), "\n");
					let e = u.error.details?.availableFunctions ?? [];
					e.length > 0 ? (console.log(D.cyan("Available functions:")), e.forEach((e) => {
						console.log(`  - ${D.white(e.name)} (${e.universalIdentifier})`);
					})) : console.log(D.yellow("No functions found for this application. Have you synced your app with `twenty dev`?"));
					break;
				}
				case c.EXECUTION_FAILED:
					console.error(D.red("Execution failed:"), u.error.message);
					break;
				default: console.error(D.red(u.error.message));
			}
			process.exit(1);
		}
		let d = u.data;
		console.log(D.cyan("─".repeat(60))), console.log(D.cyan("Execution Result")), console.log(D.cyan("─".repeat(60)));
		let f = d.status === "SUCCESS" ? D.green : D.red;
		console.log(`${D.bold("Status:")} ${f(d.status)}`), console.log(`${D.bold("Duration:")} ${d.duration}ms\n`), Fe(d.data) && (console.log(D.bold("Data:")), console.log(D.white(JSON.stringify(d.data, null, 2)))), d.error && (console.log(D.bold.red("Error:")), console.log(D.red(`  Type: ${d.error.errorType}`)), console.log(D.red(`  Message: ${d.error.errorMessage}\n`)), d.error.stackTrace && (console.log(D.gray("Stack trace:")), console.log(D.gray(d.error.stackTrace)))), d.logs && (console.log(D.bold("Logs:")), console.log(D.gray(d.logs))), console.log(D.cyan("─".repeat(60))), d.status !== "SUCCESS" && process.exit(1);
	}
}, Ht = class {
	constructor() {
		this.apiService = new l();
	}
	async execute({ appPath: e = w, functionUniversalIdentifier: t, functionName: n }) {
		try {
			let r = await d(e);
			r || process.exit(1), this.logWatchInfo({
				appName: r.application.displayName,
				functionUniversalIdentifier: t,
				functionName: n
			}), await this.apiService.subscribeToLogs({
				applicationUniversalIdentifier: r.application.universalIdentifier,
				functionUniversalIdentifier: t,
				functionName: n
			});
		} catch (e) {
			console.error(D.red("Watch logs failed:"), e instanceof Error ? e.message : e), process.exit(1);
		}
	}
	logWatchInfo({ appName: e, functionUniversalIdentifier: t, functionName: n }) {
		let r = e ?? "Twenty Application", i = t || n ? `function "${t || n}"` : "functions", a = _.getActiveRemote();
		console.log(D.blue(`🚀 Watching ${r} ${i} logs on ${a}:\n`));
	}
}, Ut = (e) => `${D.cyan(e.file)}:${D.yellow(String(e.line))}:${D.yellow(String(e.column + 1))} - ${D.red("error")} ${e.text}`, Wt = class {
	async execute(e) {
		let t = e.appPath ?? w;
		console.log(D.blue("Running type check...")), console.log(D.gray(`App path: ${t}\n`));
		let n = await je(t);
		n.length === 0 && (console.log(D.green("✓ No type errors found")), process.exit(0));
		for (let e of n) console.log(Ut(e));
		console.log("\n", D.red(`✗ Found ${n.length} type error${n.length === 1 ? "" : "s"}`)), process.exit(1);
	}
}, Gt = (e, t) => console.warn(D.yellow(`⚠ \`twenty ${e}\` is deprecated. Use \`twenty ${t}\` instead.`)), Kt = (e) => {
	let t = new Bt(), n = new Wt(), r = new Ht(), i = new Vt(), a = new zt(), o = new ut(), s = new dt(), c = new ft();
	e.command("build [appPath]", { hidden: !0 }).option("--tarball", "Also pack into a .tgz tarball").action(async (e, n) => {
		Gt("build", "dev:build"), await t.execute({
			appPath: A(e),
			tarball: n.tarball
		});
	}), e.command("typecheck [appPath]", { hidden: !0 }).action(async (e) => {
		Gt("typecheck", "dev:typecheck"), await n.execute({ appPath: A(e) });
	}), e.command("logs [appPath]", { hidden: !0 }).option("-u, --functionUniversalIdentifier <functionUniversalIdentifier>", "Only show logs for the function with this universal ID").option("-n, --functionName <functionName>", "Only show logs for the function with this name").action(async (e, t) => {
		Gt("logs", "dev:function:logs"), await r.execute({
			...t,
			appPath: A(e)
		});
	}), e.command("exec [appPath]", { hidden: !0 }).option("--postInstall", "Execute post-install logic function if defined").option("--preInstall", "Execute pre-install logic function if defined").option("-p, --payload <payload>", "JSON payload to send to the function", "{}").option("-u, --functionUniversalIdentifier <functionUniversalIdentifier>", "Universal ID of the function to execute").option("-n, --functionName <functionName>", "Name of the function to execute").action(async (e, t) => {
		Gt("exec", "dev:function:exec"), !t?.postInstall && !t?.preInstall && !t?.functionUniversalIdentifier && !t?.functionName && (console.error(D.red("Error: Either --postInstall, --preInstall, --functionName (-n), or --functionUniversalIdentifier (-u) is required.")), process.exit(1)), await i.execute({
			...t,
			payload: t?.payload ?? "{}",
			appPath: A(e)
		});
	}), e.command("add [entityType]", { hidden: !0 }).option("--path <path>", "Path in which the entity should be created.").action(async (e, t) => {
		Gt("add", "dev:add"), await a.execute(e, t?.path);
	}), e.command("publish [appPath]", { hidden: !0 }).option("--tag <tag>", "npm dist-tag (e.g. beta, next)").action(async (e, t) => {
		Gt("publish", "app:publish"), await o.execute({
			appPath: A(e),
			tag: t.tag
		});
	}), e.command("deploy [appPath]", { hidden: !0 }).option("-r, --remote <name>", "Deploy to a specific remote").action(async (e, t) => {
		Gt("deploy", "app:publish --private"), await o.execute({
			appPath: A(e),
			private: !0,
			remote: t.remote
		});
	}), e.command("install [appPath]", { hidden: !0 }).option("-r, --remote <name>", "Install on a specific remote").action(async (e, t) => {
		Gt("install", "app:install"), await s.execute({
			appPath: A(e),
			remote: t.remote
		});
	}), e.command("uninstall [appPath]", { hidden: !0 }).option("-y, --yes", "Skip confirmation prompt").action(async (e, t) => {
		Gt("uninstall", "app:uninstall");
		try {
			let n = await c.execute({
				appPath: A(e),
				askForConfirmation: !t?.yes
			});
			process.exit(+!n.success);
		} catch {
			process.exit(1);
		}
	}), e.command("catalog-sync", { hidden: !0 }).option("-r, --remote <name>", "Sync on a specific remote").action(async (e) => {
		Gt("catalog-sync", "dev:catalog-sync");
		let { CatalogSyncCommand: t } = await import("./catalog-sync-BYM3GqZl.mjs");
		await new t().execute({ remote: e.remote });
	});
}, qt = class {
	constructor({ state: e, notify: t }) {
		this.state = e, this.notify = t;
	}
	async execute(e) {
		let t = this.state.steps.buildManifest;
		t.status = "in_progress", this.state.updatePipeline({ status: "building" });
		let n = [{
			message: "Building manifest",
			status: "info"
		}], r = await x(e.appPath);
		if (!r.success) {
			for (let e of r.errors) n.push({
				message: e,
				status: "error"
			});
			return t.output = { result: null }, t.status = "error", this.state.updatePipeline({ status: "error" }), this.state.applyStepEvents(n), null;
		}
		if (r.warnings.length > 0) for (let e of r.warnings) n.push({
			message: `⚠ ${e}`,
			status: "warning"
		});
		n.push({
			message: "Successfully built manifest",
			status: "success"
		});
		let i = {
			manifest: r.manifest,
			filePaths: r.filePaths
		};
		return t.output = { result: i }, t.status = "done", this.state.updatePipeline({ appName: r.manifest.application.displayName }), this.state.updateEntitiesFromManifest(r.filePaths), this.state.applyStepEvents(n), i;
	}
}, Jt = class {
	constructor({ apiService: e, state: t, notify: n }) {
		this.hasRetried = !1, this.apiService = e, this.state = t, this.notify = n;
	}
	async execute() {
		let e = this.state.steps.checkServer, t = await this.apiService.validateAuth();
		if (!t.serverUp) {
			let t = await Te();
			return t && !this.hasRetried ? (this.hasRetried = !0, await new _().setConfig({ apiUrl: t }), this.execute()) : (e.output.errorLogged || (e.output = {
				isReady: !1,
				errorLogged: !0
			}, e.status = "error", this.state.applyStepEvents([{
				message: "Cannot reach Twenty server.\n\n  Start a local server:\n    yarn twenty docker:start\n\n  Check server status:\n    yarn twenty docker:status\n\n  Waiting for server...",
				status: "error"
			}]), this.state.updatePipeline({ status: "error" })), !1);
		}
		if (!t.authValid) return e.output.errorLogged || (e.output = {
			isReady: !1,
			errorLogged: !0
		}, e.status = "error", this.state.applyStepEvents([{
			message: "Authentication failed. Run `yarn twenty remote:add` to authenticate.",
			status: "error"
		}]), this.state.updatePipeline({ status: "error" })), !1;
		let n = e.output.isReady;
		if (e.output = {
			isReady: !0,
			errorLogged: !1
		}, e.status = "done", !Fe(this.state.frontendUrl)) {
			let e = await this.apiService.getWorkspaceFrontendUrl();
			Fe(e) && (this.state.frontendUrl = e);
		}
		return n || this.notify(), !0;
	}
}, Yt = class {
	constructor({ clientService: e, configService: t, state: n, notify: r }) {
		this.clientService = e, this.configService = t, this.state = n, this.notify = r;
	}
	async execute(e) {
		let t = this.state.steps.generateApiClient;
		t.status = "in_progress", this.notify();
		try {
			let n = await se(this.configService, e.credentials);
			await this.clientService.generateCoreClient({
				appPath: e.appPath,
				appAccessToken: n
			}), t.status = "done";
		} catch (e) {
			this.state.applyStepEvents([{
				message: `Failed to generate API client: ${e instanceof Error ? e.message : String(e)}`,
				status: "error"
			}]), t.status = "error";
		}
		this.notify();
	}
}, Xt = class {
	constructor({ apiService: e, configService: t, state: n, notify: r }) {
		this.apiService = e, this.configService = t, this.state = n, this.notify = r;
	}
	async execute(e) {
		try {
			let t = await Me(this.apiService, this.configService, {
				name: e.manifest.application.displayName,
				universalIdentifier: e.manifest.application.universalIdentifier
			});
			this.registrationCredentials = {
				clientId: t.clientId,
				clientSecret: t.clientSecret
			}, this.state.applyStepEvents([{
				message: t.isNewRegistration ? `App registration created: ${e.manifest.application.displayName}` : "Existing app registration found",
				status: t.isNewRegistration ? "success" : "info"
			}, ...t.isNewRegistration ? [{
				message: "Credentials saved to config.",
				status: "info"
			}] : []]);
		} catch (e) {
			throw this.state.applyStepEvents([{
				message: `Failed to register app: ${e instanceof Error ? e.message : String(e)}`,
				status: "error"
			}]), e;
		}
		this.notify();
	}
}, Zt = class {
	constructor(e) {
		this.watcher = null, this.appPath = e.appPath, this.fileFolder = e.fileFolder, this.watchPaths = e.watchPaths, this.handleFileBuilt = e.handleFileBuilt;
	}
	async start() {
		let e = (await Promise.all(this.watchPaths.map(async (e) => {
			let t = T(this.appPath, e);
			return await re(t) ? t : null;
		}))).filter((e) => e !== null);
		e.length !== 0 && (this.watcher = nt.watch(e, { awaitWriteFinish: {
			stabilityThreshold: 100,
			pollInterval: 50
		} }), this.watcher.on("all", (e, t) => {
			e === "addDir" || e === "unlinkDir" || ((e === "add" || e === "change") && this.copyAndNotify(t), e === "unlink" && this.handleUnlink(t));
		}));
	}
	async close() {
		await this.watcher?.close();
	}
	async copyAndNotify(e) {
		let t = He(this.appPath, e), n = T(ve, t), r = T(this.appPath, n);
		await he(Ve(r)), await Se(e, r);
		let i = await Ue(r), a = Ze.createHash("md5").update(i).digest("hex");
		this.handleFileBuilt({
			fileFolder: this.fileFolder,
			builtPath: n,
			sourcePath: t,
			checksum: a
		});
	}
	async handleUnlink(e) {
		let t = T(ve, He(this.appPath, e));
		await ke(T(this.appPath, t));
	}
}, Qt = class {
	constructor(e) {
		this.process = null, this.pendingErrors = [], this.buffer = "", this.hasErrors = !1, this.appPath = e.appPath, this.onErrors = e.onErrors;
	}
	async start() {
		let e = Xe.join(this.appPath, "node_modules", ".bin", "tsc");
		if (!await re(e)) return;
		let t = Xe.join(this.appPath, "tsconfig.json");
		this.process = $e(e, [
			"--watch",
			"--noEmit",
			"--pretty",
			"false",
			"-p",
			t
		], {
			cwd: this.appPath,
			stdio: [
				"ignore",
				"pipe",
				"pipe"
			]
		}), this.process.on("error", () => {
			this.process = null;
		}), this.process.stdout?.on("data", (e) => {
			this.handleOutput(e.toString());
		}), this.process.stderr?.on("data", (e) => {
			this.handleOutput(e.toString());
		});
	}
	close() {
		this.process?.kill(), this.process = null;
	}
	handleOutput(e) {
		this.buffer += e;
		let t = this.buffer.split("\n");
		this.buffer = t.pop() ?? "";
		for (let e of t) this.processLine(e);
	}
	processLine(e) {
		if (e.includes("Starting compilation in watch mode...") || e.includes("Starting incremental compilation...")) {
			this.pendingErrors = [];
			return;
		}
		if (e.includes("Watching for file changes.")) {
			let e = this.hasErrors;
			this.hasErrors = this.pendingErrors.length > 0, (this.hasErrors || e) && this.onErrors(this.pendingErrors), this.pendingErrors = [];
			return;
		}
		let t = le(e);
		t && this.pendingErrors.push(t);
	}
}, $t = new Set([
	"node_modules",
	"dist",
	".twenty"
]), en = class {
	constructor(e) {
		this.watcher = null, this.appPath = e.appPath, this.handleChangeDetected = e.handleChangeDetected, this.verbose = e.verbose ?? !1;
	}
	async start() {
		let e = this.appPath;
		this.watcher = nt.watch(this.appPath, {
			ignoreInitial: !this.verbose,
			ignored: (t) => {
				let n = He(e, t);
				if (n === "") return !1;
				let r = n.split(Be.sep)[0];
				return $t.has(r) || r.startsWith(".");
			},
			awaitWriteFinish: {
				stabilityThreshold: 100,
				pollInterval: 50
			}
		}), this.watcher.on("all", (e, t) => {
			if (e === "addDir" || e === "unlinkDir") return;
			let n = He(this.appPath, t), r = n.startsWith(Ne), i = ["package.json", "yarn.lock"].includes(n);
			!(n.endsWith(".ts") || n.endsWith(".tsx")) && !r && !i || this.handleChangeDetected(n, e);
		});
	}
	async close() {
		await this.watcher?.close();
	}
}, tn = class {
	constructor(e) {
		this.manifestWatcher = null, this.logicFunctionsWatcher = null, this.frontComponentsWatcher = null, this.assetWatcher = null, this.dependencyWatcher = null, this.tscWatcher = null, this.state = e.state, this.scheduleSync = e.scheduleSync, this.notify = e.notify, this.onFileBuilt = e.onFileBuilt, this.shouldSkipTypecheck = e.shouldSkipTypecheck, this.verbose = e.verbose ?? !1;
	}
	async start() {
		this.state.steps.startWatchers.status = "in_progress", this.notify(), this.manifestWatcher = new en({
			appPath: this.state.appPath,
			handleChangeDetected: this.handleChangeDetected.bind(this),
			verbose: this.verbose
		}), await this.manifestWatcher.start();
	}
	async handleWatcherRestarts(e) {
		let { logicFunctions: t, frontComponents: n } = e.filePaths;
		if (!this.state.steps.startWatchers.output.watchersStarted) {
			this.state.steps.startWatchers.output.watchersStarted = !0, this.state.steps.startWatchers.status = "done", await this.startFileWatchers(t, n);
			return;
		}
		this.logicFunctionsWatcher?.shouldRestart(t) && await this.logicFunctionsWatcher.restart(t), this.frontComponentsWatcher?.shouldRestart(n) && await this.frontComponentsWatcher.restart(n);
	}
	async close() {
		this.tscWatcher?.close(), await Promise.all([
			this.manifestWatcher?.close(),
			this.logicFunctionsWatcher?.close(),
			this.frontComponentsWatcher?.close(),
			this.assetWatcher?.close(),
			this.dependencyWatcher?.close()
		]);
	}
	handleChangeDetected(e, t) {
		this.state.addEvent({
			message: `Change detected: ${e}`,
			status: "info",
			spacingBefore: !0
		}), t === "unlink" ? this.state.removeEntity(e) : this.state.updateEntityStatus(e, "building"), this.notify(), this.scheduleSync();
	}
	handleFileBuildError(e) {
		this.state.addEvent({
			message: "Build failed:",
			status: "error"
		});
		for (let t of e) this.state.addEvent({
			message: t.error,
			status: "error"
		});
		this.notify();
	}
	handleFileBuilt(e) {
		this.verbose && this.state.addEvent({
			message: `Successfully built ${e.builtPath}`,
			status: "success"
		}), this.state.steps.uploadFiles.output.builtFileInfos.set(e.builtPath, {
			checksum: e.checksum,
			builtPath: e.builtPath,
			sourcePath: e.sourcePath,
			fileFolder: e.fileFolder,
			usesSdkClient: e.usesSdkClient
		}), this.onFileBuilt(e), this.notify(), this.scheduleSync();
	}
	async startFileWatchers(e, t) {
		await Promise.all([
			this.startTscWatcher(),
			this.startLogicFunctionsWatcher(e),
			this.startFrontComponentsWatcher(t),
			this.startAssetWatcher(),
			this.startDependencyWatcher()
		]);
	}
	async startLogicFunctionsWatcher(t) {
		this.logicFunctionsWatcher = e({
			appPath: this.state.appPath,
			sourcePaths: t,
			shouldSkipTypecheck: this.shouldSkipTypecheck,
			handleBuildError: this.handleFileBuildError.bind(this),
			handleFileBuilt: this.handleFileBuilt.bind(this)
		}), await this.logicFunctionsWatcher.start();
	}
	async startFrontComponentsWatcher(e) {
		this.frontComponentsWatcher = p({
			appPath: this.state.appPath,
			sourcePaths: e,
			shouldSkipTypecheck: this.shouldSkipTypecheck,
			handleBuildError: this.handleFileBuildError.bind(this),
			handleFileBuilt: this.handleFileBuilt.bind(this)
		}), await this.frontComponentsWatcher.start();
	}
	async startAssetWatcher() {
		this.assetWatcher = new Zt({
			appPath: this.state.appPath,
			fileFolder: be.PublicAsset,
			watchPaths: [Ne],
			handleFileBuilt: this.handleFileBuilt.bind(this)
		}), await this.assetWatcher.start();
	}
	async startDependencyWatcher() {
		this.dependencyWatcher = new Zt({
			appPath: this.state.appPath,
			fileFolder: be.Dependencies,
			watchPaths: ["package.json", "yarn.lock"],
			handleFileBuilt: this.handleFileBuilt.bind(this)
		}), this.dependencyWatcher.start();
	}
	async startTscWatcher() {
		this.tscWatcher = new Qt({
			appPath: this.state.appPath,
			onErrors: this.handleTypecheckErrors.bind(this)
		}), await this.tscWatcher.start();
	}
	handleTypecheckErrors(e) {
		e.length === 0 ? this.state.addEvent({
			message: "Typecheck passed",
			status: "success"
		}) : this.state.applyStepEvents(e.map((e) => ({
			message: `Type error in ${e.file}(${e.line},${e.column}): ${e.text}`,
			status: "error"
		}))), this.notify();
	}
}, nn = class {
	constructor({ apiService: e, state: t, notify: n, verbose: r }) {
		this.apiService = e, this.state = t, this.notify = n, this.verbose = r ?? !1;
	}
	async execute(e) {
		let t = this.state.steps.syncApplication;
		t.status = "in_progress", this.state.updatePipeline({ status: "syncing" });
		let n = [], r = y({
			manifest: e.manifest,
			builtFileInfos: e.builtFileInfos
		});
		n.push({
			message: "Manifest checksums set",
			status: "info"
		}), await b(e.appPath, r), n.push({
			message: "Manifest saved to output directory",
			status: "info"
		}), n.push({
			message: "Syncing manifest",
			status: "info"
		});
		let i = await this.apiService.syncApplication(r);
		if (i.success) {
			let e = i.data;
			n.push(..._e(e.actions)), n.push({
				message: "✓ Synced",
				status: "success"
			}), t.output = {
				syncStatus: "synced",
				error: null
			}, t.status = "done", this.state.updatePipeline({
				status: "synced",
				error: null
			}), this.state.updateAllEntitiesStatus("success"), this.state.applyStepEvents(n);
			return;
		}
		let a = this.verbose ? null : u(i.error);
		a ? (n.push(...a), n.push({
			message: "Add --verbose to see full error log",
			status: "info"
		})) : n.push({
			message: `Sync failed with error: ${i.message ?? "Sync failed"}`,
			status: "error"
		});
		let o = oe(i.message);
		o && n.push({
			message: o,
			status: "info"
		});
		let s = a ? a[0].message : "Sync failed";
		t.output = {
			syncStatus: "error",
			error: s
		}, t.status = "error", this.state.updatePipeline({
			status: "error",
			error: s
		}), this.state.updateAllEntitiesStatus("error"), this.state.applyStepEvents(n);
	}
}, rn = class {
	constructor({ state: e, notify: t, verbose: n }) {
		this.uploadedCount = 0, this.failedCount = 0, this.totalQueued = 0, this.state = e, this.notify = t, this.verbose = n ?? !1;
	}
	get isInitialized() {
		return Fe(this.state.steps.uploadFiles.output.fileUploader);
	}
	initialize(e) {
		let t = this.state.steps.uploadFiles;
		t.output = {
			...t.output,
			fileUploader: new Oe({
				appPath: e.appPath,
				applicationUniversalIdentifier: e.universalIdentifier
			})
		}, t.status = "in_progress", this.notify(), this.uploadPendingFiles();
	}
	uploadFile(e, t, n) {
		let r = this.state.steps.uploadFiles;
		if (!r.output.fileUploader) return;
		r.status = "in_progress", this.totalQueued++, this.verbose && this.state.addEvent({
			message: `Uploading ${e}`,
			status: "info"
		}), this.state.updateEntityStatus(t, "uploading"), this.notify();
		let i = r.output.fileUploader.uploadFile({
			builtPath: e,
			fileFolder: n
		}).then((n) => {
			n.success ? (this.uploadedCount++, this.verbose && this.state.addEvent({
				message: `Successfully uploaded ${e}`,
				status: "success"
			}), this.state.updateEntityStatus(t, "success")) : (this.failedCount++, this.state.addEvent({
				message: `Failed to upload ${e}: ${n.error}`,
				status: "error"
			}));
		}).catch((t) => {
			this.failedCount++, this.state.addEvent({
				message: `Upload failed for ${e}: ${t}`,
				status: "error"
			});
		}).finally(() => {
			r.output.activeUploads.delete(i), r.output.activeUploads.size === 0 && (this.logUploadSummary(), r.status = "done", this.notify());
		});
		r.output.activeUploads.add(i);
	}
	async waitForUploads() {
		let e = this.state.steps.uploadFiles;
		for (; e.output.activeUploads.size > 0;) await Promise.all(e.output.activeUploads);
		e.status = "done", this.notify();
	}
	logUploadSummary() {
		if (this.totalQueued === 0) {
			this.resetCounters();
			return;
		}
		this.failedCount > 0 && this.state.addEvent({
			message: `Uploaded ${this.uploadedCount}/${this.totalQueued} files (${this.failedCount} failed)`,
			status: "error"
		}), this.state.addEvent({
			message: `Successfully uploaded ${this.uploadedCount} file${this.uploadedCount === 1 ? "" : "s"}`,
			status: "success"
		}), this.resetCounters();
	}
	resetCounters() {
		this.uploadedCount = 0, this.failedCount = 0, this.totalQueued = 0;
	}
	uploadPendingFiles() {
		for (let [e, { fileFolder: t, sourcePath: n }] of this.state.steps.uploadFiles.output.builtFileInfos.entries()) this.uploadFile(e, n, t);
	}
}, an = class {
	constructor(e) {
		this.syncTimer = null, this.serverCheckInterval = null, this.skipTypecheck = !0, this.debounceMs = e.debounceMs ?? 1e3, this.state = e.state, this.verbose = e.verbose ?? !1, this.apiService = new l({ disableInterceptors: !0 });
		let t = this.apiService, n = new _();
		this.clientService = new ye();
		let r = {
			state: this.state,
			notify: () => this.state.notify()
		};
		this.checkServerStep = new Jt({
			...r,
			apiService: t
		}), this.buildManifestStep = new qt(r), this.registerAppStep = new Xt({
			...r,
			apiService: t,
			configService: n
		}), this.uploadFilesStep = new rn({
			...r,
			verbose: this.verbose
		}), this.generateApiClientStep = new Yt({
			...r,
			clientService: this.clientService,
			configService: n
		}), this.syncApplicationStep = new nn({
			...r,
			apiService: t,
			verbose: this.verbose
		}), this.startWatchersStep = new tn({
			...r,
			scheduleSync: this.scheduleSync.bind(this),
			onFileBuilt: this.handleFileBuilt.bind(this),
			shouldSkipTypecheck: () => this.skipTypecheck,
			verbose: this.verbose
		});
	}
	async start() {
		let e = Be.join(this.state.appPath, ve);
		await he(e), await De(e), this.state.addEvent({
			message: `Using remote "${_.getActiveRemote()}"`,
			status: "info"
		}), this.verbose || this.state.addEvent({
			message: "Add --verbose to see fully detailed logs",
			status: "info"
		}), this.state.notify(), await this.startWatchersStep.start(), this.serverCheckInterval = setInterval(() => {
			this.checkServerHealth();
		}, 2e3);
	}
	async close() {
		this.serverCheckInterval && clearInterval(this.serverCheckInterval), await this.startWatchersStep.close();
	}
	getState() {
		return this.state;
	}
	handleFileBuilt(e) {
		this.state.steps.uploadFiles.output.fileUploader && this.uploadFilesStep.uploadFile(e.builtPath, e.sourcePath, e.fileFolder);
	}
	async checkServerHealth() {
		let e = this.state.steps.checkServer.output.isReady;
		await this.checkServerStep.execute() && !e && this.scheduleSync();
	}
	scheduleSync() {
		this.syncTimer && clearTimeout(this.syncTimer), this.syncTimer = setTimeout(() => {
			this.syncTimer = null, this.performSync();
		}, this.debounceMs);
	}
	async performSync() {
		if (!this.state.pipeline.isSyncing) {
			this.state.updatePipeline({ isSyncing: !0 });
			try {
				await this.runSyncPipeline();
			} catch (e) {
				this.state.addEvent({
					message: `Sync failed with error: ${m(e)}`,
					status: "error"
				}), this.state.updatePipeline({ status: "error" }), this.state.updateAllEntitiesStatus("error");
			} finally {
				this.state.updatePipeline({ isSyncing: !1 });
			}
		}
	}
	async runSyncPipeline() {
		if (!await this.checkServerStep.execute()) return;
		this.state.steps.ensureValidTokens.status = "done";
		let e = await this.buildManifestStep.execute({ appPath: this.state.appPath });
		if (!e || (await this.startWatchersStep.handleWatcherRestarts(e), !this.uploadFilesStep.isInitialized && !await this.initializePipeline(e.manifest))) return;
		let t = this.state.hasObjectsOrFieldsChanged(e.manifest);
		await this.uploadFilesStep.waitForUploads(), await this.syncApplicationStep.execute({
			manifest: e.manifest,
			builtFileInfos: this.state.steps.uploadFiles.output.builtFileInfos,
			appPath: this.state.appPath
		}), this.state.steps.syncApplication.status !== "error" && t && (await this.generateApiClientStep.execute({
			appPath: this.state.appPath,
			credentials: this.registerAppStep.registrationCredentials
		}), this.skipTypecheck = !1);
	}
	async initializePipeline(e) {
		await this.registerAppStep.execute({ manifest: e });
		let t = await this.apiService.createDevelopmentApplication({
			universalIdentifier: e.application.universalIdentifier,
			name: e.application.displayName
		});
		return !t.success || !t.data ? (this.state.applyStepEvents([{
			message: "Failed to install development application",
			status: "error"
		}, {
			message: JSON.stringify(t, null, 2),
			status: "error"
		}]), this.state.updatePipeline({ status: "error" }), !1) : (this.state.steps.resolveApplication.output = {
			applicationId: t.data.id,
			universalIdentifier: t.data.universalIdentifier
		}, this.state.steps.resolveApplication.status = "done", this.state.applyStepEvents([{
			message: "Application installed",
			status: "success"
		}]), this.uploadFilesStep.initialize({
			appPath: this.state.appPath,
			universalIdentifier: e.application.universalIdentifier
		}), !0);
	}
}, on = {
	objects: S.Object,
	fields: S.Field,
	logicFunctions: S.LogicFunction,
	frontComponents: S.FrontComponent,
	roles: S.Role,
	skills: S.Skill,
	connectionProviders: S.ConnectionProvider,
	views: S.View,
	viewFields: S.ViewField,
	navigationMenuItems: S.NavigationMenuItem,
	pageLayouts: S.PageLayout,
	pageLayoutTabs: S.PageLayoutTab,
	commandMenuItems: S.CommandMenuItem
}, sn = 200, cn = {
	pending: [
		"building",
		"uploading",
		"success",
		"error"
	],
	building: [
		"pending",
		"uploading",
		"success",
		"error"
	],
	uploading: [
		"pending",
		"success",
		"error"
	],
	success: [
		"pending",
		"building",
		"uploading",
		"error"
	],
	error: [
		"pending",
		"building",
		"uploading",
		"success"
	]
}, ln = class {
	constructor(e) {
		this.eventIdCounter = 0, this.appPath = e.appPath, this.previousObjectsFieldsFingerprint = null, this.steps = {
			checkServer: {
				output: {
					isReady: !1,
					errorLogged: !1
				},
				status: "idle"
			},
			ensureValidTokens: {
				output: {},
				status: "idle"
			},
			resolveApplication: {
				output: {
					applicationId: null,
					universalIdentifier: null
				},
				status: "idle"
			},
			buildManifest: {
				output: { result: null },
				status: "idle"
			},
			uploadFiles: {
				output: {
					fileUploader: null,
					builtFileInfos: /* @__PURE__ */ new Map(),
					activeUploads: /* @__PURE__ */ new Set()
				},
				status: "idle"
			},
			generateApiClient: {
				output: {},
				status: "idle"
			},
			syncApplication: {
				output: {
					syncStatus: "idle",
					error: null
				},
				status: "idle"
			},
			startWatchers: {
				output: { watchersStarted: !1 },
				status: "idle"
			}
		}, this.pipeline = {
			status: "idle",
			isSyncing: !1,
			error: null,
			appName: null
		}, this.versionInfo = null, this.entities = /* @__PURE__ */ new Map(), this.events = [];
	}
	setVersionInfo(e) {
		this.versionInfo = e, this.notify();
	}
	notify() {
		this.onChange?.();
	}
	updatePipeline(e) {
		Object.assign(this.pipeline, e), this.notify();
	}
	applyStepEvents(e) {
		let t = e.map((e) => (this.eventIdCounter += 1, {
			...e,
			id: this.eventIdCounter,
			timestamp: /* @__PURE__ */ new Date(),
			message: e.message.slice(0, 5e3)
		}));
		this.events = [...this.events.slice(-(sn - t.length)), ...t];
	}
	addEvent(e) {
		this.applyStepEvents([e]);
	}
	updateEntityStatus(e, t) {
		let n = new Map(this.entities), r = n.get(e);
		r?.status && !cn[r.status].includes(t) || (n.set(e, r ? {
			...r,
			status: t
		} : {
			name: e,
			path: e,
			status: t
		}), this.entities = n);
	}
	removeEntity(e) {
		let t = new Map(this.entities);
		t.delete(e), this.entities = t;
	}
	updateAllEntitiesStatus(e) {
		let t = new Map(this.entities);
		for (let [n, r] of t) t.set(n, {
			...r,
			status: e
		});
		this.entities = t;
	}
	updateEntitiesFromManifest(e) {
		let t = /* @__PURE__ */ new Map();
		for (let [n, r] of Object.entries(e)) {
			let e = on[n];
			if (e) for (let n of r) t.set(n, e);
		}
		let n = new Map(this.entities);
		for (let [e, r] of n) n.set(e, {
			...r,
			type: t.get(e)
		});
		for (let [e, r] of t) n.has(e) || n.set(e, {
			name: e,
			path: e,
			type: r,
			status: "pending"
		});
		this.entities = n;
	}
	hasObjectsOrFieldsChanged(e) {
		let t = JSON.stringify({
			objects: e.objects,
			fields: e.fields
		}), n = t !== this.previousObjectsFieldsFingerprint;
		return this.previousObjectsFieldsFingerprint = t, n;
	}
}, P = {
	idle: {
		color: "gray",
		icon: "○"
	},
	in_progress: {
		color: "yellow",
		icon: "spinner"
	},
	uploading: {
		color: "cyan",
		icon: "upload"
	},
	done: {
		color: "green",
		icon: "✓"
	},
	error: {
		color: "red",
		icon: "✗"
	}
}, un = (e) => ({
	idle: "idle",
	in_progress: "in_progress",
	done: "done",
	error: "error"
})[e], dn = (e) => ({
	pending: "idle",
	building: "in_progress",
	uploading: "uploading",
	success: "done",
	error: "error"
})[e], F = (e) => ({
	idle: "idle",
	building: "in_progress",
	syncing: "in_progress",
	synced: "done",
	error: "error"
})[e], I = {
	idle: "Idle",
	building: "Building...",
	syncing: "Syncing...",
	synced: "Synced",
	error: "Error"
}, L = [
	"◐",
	"◓",
	"◑",
	"◒"
], fn = [
	"↑",
	"⇡",
	"↑",
	"⇡"
], pn = {
	[S.Object]: "Objects",
	[S.Field]: "Fields",
	[S.LogicFunction]: "Logic functions",
	[S.FrontComponent]: "Front components",
	[S.CommandMenuItem]: "Command menu items",
	[S.Role]: "Roles",
	[S.Skill]: "Skills",
	[S.View]: "Views",
	[S.ViewField]: "View fields",
	[S.NavigationMenuItem]: "Navigation menu items",
	[S.PageLayout]: "Page layouts",
	[S.PageLayoutTab]: "Page layout tabs",
	[S.Agent]: "Agents",
	[S.ConnectionProvider]: "Connection providers"
}, mn = Object.keys(pn), hn = {
	info: "gray",
	success: "green",
	error: "red",
	warning: "yellow"
}, gn = (e) => e.toLocaleTimeString("en-US", {
	hour12: !1,
	hour: "2-digit",
	minute: "2-digit",
	second: "2-digit"
}), _n = (e, t = 40) => {
	if (e.length <= t) return e;
	let n = e.split("/");
	return n.length <= 2 ? e : `.../${n.slice(-2).join("/")}`;
}, vn = (e) => {
	let t = /* @__PURE__ */ new Map();
	for (let e of mn) t.set(e, []);
	for (let n of e.values()) {
		if (!n.type) continue;
		let e = t.get(n.type) ?? [];
		e.push(n), t.set(n.type, e);
	}
	return t;
}, R = [
	{
		status: "success",
		label: "synced"
	},
	{
		status: "building",
		label: "building"
	},
	{
		status: "uploading",
		label: "uploading"
	},
	{
		status: "pending",
		label: "pending"
	},
	{
		status: "error",
		label: "error"
	}
], yn = (e) => {
	let t = {
		pending: 0,
		building: 0,
		uploading: 0,
		success: 0,
		error: 0
	};
	for (let n of e) t[n.status] += 1;
	return R.filter(({ status: e }) => t[e] > 0).map(({ status: e, label: n }) => ({
		status: e,
		count: t[e],
		label: n
	}));
}, bn = (e) => {
	let t = e.steps.resolveApplication.output.applicationId;
	return !Fe(e.frontendUrl) || !Fe(t) ? null : `${e.frontendUrl}/settings/applications/${t}`;
}, xn = (e) => e.some((e) => e === "error") ? "error" : e.some((e) => e === "in_progress") ? "in_progress" : e.every((e) => e === "done") ? "done" : "idle", Sn = (e) => {
	let t = [...e.entities.values()], n = t.some((e) => e.status === "error"), r = t.some((e) => e.status === "building"), i = t.length > 0 && t.every((e) => e.status === "uploading" || e.status === "success"), a = n ? "error" : r ? "in_progress" : i ? "done" : "idle";
	return [
		{
			label: "Application Initialization",
			status: xn([
				e.steps.checkServer.status,
				e.steps.ensureValidTokens.status,
				e.steps.resolveApplication.status
			])
		},
		{
			label: "Resources Build",
			status: a
		},
		{
			label: "Resources Upload",
			status: e.steps.uploadFiles.status
		},
		{
			label: "Manifest Build",
			status: e.steps.buildManifest.status
		},
		{
			label: "Application Synchronization",
			status: e.steps.syncApplication.status
		},
		{
			label: "Api Client Generation",
			status: e.steps.generateApiClient.status
		}
	];
}, z = 120, Cn = Math.round(200 / z), wn = () => {
	let e = Math.floor(Date.now() / z);
	return {
		spinnerFrame: L[e % L.length],
		uploadFrame: fn[Math.floor(e / Cn) % fn.length]
	};
}, Tn = (e) => {
	let t = P[e];
	return t.icon === "spinner" ? wn().spinnerFrame : t.icon === "upload" ? wn().uploadFrame : t.icon;
}, En = Ge.createContext(null), Dn = En.Provider, On = () => {
	let e = Ge.useContext(En);
	if (!e) throw Error("useInk must be used within InkProvider");
	return e;
}, kn = ({ uiStatus: e }) => {
	let { Text: t } = On(), n = Tn(e), r = P[e];
	return /* @__PURE__ */ k(t, {
		color: r.color,
		children: [n, " "]
	});
}, An = ({ entity: e }) => {
	let { Box: t, Text: n } = On();
	return /* @__PURE__ */ k(t, { children: [
		/* @__PURE__ */ O(kn, { uiStatus: dn(e.status) }),
		/* @__PURE__ */ O(n, { children: e.name }),
		e.path !== e.name && /* @__PURE__ */ k(n, {
			dimColor: !0,
			children: [
				" (",
				_n(e.path),
				")"
			]
		})
	] });
}, jn = ({ type: e, entities: t }) => {
	let { Box: n, Text: r } = On();
	return t.length === 0 ? null : /* @__PURE__ */ k(n, {
		flexDirection: "column",
		marginTop: 1,
		children: [/* @__PURE__ */ O(r, {
			bold: !0,
			dimColor: !0,
			children: pn[e]
		}), t.map((e) => /* @__PURE__ */ O(An, { entity: e }, e.path))]
	});
}, Mn = ({ entities: e }) => {
	let { Box: t, Text: n } = On();
	if (e.length === 0) return null;
	let r = yn(e);
	return /* @__PURE__ */ k(t, {
		marginTop: 1,
		children: [/* @__PURE__ */ k(n, {
			bold: !0,
			dimColor: !0,
			children: ["Entities", " "]
		}), r.map((e, r) => /* @__PURE__ */ k(t, { children: [
			r > 0 && /* @__PURE__ */ O(n, {
				dimColor: !0,
				children: " · "
			}),
			/* @__PURE__ */ O(kn, { uiStatus: dn(e.status) }),
			/* @__PURE__ */ k(n, { children: [
				e.count,
				" ",
				e.label
			] })
		] }, e.status))]
	});
}, Nn = () => {
	let { Box: e, Text: t } = On();
	return /* @__PURE__ */ O(e, {
		marginTop: 1,
		children: /* @__PURE__ */ k(t, {
			dimColor: !0,
			children: [
				/* @__PURE__ */ O(t, {
					color: P.idle.color,
					children: P.idle.icon
				}),
				" ",
				"pending",
				" ",
				/* @__PURE__ */ O(t, {
					color: P.in_progress.color,
					children: L[0]
				}),
				" ",
				"building",
				" ",
				/* @__PURE__ */ O(t, {
					color: P.uploading.color,
					children: fn[0]
				}),
				" ",
				"uploading",
				" ",
				/* @__PURE__ */ O(t, {
					color: P.done.color,
					children: P.done.icon
				}),
				" ",
				"success",
				" ",
				/* @__PURE__ */ O(t, {
					color: P.error.color,
					children: P.error.icon
				}),
				" ",
				"error"
			]
		})
	});
}, Pn = 7, Fn = ({ versionInfo: e }) => {
	let { Box: t, Text: n } = On();
	if (e === null) return null;
	let { cliVersion: r, localServerVersion: i, latestServerVersion: a, isMinorOrMajorBehind: o, daysBehind: s } = e, c = o && s !== null && s > Pn;
	return /* @__PURE__ */ k(t, { children: [
		/* @__PURE__ */ O(n, {
			dimColor: !0,
			children: "Versions: "
		}),
		/* @__PURE__ */ O(n, {
			dimColor: !0,
			children: "CLI "
		}),
		/* @__PURE__ */ k(n, { children: ["v", r] }),
		i !== null && /* @__PURE__ */ k(rt, { children: [
			/* @__PURE__ */ O(n, {
				dimColor: !0,
				children: " • Server "
			}),
			/* @__PURE__ */ k(n, {
				color: c ? "yellow" : void 0,
				children: ["v", i]
			}),
			i !== null && a !== null && i === a && /* @__PURE__ */ O(n, {
				color: "green",
				children: " ✓ latest"
			}),
			c && a !== null && /* @__PURE__ */ k(rt, { children: [/* @__PURE__ */ k(n, {
				color: "yellow",
				children: [" → v", a]
			}), s !== null && /* @__PURE__ */ k(n, {
				dimColor: !0,
				children: [
					" (",
					s,
					"d behind)"
				]
			})] })
		] })
	] });
}, In = ({ state: e }) => {
	let { Text: t } = On(), n = F(e.pipeline.status), r = Tn(n), i = P[n], a = I[e.pipeline.status];
	return /* @__PURE__ */ k(t, {
		color: i.color,
		children: [
			r,
			" ",
			a
		]
	});
}, B = ({ label: e, status: t }) => {
	let { Box: n, Text: r } = On(), i = un(t), a = Tn(i), o = P[i];
	return /* @__PURE__ */ k(n, { children: [/* @__PURE__ */ k(r, {
		dimColor: !0,
		children: [e, ": "]
	}), /* @__PURE__ */ k(r, {
		color: o.color,
		children: [
			a,
			" ",
			t.replace("_", " ")
		]
	})] });
}, Ln = ({ state: e, verbose: t = !1 }) => {
	let { Box: n, Text: r } = On(), i = vn(e.entities), a = bn(e);
	return /* @__PURE__ */ k(n, {
		flexDirection: "column",
		borderStyle: "classic",
		borderColor: "gray",
		paddingX: 1,
		children: [
			/* @__PURE__ */ O(r, {
				bold: !0,
				color: "cyan",
				children: "Application"
			}),
			/* @__PURE__ */ k(n, {
				marginLeft: 2,
				flexDirection: "column",
				children: [
					/* @__PURE__ */ k(n, { children: [/* @__PURE__ */ O(r, {
						dimColor: !0,
						children: "Name: "
					}), /* @__PURE__ */ O(r, {
						bold: !0,
						children: e.pipeline.appName ?? "Loading..."
					})] }),
					/* @__PURE__ */ k(n, { children: [/* @__PURE__ */ O(r, {
						dimColor: !0,
						children: "Overall Status: "
					}), /* @__PURE__ */ O(In, { state: e })] }),
					a && /* @__PURE__ */ k(n, { children: [/* @__PURE__ */ O(r, {
						dimColor: !0,
						children: "Open:"
					}), /* @__PURE__ */ k(r, {
						bold: !0,
						color: "cyan",
						children: [" ", a]
					})] }),
					/* @__PURE__ */ O(Fn, { versionInfo: e.versionInfo })
				]
			}),
			/* @__PURE__ */ O(n, {
				marginLeft: 2,
				flexDirection: "column",
				marginTop: 1,
				children: Sn(e).map((e) => /* @__PURE__ */ O(B, {
					label: e.label,
					status: e.status
				}, e.label))
			}),
			/* @__PURE__ */ O(n, {
				marginLeft: 2,
				flexDirection: "column",
				children: t ? mn.map((e) => /* @__PURE__ */ O(jn, {
					type: e,
					entities: i.get(e) ?? []
				}, e)) : /* @__PURE__ */ O(Mn, { entities: Array.from(e.entities.values()) })
			})
		]
	});
}, Rn = ({ event: e }) => {
	let { Box: t, Text: n } = On(), r = hn[e.status], i = gn(e.timestamp);
	return /* @__PURE__ */ k(t, {
		marginTop: +!!e.spacingBefore,
		children: [/* @__PURE__ */ k(n, {
			dimColor: !0,
			children: [i, " "]
		}), /* @__PURE__ */ O(n, {
			color: r,
			children: e.message
		})]
	});
}, zn = new Set(["building", "syncing"]), Bn = 120, Vn = 80, Hn = ({ uiStateManager: e, verbose: t }) => {
	let { Box: n, Static: r } = On(), [, i] = Je((e) => e + 1, 0), a = Ye(null), o = Ye(0), s = Ke(() => {
		a.current && clearTimeout(a.current), a.current = setTimeout(() => {
			a.current = null, o.current = Date.now(), i();
		}, Vn);
	}, []);
	qe(() => e.subscribe(() => {
		s();
	}), [e, s]), qe(() => {
		let t = setInterval(() => {
			let t = e.getSnapshot();
			zn.has(t.pipeline.status) && (Date.now() - o.current < Bn || i());
		}, Bn);
		return () => clearInterval(t);
	}, [e]), qe(() => () => {
		a.current && clearTimeout(a.current);
	}, []);
	let c = e.getSnapshot();
	return /* @__PURE__ */ k(rt, { children: [/* @__PURE__ */ O(r, {
		items: c.events,
		children: (e) => /* @__PURE__ */ O(Rn, { event: e }, e.id)
	}), /* @__PURE__ */ k(n, {
		marginTop: 1,
		flexDirection: "column",
		children: [/* @__PURE__ */ O(Ln, {
			state: c,
			verbose: t
		}), t && /* @__PURE__ */ O(Nn, {})]
	})] });
}, Un = async (e, t = !1) => {
	let { render: n, Box: r, Text: i, Static: a } = await import("ink"), { unmount: o } = n(/* @__PURE__ */ O(Dn, {
		value: {
			Box: r,
			Text: i,
			Static: a
		},
		children: /* @__PURE__ */ O(Hn, {
			uiStateManager: e,
			verbose: t
		})
	}));
	return { unmount: o };
}, Wn = class {
	constructor(e) {
		this.listeners = /* @__PURE__ */ new Set(), this.pendingNotify = !1, this.orchestratorState = e;
	}
	getSnapshot() {
		return this.orchestratorState;
	}
	subscribe(e) {
		return this.listeners.add(e), e(this.orchestratorState), () => this.listeners.delete(e);
	}
	notify() {
		this.pendingNotify || (this.pendingNotify = !0, queueMicrotask(() => {
			this.pendingNotify = !1;
			for (let e of this.listeners) e(this.orchestratorState);
		}));
	}
}, Gn = class {
	constructor() {
		this.orchestrator = null, this.unmountUI = null;
	}
	async close() {
		this.unmountUI?.(), await this.orchestrator?.close();
	}
	getOrchestrator() {
		return this.orchestrator;
	}
	async ensureAuthenticatedBeforeLaunch() {
		let { serverUp: e, authValid: t } = await new l({ disableInterceptors: !0 }).validateAuth();
		e && !t && await ze(_.getActiveRemote());
	}
	async execute(e) {
		let t = e.appPath ?? w;
		await lt(t), e.headless && await ee(), await this.ensureAuthenticatedBeforeLaunch();
		let n = new ln({ appPath: t });
		if (!e.headless) {
			let t = new Wn(n);
			n.onChange = () => t.notify();
			let { unmount: r } = await Un(t, e.verbose);
			this.unmountUI = r, Pe().then((e) => n.setVersionInfo(e));
		}
		this.orchestrator = new an({
			state: n,
			verbose: e.verbose,
			debounceMs: e.debounceMs
		}), await this.orchestrator.start(), e.headless || this.setupGracefulShutdown();
	}
	setupGracefulShutdown() {
		let e = () => void this.close().then(() => process.exit(0));
		process.on("SIGINT", e), process.on("SIGTERM", e);
	}
}, Kn = class {
	async execute(e) {
		let t = e.appPath ?? w;
		await lt(t);
		let n = _.getActiveRemote();
		console.log(D.blue(`${e.dryRun ? "Previewing application diff" : "Syncing application"} on ${n}...`)), console.log(D.gray(`App path: ${t}\n`));
		let r = await Ce({
			appPath: t,
			verbose: e.verbose,
			dryRun: e.dryRun,
			onProgress: (e) => console.log(D.gray(e))
		});
		if (r.success || (console.error(D.red(r.error.message)), process.exit(1)), e.dryRun) {
			console.log(D.green(`\n✓ Dry run complete for ${r.data.applicationDisplayName} — no changes were applied`));
			return;
		}
		console.log(D.green(`\n✓ Synced ${r.data.applicationDisplayName} (${r.data.fileCount} file${r.data.fileCount === 1 ? "" : "s"})`)), console.log(D.gray(`Output: ${r.data.outputDir}`));
	}
}, qn = class {
	async execute(e) {
		let t = e.appPath ?? w, n = T(t, "node_modules", "twenty-client-sdk");
		await re(n) || (console.error(D.red(`Cannot find twenty-client-sdk in ${t}.\n\n  Install it first:
    yarn add twenty-client-sdk`)), process.exit(1));
		let r = await new l({ disableInterceptors: !0 }).validateAuth();
		r.serverUp || (console.error(D.red("Cannot reach Twenty server.\n\n  Check your remotes:\n    yarn twenty remote:status")), process.exit(1)), r.authValid || (console.error(D.red("Authentication failed. Run `yarn twenty remote:add` to authenticate.")), process.exit(1)), console.log(D.blue("Generating API client..."));
		try {
			await new ye({ skipAuth: !1 }).generateCoreClient({ appPath: t });
		} catch (e) {
			console.error(D.red(`Failed to generate API client: ${m(e)}`)), process.exit(1);
		}
		console.log(D.green("✓ API client generated")), console.log(D.gray(`Output: ${T(n, "dist", "core", "generated")}`));
	}
}, Jn = (e) => {
	let t = new Ht(), n = new Vt();
	e.command("dev:function:logs [appPath]").description("Stream logic function logs").option("-u, --functionUniversalIdentifier <functionUniversalIdentifier>", "Only show logs for the function with this universal ID").option("-n, --functionName <functionName>", "Only show logs for the function with this name").action(async (e, n) => {
		await t.execute({
			...n,
			appPath: A(e)
		});
	}), e.command("dev:function:exec [appPath]").description("Execute a logic function").option("--postInstall", "Execute post-install logic function if defined").option("--preInstall", "Execute pre-install logic function if defined").option("-p, --payload <payload>", "JSON payload to send to the function", "{}").option("-u, --functionUniversalIdentifier <functionUniversalIdentifier>", "Universal ID of the function to execute").option("-n, --functionName <functionName>", "Name of the function to execute").action(async (e, t) => {
		!t?.postInstall && !t?.preInstall && !t?.functionUniversalIdentifier && !t?.functionName && (console.error(D.red("Error: Either --postInstall, --preInstall, --functionName (-n), or --functionUniversalIdentifier (-u) is required.")), process.exit(1)), await n.execute({
			...t,
			payload: t?.payload ?? "{}",
			appPath: A(e)
		});
	});
}, Yn = (e) => {
	let t = new Bt(), n = new Gn(), r = new Kn(), i = new Wt(), a = new zt(), o = new qn();
	e.command("dev [appPath]").description("Build and sync local changes").option("-o, --once", "Build and sync once, then exit (useful for CI, scripts, and pre-commit hooks)").option("--dry-run", "Preview the metadata changes without applying them (requires --once)").option("--debounceMs <ms>", "Debounce in ms (default: 1 000)").option("-v, --verbose", "Show detailed logs").option("-d, --debug", "Show detailed logs (alias for --verbose)").action(async (e, t) => {
		t.dryRun && !t.once && console.warn(D.yellow("--dry-run only applies with --once. Ignoring it; run `yarn twenty dev --once --dry-run` to preview changes."));
		let i = {
			appPath: A(e),
			verbose: t.verbose || t.debug,
			debounceMs: t.debounceMs ? parseInt(t.debounceMs, 10) : void 0
		};
		if (t.once) {
			await r.execute({
				...i,
				dryRun: t.dryRun
			});
			return;
		}
		await n.execute(i);
	}), e.command("dev:build [appPath]").description("Build and generate API client").option("--tarball", "Also pack into a .tgz tarball").action(async (e, n) => {
		await t.execute({
			appPath: A(e),
			tarball: n.tarball
		});
	}), e.command("dev:typecheck [appPath]").description("Run TypeScript type checking").action(async (e) => {
		await i.execute({ appPath: A(e) });
	}), e.command("dev:add [entityType]").description(`Scaffold a new entity (${Object.values(S).join("|")})`).option("--path <path>", "Path in which the entity should be created.").action(async (e, t) => {
		await a.execute(e, t?.path);
	}), e.command("dev:catalog-sync").description("Trigger marketplace catalog sync").option("-r, --remote <name>", "Sync on a specific remote").action(async (e) => {
		let { CatalogSyncCommand: t } = await import("./catalog-sync-BYM3GqZl.mjs");
		await new t().execute({ remote: e.remote });
	}), e.command("dev:generate-client [appPath]").description("Generate the typed API client from the active remote (no app definition required)").action(async (e) => {
		await o.execute({ appPath: A(e) });
	}), Jn(e);
}, Xn = async (e, t) => {
	let n = t.test ? Le : ie, r = t.port ? parseInt(t.port, 10) : n;
	(isNaN(r) || r < 1 || r > 65535) && (console.error(D.red("Invalid port number.")), process.exit(1));
	let i = await xe({
		port: r,
		test: t.test,
		version: e,
		onProgress: (e) => console.log(D.gray(e))
	});
	i.success || (console.error(D.red(i.error.message)), process.exit(1));
}, Zn = (e) => {
	let t = e.test ? h : Re;
	if (!Ie(t)) {
		console.log(D.yellow("No Twenty server container found."));
		return;
	}
	Qe(`docker stop ${t}`, { stdio: "ignore" }), console.log(D.green("Twenty server stopped."));
}, Qn = (e) => {
	let t = e.test ? h : Re;
	if (!Ie(t)) {
		console.log(D.yellow("No Twenty server container found."));
		return;
	}
	try {
		et("docker", [
			"logs",
			"-f",
			"--tail",
			e.lines,
			t
		], { stdio: "inherit" });
	} catch {}
}, $n = async (e) => {
	let n = e.test ? h : Re, a = e.test ? Le : ie;
	if (!Ie(n)) {
		console.log(`  Status:  ${D.gray("not created")}`), console.log(D.gray(`  Run 'yarn twenty docker:start${e.test ? " --test" : ""}' to create one.`));
		return;
	}
	let o = t(n), s = o ? r(n) : a, c = o ? await ue(s) : !1, l = c ? D.green("running (healthy)") : o ? D.yellow("running (starting...)") : D.gray("stopped"), u = i("APP_VERSION", n);
	console.log(`  Status:  ${l}`), console.log(`  URL:     http://localhost:${s}`), u && console.log(`  Version: ${D.gray(u)}`), c && console.log(D.gray("  Login:   tim@apple.dev / tim@apple.dev"));
}, er = (e) => {
	let t = e.test ? h : Re, n = e.test ? "twenty-app-dev-test-data" : "twenty-app-dev-data", r = e.test ? "twenty-app-dev-test-storage" : "twenty-app-dev-storage";
	Ie(t) && Qe(`docker rm -f ${t}`, { stdio: "ignore" });
	try {
		Qe(`docker volume rm ${n} ${r}`, { stdio: "ignore" });
	} catch {}
	console.log(D.green("Twenty server data reset.")), console.log(D.gray(`Run 'yarn twenty docker:start${e.test ? " --test" : ""}' to start a fresh instance.`));
}, tr = async (e, t) => {
	let n = await ne({
		version: e ?? "latest",
		test: t.test,
		onProgress: (e) => console.log(D.gray(e))
	});
	n.success || (console.error(D.red(n.error.message)), process.exit(1));
	let { data: r } = n;
	if (!r.imageUpdated) {
		console.log(D.green(`  Already up to date (${r.image}).`));
		return;
	}
	console.log(D.green(`  Upgraded to: ${r.image}`)), r.containerRecreated && console.log(D.gray(`  Run 'yarn twenty docker:start${t.test ? " --test" : ""}' to wait for the server to be ready.`));
}, nr = (e) => {
	e.command("docker:start [version]").description("Start the local Twenty container (version defaults to the app `engines.twenty` range, then `latest`)").option("-p, --port <port>", "HTTP port").option("--test", "Start a separate test instance (port 2021)").action(Xn), e.command("docker:stop").description("Stop the local Twenty container").option("--test", "Stop the test instance").action(Zn), e.command("docker:logs").description("Stream container logs").option("-n, --lines <lines>", "Number of lines to show", "50").option("--test", "Show logs for the test instance").action(Qn), e.command("docker:status").description("Show container status").option("--test", "Show status of the test instance").action($n), e.command("docker:reset").description("Delete all data and start fresh").option("--test", "Reset the test instance").action(er), e.command("docker:upgrade [version]").description("Upgrade the Docker image").option("--test", "Upgrade the test instance").action(tr);
	let t = e.command("server", { hidden: !0 }).description("Manage a Twenty server (local instance and server-side actions)"), n = (e, t) => console.warn(D.yellow(`⚠ \`twenty server ${e}\` is deprecated. Use \`twenty ${t}\` instead.`));
	t.command("start [version]").option("-p, --port <port>", "HTTP port").option("--test", "Start a separate test instance (port 2021)").action(async (e, t) => {
		n("start", "docker:start"), await Xn(e, t);
	}), t.command("stop").option("--test", "Stop the test instance").action((e) => {
		n("stop", "docker:stop"), Zn(e);
	}), t.command("logs").option("-n, --lines <lines>", "Number of lines to show", "50").option("--test", "Show logs for the test instance").action((e) => {
		n("logs", "docker:logs"), Qn(e);
	}), t.command("status").option("--test", "Show status of the test instance").action(async (e) => {
		n("status", "docker:status"), await $n(e);
	}), t.command("reset").option("--test", "Reset the test instance").action((e) => {
		n("reset", "docker:reset"), er(e);
	}), t.command("upgrade [version]").option("--test", "Upgrade the test instance").action(async (e, t) => {
		n("upgrade", "docker:upgrade"), await tr(e, t);
	}), t.command("catalog-sync").option("-r, --remote <name>", "Sync on a specific remote").action(async (e) => {
		n("catalog-sync", "dev:catalog-sync");
		let { CatalogSyncCommand: t } = await import("./catalog-sync-BYM3GqZl.mjs");
		await new t().execute({ remote: e.remote });
	});
}, rr = (e) => {
	try {
		return new URL(e).hostname.replace(/\./g, "-");
	} catch {
		return "remote";
	}
}, ir = async (e, t) => t ? ((await Ae({
	apiKey: t,
	apiUrl: e
})).success || (console.error(D.red("✗ Authentication failed.")), process.exit(1)), "API key") : ar(e), ar = async (e) => {
	console.log(D.gray("Opening browser for authentication..."));
	let t = await n({ apiUrl: e });
	return t.success ? "OAuth" : (console.log(D.yellow(t.error.message)), (await Ae({
		apiKey: (await tt.prompt([{
			type: "password",
			name: "apiKey",
			message: "API Key:",
			mask: "*",
			validate: (e) => e.length > 0 || "API key is required"
		}])).apiKey,
		apiUrl: e
	})).success || (console.error(D.red("✗ Authentication failed.")), process.exit(1)), "API key");
}, or = async (e) => {
	e.apiUrl && console.warn(D.yellow("⚠ --api-url is deprecated. Use --url instead."));
	let t = e.test ? o(!0) : void 0, n = new _(t ? { configPath: t } : void 0), r = await n.getRemotes();
	if (e.as !== void 0 && r.includes(e.as)) {
		let t = await n.getConfigForRemote(e.as);
		_.setActiveRemote(e.as);
		let r = await ir(t.apiUrl, e.apiKey);
		console.log(D.green(`✓ Re-authenticated "${e.as}" via ${r}.`)), await n.setDefaultRemote(e.as), console.log(D.green(`✓ Default remote set to "${e.as}".`));
		return;
	}
	let i = e.url ?? e.apiUrl;
	if (i) i = de(i);
	else {
		let t = await Te();
		e.local ? (t || (console.error(D.red("No local Twenty server found.\nStart one with: yarn twenty docker:start")), process.exit(1)), console.log(D.gray(`Found local server at ${t}`)), i = t) : i = de((await tt.prompt([{
			type: "input",
			name: "serverUrl",
			message: "Twenty server URL:",
			validate: (e) => {
				try {
					return new URL(e), !0;
				} catch {
					return "Please enter a valid URL";
				}
			}
		}])).serverUrl);
	}
	let a = e.as ?? rr(i);
	_.setActiveRemote(a);
	let s = await ir(i, e.apiKey);
	console.log(D.green(`✓ Remote "${a}" added (${i}) via ${s}.`)), await n.setDefaultRemote(a), console.log(D.green(`✓ Default remote set to "${a}".`));
}, sr = async () => {
	let e = new _(), t = await e.getRemotes(), n = await e.getDefaultRemote();
	if (t.length === 0) {
		console.log("No remotes configured."), console.log("Use 'twenty remote:add' to add one.");
		return;
	}
	console.log("");
	for (let r of t) {
		let t = await e.getConfigForRemote(r), i = t.twentyCLIAccessToken ? "oauth" : t.apiKey ? "api-key" : "none", a = r === n, o = a ? "* " : "  ", s = a ? D.bold(r) : r;
		console.log(`${o}${s}  ${D.gray(t.apiUrl)}  [${i}]`);
	}
	console.log("\n", D.gray("Use 'twenty remote:use <name>' to change default"));
}, cr = async (e) => {
	let t = new _(), n = e ?? (await tt.prompt([{
		type: "select",
		name: "remote",
		message: "Select default remote:",
		choices: await t.getRemotes()
	}])).remote;
	(await t.getRemotes()).includes(n) || (console.error(D.red(`Remote "${n}" not found.`)), process.exit(1)), await t.setDefaultRemote(n), console.log(D.green(`✓ Default remote set to "${n}".`));
}, lr = async () => {
	let e = new _(), t = new l(), n = _.getActiveRemote(), r = await e.getConfig(), i = r.twentyCLIAccessToken ? "oauth" : r.apiKey ? "api-key" : "none";
	if (console.log(`  Remote:  ${D.bold(n)}`), console.log(`  Server:  ${r.apiUrl}`), i === "none") {
		console.log(`  Auth:    ${D.yellow("not configured")}`);
		return;
	}
	let { authValid: a } = await t.validateAuth(), o = a ? D.green(`${i} (valid)`) : D.red(`${i} (invalid)`);
	console.log(`  Auth:    ${o}`);
}, ur = async (e) => {
	let t = new _();
	(await t.getRemotes()).includes(e) || (console.error(D.red(`Remote "${e}" not found.`)), process.exit(1)), _.setActiveRemote(e), await t.clearConfig(), console.log(D.green(`✓ Remote "${e}" removed.`));
}, dr = (e) => {
	e.command("remote:add").description("Add or re-authenticate a remote").option("--as <name>", "Name for this remote").option("--api-key <apiKey>", "API key for non-interactive auth").option("--url <url>", "Server URL").option("--api-url <apiUrl>", "[deprecated: use --url]").option("--local", "Connect to a local Twenty server (auto-detect)").option("--test", "Write to config.test.json (for integration tests)").action(or), e.command("remote:list").description("List all configured remotes").action(sr), e.command("remote:use [name]").description("Set the default remote").action(cr), e.command("remote:status").description("Show active remote auth status").action(lr), e.command("remote:remove <name>").description("Remove a remote").action(ur);
	let t = e.command("remote", { hidden: !0 }).description("Manage remote Twenty servers"), n = (e, t) => console.warn(D.yellow(`⚠ \`twenty remote ${e}\` is deprecated. Use \`twenty ${t}\` instead.`));
	t.command("add").option("--as <name>", "Name for this remote").option("--api-key <apiKey>", "API key for non-interactive auth").option("--url <url>", "Server URL").option("--api-url <apiUrl>", "[deprecated: use --url]").option("--local", "Connect to a local Twenty server (auto-detect)").option("--test", "Write to config.test.json (for integration tests)").action(async (e) => {
		n("add", "remote:add"), await or(e);
	}), t.command("list").action(async () => {
		n("list", "remote:list"), await sr();
	}), t.command("switch [name]").action(async (e) => {
		n("switch", "remote:use"), await cr(e);
	}), t.command("status").action(async () => {
		n("status", "remote:status"), await lr();
	}), t.command("remove <name>").action(async (e) => {
		n("remove", "remote:remove"), await ur(e);
	});
}, fr = (e) => {
	Yn(e), pt(e), nr(e), dr(e), Kt(e);
};
//#endregion
//#region src/cli/cli.ts
ot.defaultOptions.depth = 10;
var pr = new it();
pr.name("twenty").description("CLI for Twenty application development").version(f.version), pr.option("-r, --remote <name>", "Use a specific remote (overrides the default set by remote:use)"), pr.hook("preAction", async (e) => {
	let t = (e.optsWithGlobals ? e.optsWithGlobals() : e.opts()).remote;
	t ? console.log(D.gray(`Using remote: ${t}`)) : t = await new _().getDefaultRemote(), _.setActiveRemote(t);
}), fr(pr), pr.exitOverride();
try {
	pr.parse();
} catch (e) {
	e instanceof at && process.exit(e.exitCode), e instanceof Error && (console.error(D.red("Error:"), e.message), process.exit(1));
}
//#endregion
