// twenty-sdk/dist/get-front-component-build-plugins-DBupR4cF.js
var e=Object.create,t=Object.defineProperty,n=Object.getOwnPropertyDescriptor,r=Object.getOwnPropertyNames,i=Object.getPrototypeOf,a=Object.prototype.hasOwnProperty,o=(e,t)=>()=>(t||(e((t={exports:{}}).exports,t),e=null),t.exports),s=(e,n)=>{let r={};for(var i in e)t(r,i,{get:e[i],enumerable:!0});return n||t(r,Symbol.toStringTag,{value:`Module`}),r},c=(e,i,o,s)=>{if(i&&typeof i==`object`||typeof i==`function`)for(var c=r(i),l=0,u=c.length,d;l<u;l++)d=c[l],!a.call(e,d)&&d!==o&&t(e,d,{get:(e=>i[e]).bind(null,d),enumerable:!(s=n(i,d))||s.enumerable});return e},l=(n,r,a)=>(a=n==null?{}:e(i(n)),c(r||!n||!n.__esModule?t(a,`default`,{value:n,enumerable:!0}):a,n));let u=require("path");u=l(u);let d=require("node:fs/promises");var f=`export var customElementMap = globalThis.__HTML_TAG_TO_CUSTOM_ELEMENT_TAG__ || {};

var _injectedStyleKeys = {};

export function injectStyleViaHead(cssText) {
  if (!cssText) return;
  var hash = 0;
  for (var i = 0; i < cssText.length; i++) {
    hash = ((hash << 5) - hash + cssText.charCodeAt(i)) | 0;
  }
  var key = 'jsx-style-' + hash;
  if (_injectedStyleKeys[key]) return;
  _injectedStyleKeys[key] = true;
  var el = document.createElement('style');
  el.setAttribute('data-jsx-style', key);
  el.textContent = cssText;
  document.head.appendChild(el);
}

export function extractCssText(children) {
  if (typeof children === 'string') return children;
  if (Array.isArray(children))
    return children
      .filter(function (c) { return typeof c === 'string'; })
      .join('');
  return '';
}

var _reactToDomEvent = {
  ondoubleclick: 'ondblclick',
};

function _isEventProp(name) {
  return (
    name.length > 2 &&
    name.charCodeAt(0) === 111 &&
    name.charCodeAt(1) === 110 &&
    name.charCodeAt(2) >= 65 &&
    name.charCodeAt(2) <= 90
  );
}

export function splitEventProps(props) {
  if (!props) return { cleanProps: props, events: null };
  var events = null;
  var cleanProps = null;
  for (var k in props) {
    if (_isEventProp(k) && typeof props[k] === 'function') {
      if (!events) {
        events = {};
        cleanProps = {};
        for (var j in props) {
          if (j === k) break;
          cleanProps[j] = props[j];
        }
      }
      events[k] = props[k];
    } else if (events) {
      cleanProps[k] = props[k];
    }
  }
  return { cleanProps: cleanProps || props, events: events };
}

export function makeEventRef(events, userRef) {
  return function(el) {
    if (el) {
      for (var name in events) {
        var domName = _reactToDomEvent[name.toLowerCase()] || name.toLowerCase();
        el[domName] = events[name];
      }
    }
    if (typeof userRef === 'function') userRef(el);
    else if (userRef != null && typeof userRef === 'object') userRef.current = el;
  };
}`,p=`import {
  jsx as _originalJsx,
  jsxs as _originalJsxs,
  Fragment,
} from '__real_react_jsx_runtime__';

import {
  customElementMap,
  injectStyleViaHead,
  extractCssText,
  splitEventProps,
  makeEventRef,
} from '__jsx_shared_helpers__';

function _wrapJsxFactory(originalFactory) {
  return function wrappedJsx(type, props, key) {
    if (typeof type === 'string') {
      if (type === 'style') {
        var css =
          props && props.dangerouslySetInnerHTML
            ? props.dangerouslySetInnerHTML.__html || ''
            : extractCssText(props && props.children);
        injectStyleViaHead(css);
        return null;
      }

      var customTag = customElementMap[type];
      if (customTag) {
        var split = splitEventProps(props);
        if (split.events) {
          var cp = split.cleanProps;
          cp.ref = makeEventRef(split.events, cp.ref);
          return originalFactory(customTag, cp, key);
        }
        return originalFactory(customTag, props, key);
      }
    }
    return originalFactory(type, props, key);
  };
}

export var jsx = _wrapJsxFactory(_originalJsx);
export var jsxs = _wrapJsxFactory(_originalJsxs);
export { Fragment };`,m=`export * from '__real_react__';
import _React from '__real_react__';

import {
  customElementMap,
  injectStyleViaHead,
  extractCssText,
  splitEventProps,
  makeEventRef,
} from '__jsx_shared_helpers__';

var _originalCreateElement = _React.createElement;

function createElement(type) {
  var args = arguments;
  if (typeof type === 'string') {
    if (type === 'style') {
      var props = args.length > 1 ? args[1] : null;
      if (props) {
        var css = props.dangerouslySetInnerHTML
          ? props.dangerouslySetInnerHTML.__html || ''
          : extractCssText(props.children);
        injectStyleViaHead(css);
      }
      return null;
    }

    var customTag = customElementMap[type];
    if (customTag) {
      var ceProps = args.length > 1 ? args[1] : null;
      var split = splitEventProps(ceProps);
      if (split.events) {
        var cp = split.cleanProps || {};
        cp.ref = makeEventRef(split.events, cp.ref);
        var newArgs = [customTag, cp];
        for (var i = 2; i < args.length; i++) newArgs.push(args[i]);
        return _originalCreateElement.apply(null, newArgs);
      }
      var newArgs2 = [customTag];
      for (var i2 = 1; i2 < args.length; i2++) newArgs2.push(args[i2]);
      return _originalCreateElement.apply(null, newArgs2);
    }
  }
  return _originalCreateElement.apply(null, args);
}

export { createElement };
export default Object.assign({}, _React, { createElement: createElement });`,h=e=>{let t=e?.usePreact??!1,n=t?`preact/jsx-runtime`:`react/jsx-runtime`,r=t?`preact/compat`:`react`;return{name:`jsx-runtime-remote-wrapper`,setup:e=>{let t,i;e.onResolve({filter:/^__jsx_shared_helpers__$/},()=>({path:`__jsx_shared_helpers__`,namespace:`jsx-shared-helpers`})),e.onLoad({filter:/.*/,namespace:`jsx-shared-helpers`},()=>({contents:f,loader:`js`})),e.onResolve({filter:/^react\/jsx-runtime$/},async r=>{if(!r.pluginData?.skipJsxWrapper)return t||=(await e.resolve(n,{kind:r.kind,resolveDir:r.resolveDir,pluginData:{skipJsxWrapper:!0}})).path,{path:`react/jsx-runtime`,namespace:`jsx-runtime-wrapper`}}),e.onResolve({filter:/^__real_react_jsx_runtime__$/},()=>{if(!t)throw Error(`jsx-runtime-remote-wrapper: real jsx-runtime path not resolved yet`);return{path:t}}),e.onLoad({filter:/.*/,namespace:`jsx-runtime-wrapper`},()=>({contents:p,loader:`js`})),e.onResolve({filter:/^react$/},async t=>{if(!t.pluginData?.skipJsxWrapper)return i||=(await e.resolve(r,{kind:t.kind,resolveDir:t.resolveDir,pluginData:{skipJsxWrapper:!0}})).path,{path:`react`,namespace:`react-wrapper`}}),e.onResolve({filter:/^__real_react__$/},()=>{if(!i)throw Error(`jsx-runtime-remote-wrapper: real react path not resolved yet`);return{path:i}}),e.onLoad({filter:/.*/,namespace:`react-wrapper`},()=>({contents:m,loader:`js`}))}}},g=/import\s*\{\s*defineFrontComponent\s*\}\s*from\s*['"][^'"]+['"];?\n?/g,_=/export\s+default\s+defineFrontComponent\s*\(\s*\{[\s\S]*?component\s*:\s*(\w+)[\s\S]*?\}\s*\)\s*;?/s,v=e=>{let t=e.replace(g,``),n=t.match(_);if(n){let e=n[1],r=RegExp(`export\\s+(const|function)\\s+${e}\\b`);t=t.replace(r,`$1 ${e}`),t=`import { createRoot as __createRoot } from 'react-dom/client';
import { jsx as __frontComponentJsx } from 'react/jsx-runtime';
`+t,t=t.replace(_,`export default function __renderFrontComponent(__container) { __createRoot(__container).render(__frontComponentJsx(${e}, {})); }`)}return t},y={name:`jsx-transform-to-remote-dom-worker-format-plugin`,setup:e=>{e.onLoad({filter:/\.tsx$/},async({path:e})=>{try{return{contents:v(await(0,d.readFile)(e,`utf8`)),loader:`tsx`}}catch(t){return{errors:[{text:`Failed to transform front component: ${t instanceof Error?t.message:String(t)}`,location:{file:e}}]}}})}},b=()=>({name:`preact-alias`,setup:e=>{let t;e.onResolve({filter:/^react-dom\/client$/},async n=>(t||=(await e.resolve(`preact/compat/client`,{kind:n.kind,resolveDir:n.resolveDir})).path,{path:t})),e.onResolve({filter:/^react-dom$/},async t=>({path:(await e.resolve(`preact/compat`,{kind:t.kind,resolveDir:t.resolveDir})).path}))}}),x=/^\/\/.*$\n/gm,S={name:`strip-comments`,setup:e=>{e.onEnd(async e=>{if(e.errors.length>0)return;let t=Object.keys(e.metafile?.outputs??{}).filter(e=>e.endsWith(`.mjs`));for(let e of t){let t=u.default.resolve(e),n=await(0,d.readFile)(t,`utf-8`),r=n.replace(x,``);r!==n&&await(0,d.writeFile)(t,r,`utf-8`)}})}},C=e=>[h(e?.usePreact?{usePreact:!0}:void 0),...e?.usePreact?[b()]:[],y,S];Object.defineProperty(exports,"i",{enumerable:!0,get:function(){return l}}),Object.defineProperty(exports,"n",{enumerable:!0,get:function(){return o}}),Object.defineProperty(exports,"r",{enumerable:!0,get:function(){return s}}),Object.defineProperty(exports,"t",{enumerable:!0,get:function(){return C}});
