// twenty-sdk/dist/get-front-component-build-plugins-BuE2QdIr.mjs
import e from "path";
import { readFile as t, writeFile as n } from "node:fs/promises";
//#region src/cli/utilities/build/common/front-component-build/jsx-runtime-remote-wrapper-plugin.ts
var r = "export var customElementMap = globalThis.__HTML_TAG_TO_CUSTOM_ELEMENT_TAG__ || {};\n\nvar _injectedStyleKeys = {};\n\nexport function injectStyleViaHead(cssText) {\n  if (!cssText) return;\n  var hash = 0;\n  for (var i = 0; i < cssText.length; i++) {\n    hash = ((hash << 5) - hash + cssText.charCodeAt(i)) | 0;\n  }\n  var key = 'jsx-style-' + hash;\n  if (_injectedStyleKeys[key]) return;\n  _injectedStyleKeys[key] = true;\n  var el = document.createElement('style');\n  el.setAttribute('data-jsx-style', key);\n  el.textContent = cssText;\n  document.head.appendChild(el);\n}\n\nexport function extractCssText(children) {\n  if (typeof children === 'string') return children;\n  if (Array.isArray(children))\n    return children\n      .filter(function (c) { return typeof c === 'string'; })\n      .join('');\n  return '';\n}\n\nvar _reactToDomEvent = {\n  ondoubleclick: 'ondblclick',\n};\n\nfunction _isEventProp(name) {\n  return (\n    name.length > 2 &&\n    name.charCodeAt(0) === 111 &&\n    name.charCodeAt(1) === 110 &&\n    name.charCodeAt(2) >= 65 &&\n    name.charCodeAt(2) <= 90\n  );\n}\n\nexport function splitEventProps(props) {\n  if (!props) return { cleanProps: props, events: null };\n  var events = null;\n  var cleanProps = null;\n  for (var k in props) {\n    if (_isEventProp(k) && typeof props[k] === 'function') {\n      if (!events) {\n        events = {};\n        cleanProps = {};\n        for (var j in props) {\n          if (j === k) break;\n          cleanProps[j] = props[j];\n        }\n      }\n      events[k] = props[k];\n    } else if (events) {\n      cleanProps[k] = props[k];\n    }\n  }\n  return { cleanProps: cleanProps || props, events: events };\n}\n\nexport function makeEventRef(events, userRef) {\n  return function(el) {\n    if (el) {\n      for (var name in events) {\n        var domName = _reactToDomEvent[name.toLowerCase()] || name.toLowerCase();\n        el[domName] = events[name];\n      }\n    }\n    if (typeof userRef === 'function') userRef(el);\n    else if (userRef != null && typeof userRef === 'object') userRef.current = el;\n  };\n}", i = "import {\n  jsx as _originalJsx,\n  jsxs as _originalJsxs,\n  Fragment,\n} from '__real_react_jsx_runtime__';\n\nimport {\n  customElementMap,\n  injectStyleViaHead,\n  extractCssText,\n  splitEventProps,\n  makeEventRef,\n} from '__jsx_shared_helpers__';\n\nfunction _wrapJsxFactory(originalFactory) {\n  return function wrappedJsx(type, props, key) {\n    if (typeof type === 'string') {\n      if (type === 'style') {\n        var css =\n          props && props.dangerouslySetInnerHTML\n            ? props.dangerouslySetInnerHTML.__html || ''\n            : extractCssText(props && props.children);\n        injectStyleViaHead(css);\n        return null;\n      }\n\n      var customTag = customElementMap[type];\n      if (customTag) {\n        var split = splitEventProps(props);\n        if (split.events) {\n          var cp = split.cleanProps;\n          cp.ref = makeEventRef(split.events, cp.ref);\n          return originalFactory(customTag, cp, key);\n        }\n        return originalFactory(customTag, props, key);\n      }\n    }\n    return originalFactory(type, props, key);\n  };\n}\n\nexport var jsx = _wrapJsxFactory(_originalJsx);\nexport var jsxs = _wrapJsxFactory(_originalJsxs);\nexport { Fragment };", a = "export * from '__real_react__';\nimport _React from '__real_react__';\n\nimport {\n  customElementMap,\n  injectStyleViaHead,\n  extractCssText,\n  splitEventProps,\n  makeEventRef,\n} from '__jsx_shared_helpers__';\n\nvar _originalCreateElement = _React.createElement;\n\nfunction createElement(type) {\n  var args = arguments;\n  if (typeof type === 'string') {\n    if (type === 'style') {\n      var props = args.length > 1 ? args[1] : null;\n      if (props) {\n        var css = props.dangerouslySetInnerHTML\n          ? props.dangerouslySetInnerHTML.__html || ''\n          : extractCssText(props.children);\n        injectStyleViaHead(css);\n      }\n      return null;\n    }\n\n    var customTag = customElementMap[type];\n    if (customTag) {\n      var ceProps = args.length > 1 ? args[1] : null;\n      var split = splitEventProps(ceProps);\n      if (split.events) {\n        var cp = split.cleanProps || {};\n        cp.ref = makeEventRef(split.events, cp.ref);\n        var newArgs = [customTag, cp];\n        for (var i = 2; i < args.length; i++) newArgs.push(args[i]);\n        return _originalCreateElement.apply(null, newArgs);\n      }\n      var newArgs2 = [customTag];\n      for (var i2 = 1; i2 < args.length; i2++) newArgs2.push(args[i2]);\n      return _originalCreateElement.apply(null, newArgs2);\n    }\n  }\n  return _originalCreateElement.apply(null, args);\n}\n\nexport { createElement };\nexport default Object.assign({}, _React, { createElement: createElement });", o = (e) => {
	let t = e?.usePreact ?? !1, n = t ? "preact/jsx-runtime" : "react/jsx-runtime", o = t ? "preact/compat" : "react";
	return {
		name: "jsx-runtime-remote-wrapper",
		setup: (e) => {
			let t, s;
			e.onResolve({ filter: /^__jsx_shared_helpers__$/ }, () => ({
				path: "__jsx_shared_helpers__",
				namespace: "jsx-shared-helpers"
			})), e.onLoad({
				filter: /.*/,
				namespace: "jsx-shared-helpers"
			}, () => ({
				contents: r,
				loader: "js"
			})), e.onResolve({ filter: /^react\/jsx-runtime$/ }, async (r) => {
				if (!r.pluginData?.skipJsxWrapper) return t ||= (await e.resolve(n, {
					kind: r.kind,
					resolveDir: r.resolveDir,
					pluginData: { skipJsxWrapper: !0 }
				})).path, {
					path: "react/jsx-runtime",
					namespace: "jsx-runtime-wrapper"
				};
			}), e.onResolve({ filter: /^__real_react_jsx_runtime__$/ }, () => {
				if (!t) throw Error("jsx-runtime-remote-wrapper: real jsx-runtime path not resolved yet");
				return { path: t };
			}), e.onLoad({
				filter: /.*/,
				namespace: "jsx-runtime-wrapper"
			}, () => ({
				contents: i,
				loader: "js"
			})), e.onResolve({ filter: /^react$/ }, async (t) => {
				if (!t.pluginData?.skipJsxWrapper) return s ||= (await e.resolve(o, {
					kind: t.kind,
					resolveDir: t.resolveDir,
					pluginData: { skipJsxWrapper: !0 }
				})).path, {
					path: "react",
					namespace: "react-wrapper"
				};
			}), e.onResolve({ filter: /^__real_react__$/ }, () => {
				if (!s) throw Error("jsx-runtime-remote-wrapper: real react path not resolved yet");
				return { path: s };
			}), e.onLoad({
				filter: /.*/,
				namespace: "react-wrapper"
			}, () => ({
				contents: a,
				loader: "js"
			}));
		}
	};
}, s = /import\s*\{\s*defineFrontComponent\s*\}\s*from\s*['"][^'"]+['"];?\n?/g, c = /export\s+default\s+defineFrontComponent\s*\(\s*\{[\s\S]*?component\s*:\s*(\w+)[\s\S]*?\}\s*\)\s*;?/s, l = (e) => {
	let t = e.replace(s, ""), n = t.match(c);
	if (n) {
		let e = n[1], r = RegExp(`export\\s+(const|function)\\s+${e}\\b`);
		t = t.replace(r, `$1 ${e}`), t = "import { createRoot as __createRoot } from 'react-dom/client';\nimport { jsx as __frontComponentJsx } from 'react/jsx-runtime';\n" + t, t = t.replace(c, `export default function __renderFrontComponent(__container) { __createRoot(__container).render(__frontComponentJsx(${e}, {})); }`);
	}
	return t;
}, u = {
	name: "jsx-transform-to-remote-dom-worker-format-plugin",
	setup: (e) => {
		e.onLoad({ filter: /\.tsx$/ }, async ({ path: e }) => {
			try {
				return {
					contents: l(await t(e, "utf8")),
					loader: "tsx"
				};
			} catch (t) {
				return { errors: [{
					text: `Failed to transform front component: ${t instanceof Error ? t.message : String(t)}`,
					location: { file: e }
				}] };
			}
		});
	}
}, d = () => ({
	name: "preact-alias",
	setup: (e) => {
		let t;
		e.onResolve({ filter: /^react-dom\/client$/ }, async (n) => (t ||= (await e.resolve("preact/compat/client", {
			kind: n.kind,
			resolveDir: n.resolveDir
		})).path, { path: t })), e.onResolve({ filter: /^react-dom$/ }, async (t) => ({ path: (await e.resolve("preact/compat", {
			kind: t.kind,
			resolveDir: t.resolveDir
		})).path }));
	}
}), f = /^\/\/.*$\n/gm, p = {
	name: "strip-comments",
	setup: (r) => {
		r.onEnd(async (r) => {
			if (r.errors.length > 0) return;
			let i = Object.keys(r.metafile?.outputs ?? {}).filter((e) => e.endsWith(".mjs"));
			for (let r of i) {
				let i = e.resolve(r), a = await t(i, "utf-8"), o = a.replace(f, "");
				o !== a && await n(i, o, "utf-8");
			}
		});
	}
}, m = (e) => [
	o(e?.usePreact ? { usePreact: !0 } : void 0),
	...e?.usePreact ? [d()] : [],
	u,
	p
];
//#endregion
export { m as t };
