// @sniptt/guards/build/guards/convenience.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isNegativeInteger = exports.isNonNegativeInteger = exports.isPositiveInteger = exports.isInteger = exports.isNumberOrNaN = exports.isNonEmptyString = exports.isNonEmptyArray = exports.isObjectOrNull = void 0;
const primitives_1 = require("./primitives");
const structural_1 = require("./structural");
const isObjectOrNull = (term) => {
    return typeof term === 'object';
};
exports.isObjectOrNull = isObjectOrNull;
const isNonEmptyArray = (term) => {
    return structural_1.isArray(term) && term.length > 0;
};
exports.isNonEmptyArray = isNonEmptyArray;
const isNonEmptyString = (term) => {
    return primitives_1.isString(term) && term.length > 0;
};
exports.isNonEmptyString = isNonEmptyString;
const isNumberOrNaN = (term) => {
    return typeof term === 'number';
};
exports.isNumberOrNaN = isNumberOrNaN;
const isInteger = (term) => {
    return primitives_1.isNumber(term) && Number.isInteger(term);
};
exports.isInteger = isInteger;
const isPositiveInteger = (term) => {
    return exports.isInteger(term) && term > 0;
};
exports.isPositiveInteger = isPositiveInteger;
const isNonNegativeInteger = (term) => {
    return exports.isInteger(term) && term >= 0;
};
exports.isNonNegativeInteger = isNonNegativeInteger;
const isNegativeInteger = (term) => {
    return exports.isInteger(term) && term < 0;
};
exports.isNegativeInteger = isNegativeInteger;
