// @sniptt/guards/build/guards/structural.js
"use strict";
// See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures
Object.defineProperty(exports, "__esModule", { value: true });
exports.isDate = exports.isWeakSet = exports.isWeakMap = exports.isSet = exports.isMap = exports.isArray = exports.isObject = exports.isFunction = exports.isNull = void 0;
const isNull = (term) => {
    return term === null;
};
exports.isNull = isNull;
const isFunction = (term) => {
    return typeof term === 'function';
};
exports.isFunction = isFunction;
const isObject = (term) => {
    return !exports.isNull(term) && typeof term === 'object';
};
exports.isObject = isObject;
const isArray = (term) => {
    return Array.isArray(term);
};
exports.isArray = isArray;
const isMap = (term) => {
    return term instanceof Map;
};
exports.isMap = isMap;
const isSet = (term) => {
    return term instanceof Set;
};
exports.isSet = isSet;
const isWeakMap = (term) => {
    return term instanceof WeakMap;
};
exports.isWeakMap = isWeakMap;
const isWeakSet = (term) => {
    return term instanceof WeakSet;
};
exports.isWeakSet = isWeakSet;
const isDate = (term) => {
    return term instanceof Date;
};
exports.isDate = isDate;
